// node-opcua-pki/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/node-opcua-pki/lib/index.ts
var lib_exports = {};
__export(lib_exports, {
  CertificateAuthority: () => CertificateAuthority,
  CertificateManager: () => CertificateManager,
  CertificateManagerState: () => CertificateManagerState,
  ChainCompletionStatus: () => ChainCompletionStatus,
  Subject: () => import_node_opcua_crypto.Subject,
  VerificationStatus: () => VerificationStatus,
  adjustApplicationUri: () => adjustApplicationUri,
  adjustDate: () => adjustDate,
  coerceCertificateChain: () => coerceCertificateChain,
  convertPFXtoPEM: () => convertPFXtoPEM,
  createPFX: () => createPFX,
  dumpPFX: () => dumpPFX,
  extractAllFromPFX: () => extractAllFromPFX,
  extractCACertificatesFromPFX: () => extractCACertificatesFromPFX,
  extractCertificateFromPFX: () => extractCertificateFromPFX,
  extractPrivateKeyFromPFX: () => extractPrivateKeyFromPFX,
  findIssuerCertificateInChain: () => findIssuerCertificateInChain,
  install_prerequisite: () => install_prerequisite,
  isIntermediateIssuer: () => isIntermediateIssuer,
  isIssuer: () => isIssuer,
  isRootIssuer: () => isRootIssuer,
  makeFingerprint: () => makeFingerprint,
  quote: () => quote
});
module.exports = __toCommonJS(lib_exports);

// packages/node-opcua-pki/lib/ca/certificate_authority.ts
var import_node_assert7 = __toESM(require("assert"));
var import_node_fs7 = __toESM(require("fs"));
var import_node_os3 = __toESM(require("os"));
var import_node_path5 = __toESM(require("path"));
var import_chalk5 = __toESM(require("chalk"));
var import_node_opcua_crypto2 = require("node-opcua-crypto");

// packages/node-opcua-pki/lib/pki/toolbox_pfx.ts
var import_node_assert4 = __toESM(require("assert"));
var import_node_fs4 = __toESM(require("fs"));

// packages/node-opcua-pki/lib/toolbox/common.ts
var import_node_assert = __toESM(require("assert"));
function quote(str) {
  return `"${str || ""}"`;
}
function adjustDate(params) {
  (0, import_node_assert.default)(params instanceof Object);
  params.startDate = params.startDate || /* @__PURE__ */ new Date();
  (0, import_node_assert.default)(params.startDate instanceof Date);
  if (params.validityMs !== void 0) {
    if (params.validityMs <= 0) {
      throw new RangeError(`validityMs must be > 0 (got ${params.validityMs})`);
    }
    params.endDate = new Date(params.startDate.getTime() + params.validityMs);
    params.validity = Math.ceil(params.validityMs / 864e5);
  } else {
    params.validity = params.validity || 365;
    params.endDate = new Date(params.startDate.getTime());
    params.endDate.setDate(params.startDate.getDate() + params.validity);
  }
  (0, import_node_assert.default)(params.endDate instanceof Date);
  (0, import_node_assert.default)(params.startDate instanceof Date);
}
function adjustApplicationUri(params) {
  const applicationUri = params.applicationUri || "";
  if (applicationUri.length > 200) {
    throw new Error(`Openssl doesn't support urn with length greater than 200${applicationUri}`);
  }
}

// packages/node-opcua-pki/lib/toolbox/common2.ts
var import_node_assert2 = __toESM(require("assert"));
var import_node_fs = __toESM(require("fs"));
var import_node_path = __toESM(require("path"));
var import_chalk = __toESM(require("chalk"));

// packages/node-opcua-pki/lib/toolbox/config.ts
var g_config = {
  opensslVersion: "unset",
  silent: process.env.VERBOSE ? !process.env.VERBOSE : true,
  force: false
};

// packages/node-opcua-pki/lib/toolbox/debug.ts
var doDebug = process.env.NODEOPCUAPKIDEBUG || false;
var displayError = true;
var displayDebug = !!process.env.NODEOPCUAPKIDEBUG || false;
function debugLog(...args) {
  if (displayDebug) {
    console.log.apply(null, args);
  }
}
function warningLog(...args) {
  console.log.apply(null, args);
}

// packages/node-opcua-pki/lib/toolbox/common2.ts
function certificateFileExist(certificateFile) {
  if (import_node_fs.default.existsSync(certificateFile) && !g_config.force) {
    warningLog(
      import_chalk.default.yellow("        certificate ") + import_chalk.default.cyan(certificateFile) + import_chalk.default.yellow(" already exists => do not overwrite")
    );
    return false;
  }
  return true;
}
function mkdirRecursiveSync(folder) {
  if (!import_node_fs.default.existsSync(folder)) {
    debugLog(import_chalk.default.white(" .. constructing "), folder);
    import_node_fs.default.mkdirSync(folder, { recursive: true });
  }
}
function makePath(folderName, filename) {
  let s;
  if (filename) {
    s = import_node_path.default.join(import_node_path.default.normalize(folderName), filename);
  } else {
    (0, import_node_assert2.default)(folderName);
    s = folderName;
  }
  s = s.replace(/\\/g, "/");
  return s;
}

// packages/node-opcua-pki/lib/toolbox/with_openssl/execute_openssl.ts
var import_node_assert3 = __toESM(require("assert"));
var import_node_child_process2 = __toESM(require("child_process"));
var import_node_fs3 = __toESM(require("fs"));
var import_node_os2 = __toESM(require("os"));
var import_byline2 = __toESM(require("byline"));
var import_chalk3 = __toESM(require("chalk"));

// packages/node-opcua-pki/lib/toolbox/with_openssl/_env.ts
var exportedEnvVars = {};
function setEnv(varName, value) {
  if (!g_config.silent) {
    warningLog(`          set ${varName}=${value}`);
  }
  exportedEnvVars[varName] = value;
  if (["OPENSSL_CONF"].indexOf(varName) >= 0) {
    process.env[varName] = value;
  }
  if (["RANDFILE"].indexOf(varName) >= 0) {
    process.env[varName] = value;
  }
}
function hasEnv(varName) {
  return Object.prototype.hasOwnProperty.call(exportedEnvVars, varName);
}
function getEnv(varName) {
  return exportedEnvVars[varName];
}
function unsetEnv(varName) {
  delete exportedEnvVars[varName];
}
function getEnvironmentVarNames() {
  return Object.keys(exportedEnvVars).map((varName) => {
    return { key: varName, pattern: `\\$ENV\\:\\:${varName}` };
  });
}
function processAltNames(params) {
  params.dns = params.dns || [];
  params.ip = params.ip || [];
  let subjectAltName = [];
  subjectAltName.push(`URI:${params.applicationUri}`);
  subjectAltName = [].concat(
    subjectAltName,
    params.dns.map((d) => `DNS:${d}`)
  );
  subjectAltName = [].concat(
    subjectAltName,
    params.ip.map((d) => `IP:${d}`)
  );
  const subjectAltNameString = subjectAltName.join(", ");
  setEnv("ALTNAME", subjectAltNameString);
}

// packages/node-opcua-pki/lib/toolbox/with_openssl/install_prerequisite.ts
var import_node_child_process = __toESM(require("child_process"));
var import_node_fs2 = __toESM(require("fs"));
var import_node_os = __toESM(require("os"));
var import_node_path2 = __toESM(require("path"));
var import_promises = require("stream/promises");
var import_byline = __toESM(require("byline"));
var import_chalk2 = __toESM(require("chalk"));
var import_yauzl = __toESM(require("yauzl"));
var doDebug2 = process.env.NODEOPCUAPKIDEBUG || false;
async function execute(cmd, cwd) {
  let output = "";
  const options = {
    cwd,
    windowsHide: true
  };
  return await new Promise((resolve, reject) => {
    const child = import_node_child_process.default.exec(
      cmd,
      options,
      (err) => {
        const exitCode = err === null ? 0 : err.code || 1;
        if (err) reject(err);
        else {
          resolve({ exitCode, output });
        }
      }
    );
    const stream1 = (0, import_byline.default)(child.stdout);
    stream1.on("data", (line) => {
      output += `${line}
`;
      if (doDebug2) {
        process.stdout.write(`        stdout ${import_chalk2.default.yellow(line)}
`);
      }
    });
  });
}
function quote2(str) {
  return `"${str.replace(/\\/g, "/")}"`;
}
function is_expected_openssl_version(strVersion) {
  return !!strVersion.match(/OpenSSL \d/);
}
async function getopensslExecPath() {
  let result1;
  try {
    result1 = await execute("which openssl");
  } catch (err) {
    warningLog("warning: ", err.message);
    throw new Error("Cannot find openssl");
  }
  const exitCode = result1?.exitCode;
  const output = result1?.output;
  if (exitCode !== 0) {
    warningLog(import_chalk2.default.yellow(" it seems that ") + import_chalk2.default.cyan("openssl") + import_chalk2.default.yellow(" is not installed on your computer "));
    warningLog(import_chalk2.default.yellow("Please install it before running this programs"));
    throw new Error("Cannot find openssl");
  }
  const opensslExecPath = output.replace(/\n\r/g, "").trim();
  return opensslExecPath;
}
async function check_system_openssl_version() {
  const opensslExecPath = await getopensslExecPath();
  const q_opensslExecPath = quote2(opensslExecPath);
  if (doDebug2) {
    warningLog(`              OpenSSL found in : ${import_chalk2.default.yellow(opensslExecPath)}`);
  }
  const result = await execute(`${q_opensslExecPath} version`);
  const exitCode = result?.exitCode;
  const output = result?.output;
  const version = output.trim();
  const versionOK = exitCode === 0 && is_expected_openssl_version(version);
  if (!versionOK) {
    let message = import_chalk2.default.whiteBright("Warning !!!!!!!!!!!! ") + "\nyour version of openssl is " + version + ". It doesn't match the expected version";
    if (process.platform === "darwin") {
      message += import_chalk2.default.cyan("\nplease refer to :") + import_chalk2.default.yellow(" https://github.com/node-opcua/node-opcua/wiki/installing-node-opcua-or-node-red-on-MacOS");
    }
    console.log(message);
  }
  return output;
}
async function install_and_check_win32_openssl_version() {
  const downloadFolder = import_node_path2.default.join(import_node_os.default.tmpdir(), ".");
  function get_openssl_folder_win32() {
    if (process.env.LOCALAPPDATA) {
      const userProgramFolder = import_node_path2.default.join(process.env.LOCALAPPDATA, "Programs");
      if (import_node_fs2.default.existsSync(userProgramFolder)) {
        return import_node_path2.default.join(userProgramFolder, "openssl");
      }
    }
    return import_node_path2.default.join(process.cwd(), "openssl");
  }
  function get_openssl_exec_path_win32() {
    const opensslFolder2 = get_openssl_folder_win32();
    return import_node_path2.default.join(opensslFolder2, "openssl.exe");
  }
  async function check_openssl_win32() {
    const opensslExecPath2 = get_openssl_exec_path_win32();
    const exists = import_node_fs2.default.existsSync(opensslExecPath2);
    if (!exists) {
      warningLog("checking presence of ", opensslExecPath2);
      warningLog(import_chalk2.default.red(" cannot find file ") + opensslExecPath2);
      return {
        opensslOk: false,
        version: `cannot find file ${opensslExecPath2}`
      };
    } else {
      const q_openssl_exe_path = quote2(opensslExecPath2);
      const cwd = ".";
      const { exitCode, output } = await execute(`${q_openssl_exe_path} version`, cwd);
      const version = output.trim();
      if (doDebug2) {
        warningLog(" Version = ", version);
      }
      return {
        opensslOk: exitCode === 0 && is_expected_openssl_version(version),
        version
      };
    }
  }
  async function find_system_openssl_win32() {
    try {
      const result = await execute("where openssl");
      if (result.exitCode !== 0) {
        return void 0;
      }
      const opensslPath2 = result.output.split(/\r?\n/)[0].trim();
      if (!opensslPath2 || !import_node_fs2.default.existsSync(opensslPath2)) {
        return void 0;
      }
      const q4 = quote2(opensslPath2);
      const versionResult = await execute(`${q4} version`);
      const version = versionResult.output.trim();
      if (versionResult.exitCode === 0 && is_expected_openssl_version(version)) {
        warningLog(
          import_chalk2.default.green("Using system OpenSSL: ") + import_chalk2.default.cyan(version) + import_chalk2.default.green(" at ") + import_chalk2.default.cyan(opensslPath2)
        );
        return opensslPath2;
      }
      warningLog(import_chalk2.default.yellow("System OpenSSL found but version not accepted: ") + version);
      return void 0;
    } catch (_err) {
      return void 0;
    }
  }
  function win32or64() {
    if (process.env.PROCESSOR_ARCHITECTURE === "x86" && process.env.PROCESSOR_ARCHITEW6432) {
      return 64;
    }
    if (process.env.PROCESSOR_ARCHITECTURE === "AMD64") {
      return 64;
    }
    if (process.env.CURRENT_CPU === "x64") {
      return 64;
    }
    return 32;
  }
  async function download_openssl() {
    const url = win32or64() === 64 ? "https://github.com/node-opcua/node-opcua-pki/releases/download/2.14.2/openssl-1.0.2u-x64_86-win64.zip" : "https://github.com/node-opcua/node-opcua-pki/releases/download/2.14.2/openssl-1.0.2u-i386-win32.zip";
    const outputFilename = import_node_path2.default.join(downloadFolder, import_node_path2.default.basename(url));
    warningLog(`downloading ${import_chalk2.default.yellow(url)} to ${outputFilename}`);
    if (import_node_fs2.default.existsSync(outputFilename)) {
      return { downloadedFile: outputFilename };
    }
    const response = await fetch(url, { redirect: "follow" });
    if (!response.ok || !response.body) {
      throw new Error(`Failed to download OpenSSL from ${url}: ${response.status} ${response.statusText}`);
    }
    const contentLength = parseInt(response.headers.get("content-length") || "0", 10);
    let downloaded = 0;
    let lastPercent = -1;
    const fileStream = import_node_fs2.default.createWriteStream(outputFilename);
    const body = response.body;
    body.on("data", (chunk) => {
      downloaded += chunk.length;
      if (contentLength > 0) {
        const percent = Math.floor(downloaded / contentLength * 100);
        if (percent !== lastPercent && percent % 10 === 0) {
          lastPercent = percent;
          warningLog(`  download progress: ${percent}%`);
        }
      }
    });
    await (0, import_promises.pipeline)(body, fileStream);
    const stat = import_node_fs2.default.statSync(outputFilename);
    if (stat.size === 0) {
      import_node_fs2.default.unlinkSync(outputFilename);
      throw new Error(`Downloaded file is empty: ${outputFilename}`);
    }
    warningLog(import_chalk2.default.green("Download complete: ") + `${stat.size} bytes`);
    return { downloadedFile: outputFilename };
  }
  async function unzip_openssl(zipFilename) {
    const opensslFolder2 = get_openssl_folder_win32();
    const zipFile = await new Promise((resolve, reject) => {
      import_yauzl.default.open(zipFilename, { lazyEntries: true }, (err, zipfile) => {
        if (err) {
          reject(err);
        } else {
          if (!zipfile) {
            reject(new Error("zipfile is null"));
          } else {
            resolve(zipfile);
          }
        }
      });
    });
    zipFile.readEntry();
    await new Promise((resolve, reject) => {
      zipFile.on("end", (err) => {
        setImmediate(() => {
          if (doDebug2) {
            warningLog("unzip done");
          }
          if (err) {
            reject(err);
          } else {
            resolve();
          }
        });
      });
      zipFile.on("entry", (entry) => {
        zipFile.openReadStream(entry, (err, readStream) => {
          if (err) {
            return reject(err);
          }
          const file = import_node_path2.default.join(opensslFolder2, entry.fileName);
          if (doDebug2) {
            warningLog(" unzipping :", file);
          }
          const writeStream = import_node_fs2.default.createWriteStream(file, "binary");
          readStream?.pipe(writeStream);
          writeStream.on("close", () => {
            zipFile.readEntry();
          });
        });
      });
    });
  }
  const systemOpenssl = await find_system_openssl_win32();
  if (systemOpenssl) {
    return systemOpenssl;
  }
  const opensslFolder = get_openssl_folder_win32();
  const opensslExecPath = get_openssl_exec_path_win32();
  if (!import_node_fs2.default.existsSync(opensslFolder)) {
    if (doDebug2) {
      warningLog("creating openssl_folder", opensslFolder);
    }
    import_node_fs2.default.mkdirSync(opensslFolder);
  }
  const { opensslOk, version: _version } = await check_openssl_win32();
  if (!opensslOk) {
    warningLog(import_chalk2.default.yellow("openssl seems to be missing and need to be installed"));
    const { downloadedFile } = await download_openssl();
    if (doDebug2) {
      warningLog("deflating ", import_chalk2.default.yellow(downloadedFile));
    }
    await unzip_openssl(downloadedFile);
    const opensslExists = !!import_node_fs2.default.existsSync(opensslExecPath);
    if (doDebug2) {
      warningLog("verifying ", opensslExists, opensslExists ? import_chalk2.default.green("OK ") : import_chalk2.default.red(" Error"), opensslExecPath);
    }
    const _opensslExecPath2 = await check_openssl_win32();
    return opensslExecPath;
  } else {
    if (doDebug2) {
      warningLog(import_chalk2.default.green("openssl is already installed and have the expected version."));
    }
    return opensslExecPath;
  }
}
async function install_prerequisite() {
  if (process.platform !== "win32") {
    return await check_system_openssl_version();
  } else {
    return await install_and_check_win32_openssl_version();
  }
}
async function get_openssl_exec_path() {
  if (process.platform === "win32") {
    const opensslExecPath = await install_prerequisite();
    if (!import_node_fs2.default.existsSync(opensslExecPath)) {
      throw new Error(`internal error cannot find ${opensslExecPath}`);
    }
    return opensslExecPath;
  } else {
    return "openssl";
  }
}

// packages/node-opcua-pki/lib/toolbox/with_openssl/execute_openssl.ts
var opensslPath;
var n = makePath;
async function execute2(cmd, options) {
  const from = new Error();
  options.cwd = options.cwd || process.cwd();
  if (!g_config.silent) {
    warningLog(import_chalk3.default.cyan("                  CWD         "), options.cwd);
  }
  const outputs = [];
  return await new Promise((resolve, reject) => {
    const child = import_node_child_process2.default.exec(
      cmd,
      {
        cwd: options.cwd,
        windowsHide: true
      },
      (err) => {
        if (err) {
          if (!options.hideErrorMessage) {
            const fence = "###########################################";
            console.error(import_chalk3.default.bgWhiteBright.redBright(`${fence} OPENSSL ERROR ${fence}`));
            console.error(import_chalk3.default.bgWhiteBright.redBright(`CWD = ${options.cwd}`));
            console.error(import_chalk3.default.bgWhiteBright.redBright(err.message));
            console.error(import_chalk3.default.bgWhiteBright.redBright(`${fence} OPENSSL ERROR ${fence}`));
            console.error(from.stack);
          }
          reject(new Error(err.message));
          return;
        }
        resolve(outputs.join(""));
      }
    );
    if (child.stdout) {
      const stream2 = (0, import_byline2.default)(child.stdout);
      stream2.on("data", (line) => {
        outputs.push(`${line}
`);
      });
      if (!g_config.silent) {
        stream2.on("data", (line) => {
          line = line.toString();
          if (doDebug) {
            process.stdout.write(`${import_chalk3.default.white("        stdout ") + import_chalk3.default.whiteBright(line)}
`);
          }
        });
      }
    }
    if (!g_config.silent) {
      if (child.stderr) {
        const stream1 = (0, import_byline2.default)(child.stderr);
        stream1.on("data", (line) => {
          line = line.toString();
          if (displayError) {
            process.stdout.write(`${import_chalk3.default.white("        stderr ") + import_chalk3.default.red(line)}
`);
          }
        });
      }
    }
  });
}
async function find_openssl() {
  return await get_openssl_exec_path();
}
async function ensure_openssl_installed() {
  if (!opensslPath) {
    opensslPath = await find_openssl();
    const outputs = await execute_openssl("version", { cwd: "." });
    g_config.opensslVersion = outputs.trim();
    if (doDebug) {
      warningLog("OpenSSL version : ", g_config.opensslVersion);
    }
  }
}
async function execute_openssl_no_failure(cmd, options) {
  options = options || {};
  options.hideErrorMessage = true;
  try {
    return await execute_openssl(cmd, options);
  } catch (err) {
    debugLog(" (ignored error =  ERROR : )", err.message);
  }
}
function getTempFolder() {
  return import_node_os2.default.tmpdir();
}
async function execute_openssl(cmd, options) {
  debugLog("execute_openssl", cmd, options);
  const empty_config_file = n(getTempFolder(), "empty_config.cnf");
  if (!import_node_fs3.default.existsSync(empty_config_file)) {
    await import_node_fs3.default.promises.writeFile(empty_config_file, "# empty config file");
  }
  options = options || {};
  options.openssl_conf = options.openssl_conf || empty_config_file;
  (0, import_node_assert3.default)(options.openssl_conf);
  setEnv("OPENSSL_CONF", options.openssl_conf);
  if (!g_config.silent) {
    warningLog(import_chalk3.default.cyan("                  OPENSSL_CONF"), process.env.OPENSSL_CONF);
    warningLog(import_chalk3.default.cyan("                  RANDFILE    "), process.env.RANDFILE);
    warningLog(import_chalk3.default.cyan("                  CMD         openssl "), import_chalk3.default.cyanBright(cmd));
  }
  await ensure_openssl_installed();
  return await execute2(`${quote(opensslPath)} ${cmd}`, options);
}

// packages/node-opcua-pki/lib/pki/toolbox_pfx.ts
var q = quote;
var n2 = makePath;
async function createPFX(options) {
  const { certificateFile, privateKeyFile, outputFile, passphrase = "", caCertificateFiles } = options;
  (0, import_node_assert4.default)(import_node_fs4.default.existsSync(certificateFile), `Certificate file does not exist: ${certificateFile}`);
  (0, import_node_assert4.default)(import_node_fs4.default.existsSync(privateKeyFile), `Private key file does not exist: ${privateKeyFile}`);
  let cmd = `pkcs12 -export`;
  cmd += ` -in ${q(n2(certificateFile))}`;
  cmd += ` -inkey ${q(n2(privateKeyFile))}`;
  if (caCertificateFiles) {
    for (const caFile of caCertificateFiles) {
      (0, import_node_assert4.default)(import_node_fs4.default.existsSync(caFile), `CA certificate file does not exist: ${caFile}`);
      cmd += ` -certfile ${q(n2(caFile))}`;
    }
  }
  cmd += ` -out ${q(n2(outputFile))}`;
  cmd += ` -passout pass:${passphrase}`;
  await execute_openssl(cmd, {});
}
async function extractCertificateFromPFX(options) {
  const { pfxFile, passphrase = "" } = options;
  (0, import_node_assert4.default)(import_node_fs4.default.existsSync(pfxFile), `PFX file does not exist: ${pfxFile}`);
  const cmd = `pkcs12 -in ${q(n2(pfxFile))} -clcerts -nokeys -nodes -passin pass:${passphrase}`;
  return await execute_openssl(cmd, {});
}
async function extractPrivateKeyFromPFX(options) {
  const { pfxFile, passphrase = "" } = options;
  (0, import_node_assert4.default)(import_node_fs4.default.existsSync(pfxFile), `PFX file does not exist: ${pfxFile}`);
  const cmd = `pkcs12 -in ${q(n2(pfxFile))} -nocerts -nodes -passin pass:${passphrase}`;
  return await execute_openssl(cmd, {});
}
async function extractCACertificatesFromPFX(options) {
  const { pfxFile, passphrase = "" } = options;
  (0, import_node_assert4.default)(import_node_fs4.default.existsSync(pfxFile), `PFX file does not exist: ${pfxFile}`);
  const cmd = `pkcs12 -in ${q(n2(pfxFile))} -cacerts -nokeys -nodes -passin pass:${passphrase}`;
  return await execute_openssl(cmd, {});
}
async function extractAllFromPFX(options) {
  const [certificate, privateKey, caCertificates] = await Promise.all([
    extractCertificateFromPFX(options),
    extractPrivateKeyFromPFX(options),
    extractCACertificatesFromPFX(options)
  ]);
  return { certificate, privateKey, caCertificates };
}
async function convertPFXtoPEM(pfxFile, pemFile, passphrase = "") {
  (0, import_node_assert4.default)(import_node_fs4.default.existsSync(pfxFile), `PFX file does not exist: ${pfxFile}`);
  const cmd = `pkcs12 -in ${q(n2(pfxFile))} -out ${q(n2(pemFile))} -nodes -passin pass:${passphrase}`;
  await execute_openssl(cmd, {});
}
async function dumpPFX(pfxFile, passphrase = "") {
  (0, import_node_assert4.default)(import_node_fs4.default.existsSync(pfxFile), `PFX file does not exist: ${pfxFile}`);
  const cmd = `pkcs12 -in ${q(n2(pfxFile))} -info -nodes -passin pass:${passphrase}`;
  return await execute_openssl(cmd, {});
}

// packages/node-opcua-pki/lib/toolbox/display.ts
var import_chalk4 = __toESM(require("chalk"));
function displayTitle(str) {
  if (!g_config.silent) {
    warningLog("");
    warningLog(import_chalk4.default.yellowBright(str));
    warningLog(import_chalk4.default.yellow(new Array(str.length + 1).join("=")), "\n");
  }
}
function displaySubtitle(str) {
  if (!g_config.silent) {
    warningLog("");
    warningLog(`    ${import_chalk4.default.yellowBright(str)}`);
    warningLog(`    ${import_chalk4.default.white(new Array(str.length + 1).join("-"))}`, "\n");
  }
}
function display(str) {
  if (!g_config.silent) {
    warningLog(`       ${str}`);
  }
}

// packages/node-opcua-pki/lib/toolbox/with_openssl/index.ts
var import_node_constants = __toESM(require("constants"));

// packages/node-opcua-pki/lib/toolbox/with_openssl/create_certificate_signing_request.ts
var import_node_assert6 = __toESM(require("assert"));
var import_node_fs6 = __toESM(require("fs"));
var import_node_path4 = __toESM(require("path"));

// packages/node-opcua-pki/lib/misc/subject.ts
var import_node_opcua_crypto = require("node-opcua-crypto");

// packages/node-opcua-pki/lib/toolbox/with_openssl/toolbox.ts
var import_node_assert5 = __toESM(require("assert"));
var import_node_fs5 = __toESM(require("fs"));
var import_node_path3 = __toESM(require("path"));
function openssl_require2DigitYearInDate() {
  if (!g_config.opensslVersion) {
    throw new Error(
      "openssl_require2DigitYearInDate : openssl version is not known:  please call ensure_openssl_installed()"
    );
  }
  return g_config.opensslVersion.match(/OpenSSL 0\.9/);
}
g_config.opensslVersion = "";
var _counter = 0;
function stripConditionalBlocks(template) {
  return template.replace(/\{\{#([A-Z_][A-Z0-9_]*)\}\}([\s\S]*?)\{\{\/\1\}\}\r?\n?/g, (_match, key, content) => {
    const keep = hasEnv(key) && getEnv(key) !== "";
    return keep ? content : "";
  });
}
function generateStaticConfig(configPath, options) {
  const prePath = options?.cwd || "";
  const originalFilename = !import_node_path3.default.isAbsolute(configPath) ? import_node_path3.default.join(prePath, configPath) : configPath;
  let staticConfig = import_node_fs5.default.readFileSync(originalFilename, { encoding: "utf8" });
  staticConfig = stripConditionalBlocks(staticConfig);
  for (const envVar of getEnvironmentVarNames()) {
    staticConfig = staticConfig.replace(new RegExp(envVar.pattern, "gi"), getEnv(envVar.key));
  }
  const staticConfigPath = `${configPath}.${process.pid}-${_counter++}.tmp`;
  const temporaryConfigPath = !import_node_path3.default.isAbsolute(configPath) ? import_node_path3.default.join(prePath, staticConfigPath) : staticConfigPath;
  import_node_fs5.default.writeFileSync(temporaryConfigPath, staticConfig);
  if (options?.cwd) {
    return import_node_path3.default.relative(options.cwd, temporaryConfigPath);
  } else {
    return temporaryConfigPath;
  }
}
function x509Date(date) {
  date = date || /* @__PURE__ */ new Date();
  const Y = date.getUTCFullYear();
  const M = date.getUTCMonth() + 1;
  const D = date.getUTCDate();
  const h = date.getUTCHours();
  const m = date.getUTCMinutes();
  const s = date.getUTCSeconds();
  function w(s2, l) {
    return `${s2}`.padStart(l, "0");
  }
  if (openssl_require2DigitYearInDate()) {
    return `${w(Y, 2) + w(M, 2) + w(D, 2) + w(h, 2) + w(m, 2) + w(s, 2)}Z`;
  } else {
    return `${w(Y, 4) + w(M, 2) + w(D, 2) + w(h, 2) + w(m, 2) + w(s, 2)}Z`;
  }
}

// packages/node-opcua-pki/lib/toolbox/with_openssl/create_certificate_signing_request.ts
var q2 = quote;
var n3 = makePath;
async function createCertificateSigningRequestWithOpenSSL(certificateSigningRequestFilename, params) {
  (0, import_node_assert6.default)(params);
  (0, import_node_assert6.default)(params.rootDir);
  (0, import_node_assert6.default)(params.configFile);
  (0, import_node_assert6.default)(params.privateKey);
  (0, import_node_assert6.default)(typeof params.privateKey === "string");
  (0, import_node_assert6.default)(import_node_fs6.default.existsSync(params.configFile), `config file must exist ${params.configFile}`);
  (0, import_node_assert6.default)(import_node_fs6.default.existsSync(params.privateKey), `Private key must exist${params.privateKey}`);
  (0, import_node_assert6.default)(import_node_fs6.default.existsSync(params.rootDir), "RootDir key must exist");
  (0, import_node_assert6.default)(typeof certificateSigningRequestFilename === "string");
  processAltNames(params);
  const configFile = generateStaticConfig(params.configFile, { cwd: params.rootDir });
  const options = { cwd: params.rootDir, openssl_conf: import_node_path4.default.relative(params.rootDir, configFile) };
  const configOption = ` -config ${q2(n3(configFile))}`;
  const subject = params.subject ? new import_node_opcua_crypto.Subject(params.subject).toString() : void 0;
  const subjectOptions = subject ? ` -subj "${subject}"` : "";
  displaySubtitle("- Creating a Certificate Signing Request with openssl");
  await execute_openssl(
    "req -new  -sha256  -batch  -text " + configOption + " -key " + q2(n3(params.privateKey)) + subjectOptions + " -out " + q2(n3(certificateSigningRequestFilename)),
    options
  );
}

// packages/node-opcua-pki/lib/pki/templates/simple_config_template.cnf.ts
var config = '##################################################################################################\n## SIMPLE OPENSSL CONFIG FILE FOR SELF-SIGNED CERTIFICATE GENERATION\n################################################################################################################\n\ndistinguished_name       = req_distinguished_name\ndefault_md               = sha1\n\ndefault_md                = sha256                      # The default digest algorithm\n\n[ v3_ca ]\nsubjectKeyIdentifier        = hash\nauthorityKeyIdentifier      = keyid:always,issuer:always\n\n# authorityKeyIdentifier    = keyid\nbasicConstraints            = CA:TRUE\nkeyUsage                    = critical, cRLSign, keyCertSign\nnsComment                   = "Self-signed Certificate for CA generated by Node-OPCUA Certificate utility"\n#nsCertType                 = sslCA, emailCA\n#subjectAltName             = email:copy\n#issuerAltName              = issuer:copy\n#obj                        = DER:02:03\n# crlDistributionPoints       = @crl_info\n# [ crl_info ]\n# URI.0                     = http://localhost:8900/crl.pem\nsubjectAltName              = $ENV::ALTNAME\n\n[ req ]\ndays                      = 390\nreq_extensions            = v3_req\nx509_extensions           = v3_ca\n\n[v3_req]\nbasicConstraints       = CA:false\nkeyUsage               = critical, nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment\nsubjectAltName         = $ENV::ALTNAME\n\n[ v3_ca_signed]\nsubjectKeyIdentifier      = hash\nauthorityKeyIdentifier    = keyid,issuer\nbasicConstraints          = critical, CA:FALSE\nkeyUsage                  = nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment\nextendedKeyUsage          = clientAuth,serverAuth \nnsComment                 = "certificate generated by Node-OPCUA Certificate utility and signed by a CA"\nsubjectAltName            = $ENV::ALTNAME\n[ v3_selfsigned]\nsubjectKeyIdentifier      = hash\nauthorityKeyIdentifier    = keyid,issuer\nbasicConstraints          = critical, CA:FALSE\nkeyUsage                  = nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment\nextendedKeyUsage          = clientAuth,serverAuth \nnsComment                 = "Self-signed certificate generated by Node-OPCUA Certificate utility"\nsubjectAltName            = $ENV::ALTNAME\n[ req_distinguished_name ]\ncountryName             = Country Name (2 letter code)\ncountryName_default     = FR\ncountryName_min         = 2\ncountryName_max         = 2\n# stateOrProvinceName     = State or Province Name (full name)\n# stateOrProvinceName_default = Ile de France\n# localityName            = Locality Name (city, district)\n# localityName_default    = Paris\norganizationName          = Organization Name (company)\norganizationName_default  = NodeOPCUA\n# organizationalUnitName  = Organizational Unit Name (department, division)\n# organizationalUnitName_default = R&D\ncommonName                = Common Name (hostname, FQDN, IP, or your name)\ncommonName_max            = 256\ncommonName_default        = NodeOPCUA\n# emailAddress            = Email Address\n# emailAddress_max        = 40\n# emailAddress_default    = node-opcua (at) node-opcua (dot) com\nsubjectAltName            = $ENV::ALTNAME';
var simple_config_template_cnf_default = config;

// packages/node-opcua-pki/lib/ca/templates/ca_config_template.cnf.ts
var config2 = `#.........DO NOT MODIFY BY HAND .........................
[ ca ]
default_ca               = CA_default
[ CA_default ]
dir                      = %%ROOT_FOLDER%%            # the main CA folder
certs                    = $dir/certs                 # where to store certificates
new_certs_dir            = $dir/certs                 #
database                 = $dir/index.txt             # the certificate database
serial                   = $dir/serial                # the serial number counter
certificate              = $dir/public/cacert.pem     # The root CA certificate
private_key              = $dir/private/cakey.pem     # the CA private key
x509_extensions          = usr_cert                   #
default_days             = 3650                       # default validity : 10 years

# default_md               = sha1

default_md                = sha256                      # The default digest algorithm

preserve                 = no
policy                   = policy_match
# randfile                 = $dir/random.rnd
# default_startdate        = YYMMDDHHMMSSZ
# default_enddate          = YYMMDDHHMMSSZ
crl_dir                  = $dir/crl
crl_extensions           = crl_ext
crl                      = $dir/revocation_list.crl # the Revocation list
crlnumber                = $dir/crlnumber           # CRL number file
default_crl_days         = 30
default_crl_hours        = 24
#msie_hack

[ policy_match ]
countryName              = optional
stateOrProvinceName      = optional
localityName             = optional
organizationName         = optional
organizationalUnitName   = optional
commonName               = optional
emailAddress             = optional

[ req ]
default_bits             = 4096                     # Size of keys
default_keyfile          = key.pem                  # name of generated keys
distinguished_name       = req_distinguished_name
attributes               = req_attributes
x509_extensions          = v3_ca
#input_password
#output_password
string_mask              = nombstr                  # permitted characters
req_extensions           = v3_req

[ req_distinguished_name ]

#0 countryName             = Country Name (2 letter code)
# countryName_default     = FR
# countryName_min         = 2
# countryName_max         = 2
# stateOrProvinceName     = State or Province Name (full name)
# stateOrProvinceName_default = Ile de France
# localityName            = Locality Name (city, district)
# localityName_default    = Paris
organizationName          = Organization Name (company)
organizationName_default  = NodeOPCUA
# organizationalUnitName  = Organizational Unit Name (department, division)
# organizationalUnitName_default = R&D
commonName                = Common Name (hostname, FQDN, IP, or your name)
commonName_max            = 256
commonName_default        = NodeOPCUA
# emailAddress            = Email Address
# emailAddress_max        = 40
# emailAddress_default    = node-opcua (at) node-opcua (dot) com

[ req_attributes ]
#challengePassword        = A challenge password
#challengePassword_min    = 4
#challengePassword_max    = 20
#unstructuredName         = An optional company name
[ usr_cert ]
basicConstraints          = critical, CA:FALSE
subjectKeyIdentifier      = hash
authorityKeyIdentifier    = keyid,issuer:always
#authorityKeyIdentifier    = keyid
subjectAltName            = $ENV::ALTNAME
# issuerAltName            = issuer:copy
nsComment                 = ''OpenSSL Generated Certificate''
#nsCertType               = client, email, objsign for ''everything including object signing''
#nsCaRevocationUrl        = http://www.domain.dom/ca-crl.pem
#nsBaseUrl                =
#nsRenewalUrl             =
#nsCaPolicyUrl            =
#nsSslServerName          =
keyUsage                  = critical, digitalSignature, nonRepudiation, keyEncipherment, dataEncipherment, keyAgreement
extendedKeyUsage          = critical,serverAuth ,clientAuth
{{#CDP_URL}}crlDistributionPoints     = URI:$ENV::CDP_URL
{{/CDP_URL}}{{#AIA_VALUE}}authorityInfoAccess       = $ENV::AIA_VALUE
{{/AIA_VALUE}}
[ v3_req ]
basicConstraints          = critical, CA:FALSE
keyUsage                  = nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment, keyAgreement
extendedKeyUsage          = critical,serverAuth ,clientAuth
subjectAltName            = $ENV::ALTNAME
nsComment                 = "CA Generated by Node-OPCUA Certificate utility using openssl"
[ v3_ca_req ]
subjectKeyIdentifier      = hash
basicConstraints          = CA:TRUE
keyUsage                  = critical, cRLSign, keyCertSign
subjectAltName            = $ENV::ALTNAME
nsComment                 = "CA CSR generated by Node-OPCUA Certificate utility using openssl"
[ v3_ca ]
subjectKeyIdentifier      = hash
authorityKeyIdentifier    = keyid:always,issuer:always
basicConstraints          = CA:TRUE
keyUsage                  = critical, cRLSign, keyCertSign
subjectAltName            = $ENV::ALTNAME
nsComment                 = "CA Certificate generated by Node-OPCUA Certificate utility using openssl"
#nsCertType                 = sslCA, emailCA
#issuerAltName              = issuer:copy
#obj                        = DER:02:03
{{#CDP_URL}}crlDistributionPoints     = URI:$ENV::CDP_URL
{{/CDP_URL}}{{#AIA_VALUE}}authorityInfoAccess       = $ENV::AIA_VALUE
{{/AIA_VALUE}}[ v3_selfsigned]
basicConstraints          = critical, CA:FALSE
keyUsage                  = nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment, keyAgreement
extendedKeyUsage          = critical,serverAuth ,clientAuth
nsComment                 = "Self-signed certificate, generated by NodeOPCUA"
subjectAltName            = $ENV::ALTNAME

[ crl_ext ]
#issuerAltName            = issuer:copy
authorityKeyIdentifier    = keyid:always,issuer:always
#authorityInfoAccess       = @issuer_info`;
var ca_config_template_cnf_default = config2;

// packages/node-opcua-pki/lib/ca/certificate_authority.ts
var defaultSubject = "/C=FR/ST=IDF/L=Paris/O=Local NODE-OPCUA Certificate Authority/CN=NodeOPCUA-CA";
var configurationFileTemplate = ca_config_template_cnf_default;
var configurationFileSimpleTemplate = simple_config_template_cnf_default;
var config3 = {
  certificateDir: "INVALID",
  forceCA: false,
  pkiDir: "INVALID"
};
var n4 = makePath;
var q3 = quote;
function octetStringToIpAddress(a) {
  return parseInt(a.substring(0, 2), 16).toString() + "." + parseInt(a.substring(2, 4), 16).toString() + "." + parseInt(a.substring(4, 6), 16).toString() + "." + parseInt(a.substring(6, 8), 16).toString();
}
(0, import_node_assert7.default)(octetStringToIpAddress("c07b9179") === "192.123.145.121");
async function construct_CertificateAuthority(certificateAuthority) {
  const subject = certificateAuthority.subject;
  const caRootDir = import_node_path5.default.resolve(certificateAuthority.rootDir);
  async function make_folders() {
    mkdirRecursiveSync(caRootDir);
    mkdirRecursiveSync(import_node_path5.default.join(caRootDir, "private"));
    mkdirRecursiveSync(import_node_path5.default.join(caRootDir, "public"));
    mkdirRecursiveSync(import_node_path5.default.join(caRootDir, "certs"));
    mkdirRecursiveSync(import_node_path5.default.join(caRootDir, "crl"));
    mkdirRecursiveSync(import_node_path5.default.join(caRootDir, "conf"));
  }
  await make_folders();
  async function construct_default_files() {
    const serial = import_node_path5.default.join(caRootDir, "serial");
    if (!import_node_fs7.default.existsSync(serial)) {
      await import_node_fs7.default.promises.writeFile(serial, "1000");
    }
    const crlNumber = import_node_path5.default.join(caRootDir, "crlnumber");
    if (!import_node_fs7.default.existsSync(crlNumber)) {
      await import_node_fs7.default.promises.writeFile(crlNumber, "1000");
    }
    const indexFile = import_node_path5.default.join(caRootDir, "index.txt");
    if (!import_node_fs7.default.existsSync(indexFile)) {
      await import_node_fs7.default.promises.writeFile(indexFile, "");
    }
  }
  await construct_default_files();
  const caKeyExists = import_node_fs7.default.existsSync(import_node_path5.default.join(caRootDir, "private/cakey.pem"));
  const caCertExists = import_node_fs7.default.existsSync(import_node_path5.default.join(caRootDir, "public/cacert.pem"));
  if (caKeyExists && caCertExists && !config3.forceCA) {
    debugLog("CA private key and certificate already exist ... skipping");
    return;
  }
  if (caKeyExists && !caCertExists) {
    debugLog("CA private key exists but cacert.pem is missing \u2014 rebuilding CA");
    import_node_fs7.default.unlinkSync(import_node_path5.default.join(caRootDir, "private/cakey.pem"));
    const staleCsr = import_node_path5.default.join(caRootDir, "private/cakey.csr");
    if (import_node_fs7.default.existsSync(staleCsr)) {
      import_node_fs7.default.unlinkSync(staleCsr);
    }
  }
  displayTitle("Create Certificate Authority (CA)");
  const indexFileAttr = import_node_path5.default.join(caRootDir, "index.txt.attr");
  if (!import_node_fs7.default.existsSync(indexFileAttr)) {
    await import_node_fs7.default.promises.writeFile(indexFileAttr, "unique_subject = no");
  }
  const caConfigFile = certificateAuthority.configFile;
  if (1) {
    let data = configurationFileTemplate;
    data = makePath(data.replace(/%%ROOT_FOLDER%%/, caRootDir));
    await import_node_fs7.default.promises.writeFile(caConfigFile, data);
  }
  const subjectOpt = ` -subj "${subject.toString()}" `;
  const caCommonName = subject.commonName || "NodeOPCUA-CA";
  setEnv("ALTNAME", `URI:urn:${caCommonName}`);
  certificateAuthority._wireRevocationEnvVars();
  const options = { cwd: caRootDir };
  const configFile = generateStaticConfig("conf/caconfig.cnf", options);
  const configOption = ` -config ${q3(n4(configFile))}`;
  const keySize = certificateAuthority.keySize;
  const privateKeyFilename = import_node_path5.default.join(caRootDir, "private/cakey.pem");
  const csrFilename = import_node_path5.default.join(caRootDir, "private/cakey.csr");
  displayTitle(`Generate the CA private Key - ${keySize}`);
  await (0, import_node_opcua_crypto2.generatePrivateKeyFile)(privateKeyFilename, keySize);
  displayTitle("Generate a certificate request for the CA key");
  await execute_openssl(
    "req -new -sha256  -text  -extensions v3_ca_req" + configOption + " -key " + q3(n4(privateKeyFilename)) + " -out " + q3(n4(csrFilename)) + " " + subjectOpt,
    options
  );
  const issuerCA = certificateAuthority._issuerCA;
  if (issuerCA) {
    displayTitle("Generate CA Certificate (signed by issuer CA)");
    const issuerCert = import_node_path5.default.resolve(issuerCA.caCertificate);
    const issuerKey = import_node_path5.default.resolve(issuerCA.rootDir, "private/cakey.pem");
    const issuerSerial = import_node_path5.default.resolve(issuerCA.rootDir, "serial");
    await execute_openssl(
      " x509 -sha256 -req -days 3650  -text  -extensions v3_ca -extfile " + q3(n4(configFile)) + " -in private/cakey.csr  -CA " + q3(n4(issuerCert)) + " -CAkey " + q3(n4(issuerKey)) + " -CAserial " + q3(n4(issuerSerial)) + " -out public/cacert.pem",
      options
    );
  } else {
    displayTitle("Generate CA Certificate (self-signed)");
    await execute_openssl(
      " x509 -sha256 -req -days 3650  -text  -extensions v3_ca -extfile " + q3(n4(configFile)) + " -in private/cakey.csr  -signkey " + q3(n4(privateKeyFilename)) + " -out public/cacert.pem",
      options
    );
  }
  displaySubtitle("generate initial CRL (Certificate Revocation List)");
  await regenerateCrl(certificateAuthority.revocationList, configOption, options);
  displayTitle("Create Certificate Authority (CA) ---> DONE");
}
async function regenerateCrl(revocationList, configOption, options) {
  displaySubtitle("regenerate CRL (Certificate Revocation List)");
  await execute_openssl(`ca -gencrl ${configOption} -out crl/revocation_list.crl`, options);
  await execute_openssl("crl  -in  crl/revocation_list.crl -out  crl/revocation_list.der  -outform der", options);
  displaySubtitle("Display (Certificate Revocation List)");
  await execute_openssl(`crl  -in ${q3(n4(revocationList))} -text  -noout`, options);
}
function parseOpenSSLDate(dateStr) {
  const raw = dateStr?.split(",")[0] ?? "";
  if (raw.length < 12) return "";
  const yy = parseInt(raw.substring(0, 2), 10);
  const year = yy >= 70 ? 1900 + yy : 2e3 + yy;
  const month = raw.substring(2, 4);
  const day = raw.substring(4, 6);
  const hour = raw.substring(6, 8);
  const min = raw.substring(8, 10);
  const sec = raw.substring(10, 12);
  return `${year}-${month}-${day}T${hour}:${min}:${sec}Z`;
}
function validateRevocationUrl(url, fieldName) {
  if (url === void 0) {
    return void 0;
  }
  if (url === "") {
    throw new Error(`${fieldName} must not be empty \u2014 pass undefined to disable the extension`);
  }
  let parsed;
  try {
    parsed = new URL(url);
  } catch {
    throw new Error(`${fieldName} is not a valid URL: ${url}`);
  }
  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    throw new Error(`${fieldName} must use http: or https: (got ${parsed.protocol} in ${url})`);
  }
  if (!parsed.pathname || parsed.pathname === "/") {
    throw new Error(`${fieldName} must include a path component (got ${url})`);
  }
  const isLoopback = parsed.hostname === "localhost" || parsed.hostname === "::1" || parsed.hostname.startsWith("127.");
  if (isLoopback) {
    console.warn(
      `[node-opcua-pki] ${fieldName} points at loopback (${url}) \u2014 certificates issued with this URL will be unreachable from any other host.`
    );
  }
  return url;
}
var CertificateAuthority = class {
  /** RSA key size used when generating the CA private key. */
  keySize;
  /** Root filesystem path of the CA directory structure. */
  location;
  /** X.500 subject of the CA certificate. */
  subject;
  /** @internal Parent CA (undefined for root CAs). */
  _issuerCA;
  /** @internal Configured CDP / AIA URLs (US-202). */
  _crlDistributionUrl;
  _ocspResponderUrl;
  _caIssuersUrl;
  constructor(options) {
    (0, import_node_assert7.default)(Object.prototype.hasOwnProperty.call(options, "location"));
    (0, import_node_assert7.default)(Object.prototype.hasOwnProperty.call(options, "keySize"));
    this.location = options.location;
    this.keySize = options.keySize || 2048;
    this.subject = new import_node_opcua_crypto2.Subject(options.subject || defaultSubject);
    this._issuerCA = options.issuerCA;
    if (options.crlDistributionUrl !== void 0) {
      this.setCrlDistributionUrl(options.crlDistributionUrl);
    }
    if (options.ocspResponderUrl !== void 0) {
      this.setOcspResponderUrl(options.ocspResponderUrl);
    }
    if (options.caIssuersUrl !== void 0) {
      this.setCaIssuersUrl(options.caIssuersUrl);
    }
  }
  /**
   * Public URL where the CRL produced by this CA is reachable, or
   * `undefined` if no CDP extension should be emitted on issued certs.
   */
  get crlDistributionUrl() {
    return this._crlDistributionUrl;
  }
  /**
   * Public URL of the OCSP responder, or `undefined` if no AIA OCSP
   * leg should be emitted on issued certs.
   */
  get ocspResponderUrl() {
    return this._ocspResponderUrl;
  }
  /**
   * Public URL where the issuer's certificate can be fetched, or
   * `undefined` if no AIA caIssuers leg should be emitted.
   */
  get caIssuersUrl() {
    return this._caIssuersUrl;
  }
  /**
   * Configure the URL embedded as `crlDistributionPoints` in every
   * subsequently-issued certificate. Pass `undefined` to disable
   * the extension entirely. Validated synchronously — throws on
   * empty string, non-http(s) protocol, missing path. Warns (does
   * not throw) when the URL points at loopback.
   *
   * @see US-202
   */
  setCrlDistributionUrl(url) {
    this._crlDistributionUrl = validateRevocationUrl(url, "crlDistributionUrl");
  }
  /**
   * Configure the OCSP responder URL embedded as the `OCSP` leg of
   * the `authorityInfoAccess` extension on every subsequently-issued
   * certificate. Pass `undefined` to disable.
   *
   * @see US-202
   */
  setOcspResponderUrl(url) {
    this._ocspResponderUrl = validateRevocationUrl(url, "ocspResponderUrl");
  }
  /**
   * Configure the caIssuers URL embedded as the `caIssuers` leg of
   * the `authorityInfoAccess` extension on every subsequently-issued
   * certificate. Pass `undefined` to disable.
   *
   * @see US-202
   */
  setCaIssuersUrl(url) {
    this._caIssuersUrl = validateRevocationUrl(url, "caIssuersUrl");
  }
  /**
   * @internal
   * Populate the OpenSSL config substitution env vars (`CDP_URL` and
   * `AIA_VALUE`) from the configured URLs, or unset them so the
   * matching `{{#KEY}}...{{/KEY}}` blocks in the templates are
   * stripped. MUST be called before every `generateStaticConfig`
   * invocation that signs a certificate.
   */
  _wireRevocationEnvVars() {
    unsetEnv("CDP_URL");
    unsetEnv("AIA_VALUE");
    if (this._crlDistributionUrl) {
      setEnv("CDP_URL", this._crlDistributionUrl);
    }
    const aiaLegs = [];
    if (this._ocspResponderUrl) {
      aiaLegs.push(`OCSP;URI:${this._ocspResponderUrl}`);
    }
    if (this._caIssuersUrl) {
      aiaLegs.push(`caIssuers;URI:${this._caIssuersUrl}`);
    }
    if (aiaLegs.length > 0) {
      setEnv("AIA_VALUE", aiaLegs.join(","));
    }
  }
  /** Absolute path to the CA root directory (alias for {@link location}). */
  get rootDir() {
    return this.location;
  }
  /** Path to the OpenSSL configuration file (`conf/caconfig.cnf`). */
  get configFile() {
    return import_node_path5.default.normalize(import_node_path5.default.join(this.rootDir, "./conf/caconfig.cnf"));
  }
  /** Path to the CA certificate in PEM format (`public/cacert.pem`). */
  get caCertificate() {
    return makePath(this.rootDir, "./public/cacert.pem");
  }
  /**
   * Path to the issuer certificate chain (`public/issuer_chain.pem`).
   *
   * This file is created by {@link installCACertificate} when the
   * provided cert file contains additional issuer certificates
   * (e.g. intermediate + root). It is appended to signed certs
   * by {@link constructCertificateChain} to produce a full chain
   * per OPC UA Part 6 §6.2.6.
   */
  get issuerCertificateChain() {
    return makePath(this.rootDir, "./public/issuer_chain.pem");
  }
  /**
   * Path to the current Certificate Revocation List in DER format.
   * (`crl/revocation_list.der`)
   */
  get revocationListDER() {
    return makePath(this.rootDir, "./crl/revocation_list.der");
  }
  /**
   * Path to the current Certificate Revocation List in PEM format.
   * (`crl/revocation_list.crl`)
   */
  get revocationList() {
    return makePath(this.rootDir, "./crl/revocation_list.crl");
  }
  /**
   * Path to the concatenated CA certificate + CRL file.
   * Used by OpenSSL for CRL-based verification.
   */
  get caCertificateWithCrl() {
    return makePath(this.rootDir, "./public/cacertificate_with_crl.pem");
  }
  // ---------------------------------------------------------------
  // Buffer-based accessors (US-059)
  // ---------------------------------------------------------------
  /**
   * Return the CA certificate as a DER-encoded buffer.
   *
   * @throws if the CA certificate file does not exist
   *   (call {@link initialize} first).
   */
  getCACertificateDER() {
    const pem = (0, import_node_opcua_crypto2.readCertificatePEM)(this.caCertificate);
    return (0, import_node_opcua_crypto2.convertPEMtoDER)(pem);
  }
  /**
   * Return the CA certificate as a PEM-encoded string.
   *
   * @throws if the CA certificate file does not exist
   *   (call {@link initialize} first).
   */
  getCACertificatePEM() {
    const raw = (0, import_node_opcua_crypto2.readCertificatePEM)(this.caCertificate);
    const beginMarker = "-----BEGIN CERTIFICATE-----";
    const idx = raw.indexOf(beginMarker);
    if (idx > 0) {
      return raw.substring(idx);
    }
    return raw;
  }
  /**
   * Return the current Certificate Revocation List as a
   * DER-encoded buffer.
   *
   * Returns an empty buffer if no CRL has been generated yet.
   */
  getCRLDER() {
    const crlPath = this.revocationListDER;
    if (!import_node_fs7.default.existsSync(crlPath)) {
      return Buffer.alloc(0);
    }
    return import_node_fs7.default.readFileSync(crlPath);
  }
  /**
   * Return the current Certificate Revocation List as a
   * PEM-encoded string.
   *
   * Returns an empty string if no CRL has been generated yet.
   */
  getCRLPEM() {
    const crlPath = this.revocationList;
    if (!import_node_fs7.default.existsSync(crlPath)) {
      return "";
    }
    const raw = import_node_fs7.default.readFileSync(crlPath, "utf-8");
    const beginMarker = "-----BEGIN X509 CRL-----";
    const idx = raw.indexOf(beginMarker);
    if (idx > 0) {
      return raw.substring(idx);
    }
    return raw;
  }
  // ---------------------------------------------------------------
  // Certificate database API (US-057)
  // ---------------------------------------------------------------
  /**
   * Return a list of all issued certificates recorded in the
   * OpenSSL `index.txt` database.
   *
   * Each entry includes the serial number, subject, status,
   * expiry date, and (for revoked certs) the revocation date.
   */
  getIssuedCertificates() {
    return this._parseIndexTxt();
  }
  /**
   * Return the total number of certificates recorded in
   * `index.txt`.
   */
  getIssuedCertificateCount() {
    return this._parseIndexTxt().length;
  }
  /**
   * Return the status of a certificate by its serial number.
   *
   * @param serial - hex-encoded serial number (e.g. `"1000"`)
   * @returns `"valid"`, `"revoked"`, `"expired"`, or
   *   `undefined` if not found
   */
  getCertificateStatus(serial) {
    const upper = serial.toUpperCase();
    const record = this._parseIndexTxt().find((r) => r.serial.toUpperCase() === upper);
    return record?.status;
  }
  /**
   * Read a specific issued certificate by serial number and
   * return its content as a DER-encoded buffer.
   *
   * OpenSSL stores signed certificates in the `certs/`
   * directory using the naming convention `<SERIAL>.pem`.
   *
   * @param serial - hex-encoded serial number (e.g. `"1000"`)
   * @returns the DER buffer, or `undefined` if not found
   */
  getCertificateBySerial(serial) {
    const upper = serial.toUpperCase();
    const certFile = import_node_path5.default.join(this.rootDir, "certs", `${upper}.pem`);
    if (!import_node_fs7.default.existsSync(certFile)) {
      return void 0;
    }
    const pem = (0, import_node_opcua_crypto2.readCertificatePEM)(certFile);
    return (0, import_node_opcua_crypto2.convertPEMtoDER)(pem);
  }
  /**
   * Path to the OpenSSL certificate database file.
   */
  get indexFile() {
    return import_node_path5.default.join(this.rootDir, "index.txt");
  }
  /**
   * Parse the OpenSSL `index.txt` certificate database.
   *
   * Each line has tab-separated fields:
   * ```
   * status  expiry  [revocationDate]  serial  unknown  subject
   * ```
   *
   * - status: `V` (valid), `R` (revoked), `E` (expired)
   * - expiry: `YYMMDDHHmmssZ`
   * - revocationDate: present only for revoked certs
   * - serial: hex string
   * - unknown: always `"unknown"`
   * - subject: X.500 slash-delimited string
   */
  _parseIndexTxt() {
    const indexPath = this.indexFile;
    if (!import_node_fs7.default.existsSync(indexPath)) {
      return [];
    }
    const content = import_node_fs7.default.readFileSync(indexPath, "utf-8");
    const lines = content.split("\n").filter((l) => l.trim().length > 0);
    const records = [];
    for (const line of lines) {
      const fields = line.split("	");
      if (fields.length < 4) continue;
      const statusChar = fields[0];
      const expiryStr = fields[1];
      let serial;
      let subject;
      let revocationDate;
      if (statusChar === "R") {
        revocationDate = fields[2];
        serial = fields[3];
        subject = fields.length >= 6 ? fields[5] : "";
      } else {
        serial = fields[3];
        subject = fields.length >= 6 ? fields[5] : "";
      }
      let status;
      switch (statusChar) {
        case "V":
          status = "valid";
          break;
        case "R":
          status = "revoked";
          break;
        case "E":
          status = "expired";
          break;
        default:
          continue;
      }
      records.push({
        serial,
        status,
        subject,
        expiryDate: parseOpenSSLDate(expiryStr),
        revocationDate: revocationDate ? parseOpenSSLDate(revocationDate) : void 0
      });
    }
    return records;
  }
  // ---------------------------------------------------------------
  // Buffer-based CA operations (US-058)
  // ---------------------------------------------------------------
  /**
   * Sign a DER-encoded Certificate Signing Request and return
   * the signed certificate as a DER buffer.
   *
   * This method handles temp-file creation and cleanup
   * internally so that callers can work with in-memory
   * buffers only.
   *
   * The CA can override fields from the CSR by passing
   * `options.dns`, `options.ip`, `options.applicationUri`,
   * `options.startDate`, or `options.subject`.
   *
   * @param csrDer - the CSR as a DER-encoded buffer
   * @param options - signing options and CA overrides
   * @returns the signed certificate as a DER-encoded buffer
   */
  async signCertificateRequestFromDER(csrDer, options) {
    const tmpDir = await import_node_fs7.default.promises.mkdtemp(import_node_path5.default.join(import_node_os3.default.tmpdir(), "pki-sign-"));
    try {
      const csrFile = import_node_path5.default.join(tmpDir, "request.csr");
      const certFile = import_node_path5.default.join(tmpDir, "certificate.pem");
      const csrPem = (0, import_node_opcua_crypto2.toPem)(csrDer, "CERTIFICATE REQUEST");
      await import_node_fs7.default.promises.writeFile(csrFile, csrPem, "utf-8");
      const signingParams = {};
      if (options?.validityMs !== void 0) signingParams.validityMs = options.validityMs;
      else signingParams.validity = options?.validity ?? 365;
      if (options?.startDate) signingParams.startDate = options.startDate;
      if (options?.dns) signingParams.dns = options.dns;
      if (options?.ip) signingParams.ip = options.ip;
      if (options?.applicationUri) signingParams.applicationUri = options.applicationUri;
      if (options?.subject) signingParams.subject = options.subject;
      await this.signCertificateRequest(certFile, csrFile, signingParams);
      const certPem = (0, import_node_opcua_crypto2.readCertificatePEM)(certFile);
      return (0, import_node_opcua_crypto2.convertPEMtoDER)(certPem);
    } finally {
      await import_node_fs7.default.promises.rm(tmpDir, {
        recursive: true,
        force: true
      });
    }
  }
  /**
   * Advertise the validity limits this CA can honor.
   *
   * Consumers (notably the GDS server in [`cert_auth.ts`](https://github.com/sterfive/node-opcua-gds))
   * clamp a requested validity against these bounds before calling
   * {@link signCertificateRequestFromDER}, so a misconfigured
   * `defaultCertValidity` cannot ask the CA for something it cannot
   * produce.
   *
   * Defaults match the OpenSSL-backed implementation:
   * - `minValidityMs = 60_000` (1 minute) — practical floor; the
   *   X.509 spec floor is 1 second but very short certs are rarely
   *   useful and pathological for any real deployment.
   * - `maxValidityMs = 10 * 365 * 86_400_000` (≈ 10 years) — long
   *   enough for root CAs.
   * - `validityGranularityMs = 1_000` (1 second) — RFC 5280 §4.1.2.5
   *   floor on `notBefore` / `notAfter`.
   * - `nativeUnit = "second"` — what `x509Date()` actually encodes.
   *
   * @see US-208 — the consumer-side capability story.
   */
  getCapabilities() {
    return {
      minValidityMs: 6e4,
      maxValidityMs: 10 * 365 * 864e5,
      validityGranularityMs: 1e3,
      nativeUnit: "second"
    };
  }
  /**
   * Generate a new RSA key pair, create an internal CSR, sign it
   * with this CA, and return both the certificate and private key
   * as DER-encoded buffers.
   *
   * The private key is **never stored** by the CA — it exists only
   * in a temporary directory that is cleaned up after the operation.
   *
   * This is used by `StartNewKeyPairRequest` (OPC UA Part 12) for
   * constrained devices that cannot generate their own keys.
   *
   * @param options - key generation and certificate parameters
   * @returns `{ certificateDer, privateKey }` — certificate as DER,
   *   private key as a branded `PrivateKey` buffer
   */
  async generateKeyPairAndSignDER(options) {
    const keySize = options.keySize ?? 2048;
    const startDate = options.startDate ?? /* @__PURE__ */ new Date();
    const tmpDir = await import_node_fs7.default.promises.mkdtemp(import_node_path5.default.join(import_node_os3.default.tmpdir(), "pki-keygen-"));
    try {
      const privateKeyFile = import_node_path5.default.join(tmpDir, "private_key.pem");
      await (0, import_node_opcua_crypto2.generatePrivateKeyFile)(privateKeyFile, keySize);
      const configFile = import_node_path5.default.join(tmpDir, "openssl.cnf");
      await import_node_fs7.default.promises.writeFile(configFile, configurationFileSimpleTemplate, "utf-8");
      const csrFile = import_node_path5.default.join(tmpDir, "request.csr");
      await createCertificateSigningRequestWithOpenSSL(csrFile, {
        rootDir: tmpDir,
        configFile,
        privateKey: privateKeyFile,
        applicationUri: options.applicationUri,
        subject: options.subject,
        dns: options.dns ?? [],
        ip: options.ip ?? [],
        purpose: import_node_opcua_crypto2.CertificatePurpose.ForApplication
      });
      const certFile = import_node_path5.default.join(tmpDir, "certificate.pem");
      const signingParams = {
        applicationUri: options.applicationUri,
        dns: options.dns,
        ip: options.ip,
        startDate
      };
      if (options.validityMs !== void 0) signingParams.validityMs = options.validityMs;
      else signingParams.validity = options.validity ?? 365;
      await this.signCertificateRequest(certFile, csrFile, signingParams);
      const certPem = (0, import_node_opcua_crypto2.readCertificatePEM)(certFile);
      const certificateDer = (0, import_node_opcua_crypto2.convertPEMtoDER)(certPem);
      const privateKey = (0, import_node_opcua_crypto2.readPrivateKey)(privateKeyFile);
      return { certificateDer, privateKey };
    } finally {
      await import_node_fs7.default.promises.rm(tmpDir, {
        recursive: true,
        force: true
      });
    }
  }
  /**
   * Generate a new RSA key pair, create an internal CSR, sign it
   * with this CA, and return the result as a PKCS#12 (PFX)
   * buffer bundling the certificate, private key, and CA chain.
   *
   * The private key is **never stored** by the CA — it exists only
   * in a temporary directory that is cleaned up after the operation.
   *
   * @param options - key generation, certificate, and PFX options
   * @returns the PFX as a `Buffer`
   */
  async generateKeyPairAndSignPFX(options) {
    const keySize = options.keySize ?? 2048;
    const startDate = options.startDate ?? /* @__PURE__ */ new Date();
    const passphrase = options.passphrase ?? "";
    const tmpDir = await import_node_fs7.default.promises.mkdtemp(import_node_path5.default.join(import_node_os3.default.tmpdir(), "pki-keygen-pfx-"));
    try {
      const privateKeyFile = import_node_path5.default.join(tmpDir, "private_key.pem");
      await (0, import_node_opcua_crypto2.generatePrivateKeyFile)(privateKeyFile, keySize);
      const configFile = import_node_path5.default.join(tmpDir, "openssl.cnf");
      await import_node_fs7.default.promises.writeFile(configFile, configurationFileSimpleTemplate, "utf-8");
      const csrFile = import_node_path5.default.join(tmpDir, "request.csr");
      await createCertificateSigningRequestWithOpenSSL(csrFile, {
        rootDir: tmpDir,
        configFile,
        privateKey: privateKeyFile,
        applicationUri: options.applicationUri,
        subject: options.subject,
        dns: options.dns ?? [],
        ip: options.ip ?? [],
        purpose: import_node_opcua_crypto2.CertificatePurpose.ForApplication
      });
      const certFile = import_node_path5.default.join(tmpDir, "certificate.pem");
      const signingParams = {
        applicationUri: options.applicationUri,
        dns: options.dns,
        ip: options.ip,
        startDate
      };
      if (options.validityMs !== void 0) signingParams.validityMs = options.validityMs;
      else signingParams.validity = options.validity ?? 365;
      await this.signCertificateRequest(certFile, csrFile, signingParams);
      const pfxFile = import_node_path5.default.join(tmpDir, "bundle.pfx");
      await createPFX({
        certificateFile: certFile,
        privateKeyFile,
        outputFile: pfxFile,
        passphrase,
        caCertificateFiles: [this.caCertificate]
      });
      return await import_node_fs7.default.promises.readFile(pfxFile);
    } finally {
      await import_node_fs7.default.promises.rm(tmpDir, {
        recursive: true,
        force: true
      });
    }
  }
  /**
   * Revoke a DER-encoded certificate and regenerate the CRL.
   *
   * Extracts the serial number from the certificate, then
   * uses the stored cert file at `certs/<serial>.pem` for
   * revocation — avoiding temp-file PEM format mismatches.
   *
   * @param certDer - the certificate as a DER-encoded buffer
   * @param reason - CRL reason code
   *   (default: `"keyCompromise"`)
   * @throws if the certificate's serial is not found in the CA
   */
  async revokeCertificateDER(certDer, reason) {
    const info = (0, import_node_opcua_crypto2.exploreCertificate)(certDer);
    const serial = info.tbsCertificate.serialNumber.replace(/:/g, "").toUpperCase();
    const storedCertFile = import_node_path5.default.join(this.rootDir, "certs", `${serial}.pem`);
    if (!import_node_fs7.default.existsSync(storedCertFile)) {
      throw new Error(`Cannot revoke: no stored certificate found for serial ${serial} at ${storedCertFile}`);
    }
    await this.revokeCertificate(storedCertFile, {
      reason: reason ?? "keyCompromise"
    });
  }
  /**
   * Initialize the CA directory structure, generate the CA
   * private key and self-signed certificate if they do not
   * already exist.
   */
  async initialize() {
    await construct_CertificateAuthority(this);
  }
  /**
   * Initialize the CA directory structure and generate the
   * private key + CSR **without signing**.
   *
   * Use this when the CA certificate will be signed by an
   * external (third-party) root CA.  After receiving the signed
   * certificate, call {@link installCACertificate} to complete
   * the setup.
   *
   * **Idempotent / restart-safe:**
   * - If the CA certificate exists and is valid → `{ status: "ready" }`
   * - If the CA certificate has expired → `{ status: "expired", csrPath, expiryDate }`
   *   (a new CSR is generated, preserving the existing private key)
   * - If key + CSR exist but no cert (restart before install) →
   *   `{ status: "pending", csrPath }` without regenerating
   * - Otherwise → generates key + CSR → `{ status: "created", csrPath }`
   *
   * @returns an {@link InitializeCSRResult} describing the CA state
   */
  async initializeCSR() {
    const caRootDir = import_node_path5.default.resolve(this.rootDir);
    mkdirRecursiveSync(caRootDir);
    for (const dir of ["private", "public", "certs", "crl", "conf"]) {
      mkdirRecursiveSync(import_node_path5.default.join(caRootDir, dir));
    }
    const caCertFile = this.caCertificate;
    const privateKeyFile = import_node_path5.default.join(caRootDir, "private/cakey.pem");
    const csrFile = import_node_path5.default.join(caRootDir, "private/cakey.csr");
    if (import_node_fs7.default.existsSync(caCertFile)) {
      const certDer = (0, import_node_opcua_crypto2.convertPEMtoDER)((0, import_node_opcua_crypto2.readCertificatePEM)(caCertFile));
      const certInfo = (0, import_node_opcua_crypto2.exploreCertificate)(certDer);
      const notAfter = certInfo.tbsCertificate.validity.notAfter;
      if (notAfter.getTime() < Date.now()) {
        debugLog("CA certificate has expired \u2014 generating renewal CSR");
        await this._generateCSR(caRootDir, privateKeyFile, csrFile);
        return { status: "expired", csrPath: csrFile, expiryDate: notAfter };
      }
      debugLog("CA certificate already exists and is valid \u2014 ready");
      return { status: "ready" };
    }
    if (import_node_fs7.default.existsSync(privateKeyFile) && import_node_fs7.default.existsSync(csrFile)) {
      debugLog("CA key + CSR already exist \u2014 pending external signing");
      return { status: "pending", csrPath: csrFile };
    }
    const serial = import_node_path5.default.join(caRootDir, "serial");
    if (!import_node_fs7.default.existsSync(serial)) {
      await import_node_fs7.default.promises.writeFile(serial, "1000");
    }
    const crlNumber = import_node_path5.default.join(caRootDir, "crlnumber");
    if (!import_node_fs7.default.existsSync(crlNumber)) {
      await import_node_fs7.default.promises.writeFile(crlNumber, "1000");
    }
    const indexFile = import_node_path5.default.join(caRootDir, "index.txt");
    if (!import_node_fs7.default.existsSync(indexFile)) {
      await import_node_fs7.default.promises.writeFile(indexFile, "");
    }
    const indexFileAttr = import_node_path5.default.join(caRootDir, "index.txt.attr");
    if (!import_node_fs7.default.existsSync(indexFileAttr)) {
      await import_node_fs7.default.promises.writeFile(indexFileAttr, "unique_subject = no");
    }
    const caConfigFile = this.configFile;
    let data = configurationFileTemplate;
    data = makePath(data.replace(/%%ROOT_FOLDER%%/, caRootDir));
    await import_node_fs7.default.promises.writeFile(caConfigFile, data);
    if (!import_node_fs7.default.existsSync(privateKeyFile)) {
      await (0, import_node_opcua_crypto2.generatePrivateKeyFile)(privateKeyFile, this.keySize);
    }
    await this._generateCSR(caRootDir, privateKeyFile, csrFile);
    return { status: "created", csrPath: csrFile };
  }
  /**
   * Check whether the CA certificate needs renewal and, if so,
   * generate a new CSR for re-signing by the external root CA.
   *
   * Use this while the CA is running to detect upcoming expiry
   * **before** it actually expires. The existing private key is
   * preserved so previously issued certs remain valid.
   *
   * @param thresholdDays - number of days before expiry at which
   *   to trigger renewal (default: 30)
   * @returns an {@link InitializeCSRResult} — `"expired"` if
   *   renewal is needed, `"ready"` if the cert is still valid
   */
  async renewCSR(thresholdDays = 30) {
    const caRootDir = import_node_path5.default.resolve(this.rootDir);
    const caCertFile = this.caCertificate;
    const privateKeyFile = import_node_path5.default.join(caRootDir, "private/cakey.pem");
    const csrFile = import_node_path5.default.join(caRootDir, "private/cakey.csr");
    if (!import_node_fs7.default.existsSync(caCertFile)) {
      return this.initializeCSR();
    }
    const certDer = (0, import_node_opcua_crypto2.convertPEMtoDER)((0, import_node_opcua_crypto2.readCertificatePEM)(caCertFile));
    const certInfo = (0, import_node_opcua_crypto2.exploreCertificate)(certDer);
    const notAfter = certInfo.tbsCertificate.validity.notAfter;
    const thresholdMs = thresholdDays * 24 * 60 * 60 * 1e3;
    if (notAfter.getTime() - Date.now() < thresholdMs) {
      debugLog(`CA certificate expires within ${thresholdDays} days \u2014 generating renewal CSR`);
      await this._generateCSR(caRootDir, privateKeyFile, csrFile);
      return { status: "expired", csrPath: csrFile, expiryDate: notAfter };
    }
    return { status: "ready" };
  }
  /**
   * Generate a CSR using the existing private key.
   * @internal
   */
  async _generateCSR(caRootDir, privateKeyFile, csrFile) {
    const subjectOpt = ` -subj "${this.subject.toString()}" `;
    processAltNames({});
    const options = { cwd: caRootDir };
    const configFile = generateStaticConfig("conf/caconfig.cnf", options);
    const configOption = ` -config ${q3(n4(configFile))}`;
    await execute_openssl(
      "req -new -sha256  -text  -extensions v3_ca_req" + configOption + " -key " + q3(n4(privateKeyFile)) + " -out " + q3(n4(csrFile)) + " " + subjectOpt,
      options
    );
  }
  /**
   * Install an externally-signed CA certificate and generate
   * the initial CRL.
   *
   * Call this after {@link initializeCSR} once the external
   * root CA has signed the CSR.
   *
   * **Safety checks:**
   * - Verifies that the certificate's public key matches the
   *   CA private key before installing.
   *
   * @param signedCertFile - path to the PEM-encoded signed
   *   CA certificate (issued by the external root CA)
   * @returns an {@link InstallCACertificateResult} with
   *   `status: "success"` or `status: "error"` and a `reason`
   */
  async installCACertificate(signedCertFile) {
    const caRootDir = import_node_path5.default.resolve(this.rootDir);
    const caCertFile = this.caCertificate;
    const privateKeyFile = import_node_path5.default.join(caRootDir, "private/cakey.pem");
    const fullPem = await import_node_fs7.default.promises.readFile(signedCertFile, "utf8");
    const pemBlocks = fullPem.match(/-----BEGIN CERTIFICATE-----[\s\S]*?-----END CERTIFICATE-----/g);
    if (!pemBlocks || pemBlocks.length === 0) {
      return {
        status: "error",
        reason: "no_certificate_found",
        message: "The provided file does not contain any PEM-encoded certificate."
      };
    }
    const certDer = (0, import_node_opcua_crypto2.convertPEMtoDER)(pemBlocks[0]);
    const privateKey = (0, import_node_opcua_crypto2.readPrivateKey)(privateKeyFile);
    if (!(0, import_node_opcua_crypto2.certificateMatchesPrivateKey)(certDer, privateKey)) {
      return {
        status: "error",
        reason: "certificate_key_mismatch",
        message: "The provided certificate does not match the CA private key. Ensure the certificate was signed from the CSR generated by initializeCSR()."
      };
    }
    await import_node_fs7.default.promises.writeFile(caCertFile, `${pemBlocks[0]}
`);
    const issuerChainFile = this.issuerCertificateChain;
    if (pemBlocks.length > 1) {
      const issuerPem = `${pemBlocks.slice(1).join("\n")}
`;
      await import_node_fs7.default.promises.writeFile(issuerChainFile, issuerPem);
      debugLog(`Stored ${pemBlocks.length - 1} issuer certificate(s) in issuer_chain.pem`);
    } else {
      if (import_node_fs7.default.existsSync(issuerChainFile)) {
        await import_node_fs7.default.promises.unlink(issuerChainFile);
      }
    }
    const options = { cwd: caRootDir };
    const configFile = generateStaticConfig("conf/caconfig.cnf", options);
    const configOption = ` -config ${q3(n4(configFile))}`;
    await regenerateCrl(this.revocationList, configOption, options);
    return { status: "success" };
  }
  /**
   * Sign a CSR with CA extensions (`v3_ca`), producing a
   * subordinate CA certificate.
   *
   * Unlike {@link signCertificateRequest} which signs with
   * end-entity extensions (SANs, etc.), this method signs
   * with `basicConstraints = CA:TRUE` and `keyUsage =
   * keyCertSign, cRLSign`.
   *
   * @param certFile - output path for the signed CA cert (PEM)
   * @param csrFile - path to the subordinate CA's CSR
   * @param params - signing parameters
   */
  async signCACertificateRequest(certFile, csrFile, params) {
    const caRootDir = import_node_path5.default.resolve(this.rootDir);
    const options = { cwd: caRootDir };
    this._wireRevocationEnvVars();
    const configFile = generateStaticConfig("conf/caconfig.cnf", options);
    const validity = params.validity ?? 3650;
    await execute_openssl(
      ` x509 -sha256 -req -days ${validity} -text  -extensions v3_ca -extfile ` + q3(n4(configFile)) + " -in " + q3(n4(csrFile)) + " -CA " + q3(n4(this.caCertificate)) + " -CAkey " + q3(n4(import_node_path5.default.join(caRootDir, "private/cakey.pem"))) + " -CAserial " + q3(n4(import_node_path5.default.join(caRootDir, "serial"))) + " -out " + q3(n4(certFile)),
      options
    );
    await this.constructCertificateChain(certFile);
  }
  /**
   * Rebuild the combined CA certificate + CRL file.
   *
   * This concatenates the CA certificate with the current
   * revocation list so that OpenSSL can verify certificates
   * with CRL checking enabled.
   */
  async constructCACertificateWithCRL() {
    const cacertWithCRL = this.caCertificateWithCrl;
    if (import_node_fs7.default.existsSync(this.revocationList)) {
      await import_node_fs7.default.promises.writeFile(
        cacertWithCRL,
        import_node_fs7.default.readFileSync(this.caCertificate, "utf8") + import_node_fs7.default.readFileSync(this.revocationList, "utf8")
      );
    } else {
      await import_node_fs7.default.promises.writeFile(cacertWithCRL, import_node_fs7.default.readFileSync(this.caCertificate));
    }
  }
  /**
   * Append the CA certificate to a signed certificate file,
   * creating a PEM certificate chain.
   *
   * @param certificate - path to the certificate file to extend
   */
  async constructCertificateChain(certificate) {
    (0, import_node_assert7.default)(import_node_fs7.default.existsSync(certificate));
    (0, import_node_assert7.default)(import_node_fs7.default.existsSync(this.caCertificate));
    debugLog(import_chalk5.default.yellow("        certificate file :"), import_chalk5.default.cyan(certificate));
    let chain = await import_node_fs7.default.promises.readFile(certificate, "utf8");
    chain += await import_node_fs7.default.promises.readFile(this.caCertificate, "utf8");
    if (import_node_fs7.default.existsSync(this.issuerCertificateChain)) {
      chain += await import_node_fs7.default.promises.readFile(this.issuerCertificateChain, "utf8");
    }
    await import_node_fs7.default.promises.writeFile(certificate, chain);
  }
  /**
   * Create a self-signed certificate using OpenSSL.
   *
   * @param certificateFile - output path for the signed certificate
   * @param privateKey - path to the private key file
   * @param params - certificate parameters (subject, validity, SANs)
   */
  async createSelfSignedCertificate(certificateFile, privateKey, params) {
    (0, import_node_assert7.default)(typeof privateKey === "string");
    (0, import_node_assert7.default)(import_node_fs7.default.existsSync(privateKey));
    if (!certificateFileExist(certificateFile)) {
      return;
    }
    adjustDate(params);
    adjustApplicationUri(params);
    processAltNames(params);
    const csrFile = `${certificateFile}_csr`;
    (0, import_node_assert7.default)(csrFile);
    const configFile = generateStaticConfig(this.configFile, { cwd: this.rootDir });
    const options = {
      cwd: this.rootDir,
      openssl_conf: makePath(configFile)
    };
    const configOption = "";
    const subject = params.subject ? new import_node_opcua_crypto2.Subject(params.subject).toString() : "";
    const subjectOptions = subject && subject.length > 1 ? ` -subj ${subject} ` : "";
    displaySubtitle("- the certificate signing request");
    await execute_openssl(
      "req  -new -sha256 -text " + configOption + subjectOptions + " -batch -key " + q3(n4(privateKey)) + " -out " + q3(n4(csrFile)),
      options
    );
    displaySubtitle("- creating the self-signed certificate");
    await execute_openssl(
      "ca  -selfsign  -keyfile " + q3(n4(privateKey)) + " -startdate " + x509Date(params.startDate) + " -enddate " + x509Date(params.endDate) + " -batch -out " + q3(n4(certificateFile)) + " -in " + q3(n4(csrFile)),
      options
    );
    displaySubtitle("- dump the certificate for a check");
    await execute_openssl(`x509 -in ${q3(n4(certificateFile))}  -dates -fingerprint -purpose -noout`, {});
    displaySubtitle("- verify self-signed certificate");
    await execute_openssl_no_failure(`verify -verbose -CAfile ${q3(n4(certificateFile))} ${q3(n4(certificateFile))}`, options);
    await import_node_fs7.default.promises.unlink(csrFile);
  }
  /**
   * Revoke a certificate and regenerate the CRL.
   *
   * @param certificate - path to the certificate file to revoke
   * @param params - revocation parameters
   * @param params.reason - CRL reason code
   *   (default `"keyCompromise"`)
   */
  async revokeCertificate(certificate, params) {
    const crlReasons = [
      "unspecified",
      "keyCompromise",
      "CACompromise",
      "affiliationChanged",
      "superseded",
      "cessationOfOperation",
      "certificateHold",
      "removeFromCRL"
    ];
    const configFile = generateStaticConfig("conf/caconfig.cnf", { cwd: this.rootDir });
    const options = {
      cwd: this.rootDir,
      openssl_conf: makePath(configFile)
    };
    setEnv("ALTNAME", "");
    const randomFile = import_node_path5.default.join(this.rootDir, "random.rnd");
    setEnv("RANDFILE", randomFile);
    const configOption = ` -config ${q3(n4(configFile))}`;
    const reason = params.reason || "keyCompromise";
    (0, import_node_assert7.default)(crlReasons.indexOf(reason) >= 0);
    displayTitle(`Revoking certificate  ${certificate}`);
    displaySubtitle("Revoke certificate");
    await execute_openssl_no_failure(`ca -verbose ${configOption} -revoke ${q3(certificate)} -crl_reason ${reason}`, options);
    await regenerateCrl(this.revocationList, configOption, options);
    displaySubtitle("Verify that certificate is revoked");
    await execute_openssl_no_failure(
      "verify -verbose -CRLfile " + q3(n4(this.revocationList)) + " -CAfile " + q3(n4(this.caCertificate)) + " -crl_check " + q3(n4(certificate)),
      options
    );
    displaySubtitle("Produce CRL in DER form ");
    await execute_openssl(`crl  -in ${q3(n4(this.revocationList))} -out crl/revocation_list.der  -outform der`, options);
    displaySubtitle("Produce CRL in PEM form ");
    await execute_openssl(`crl  -in ${q3(n4(this.revocationList))} -out crl/revocation_list.pem  -outform pem -text `, options);
  }
  /**
   * Sign a Certificate Signing Request (CSR) with this CA.
   *
   * The signed certificate is written to `certificate`, and the
   * CA certificate chain plus CRL are appended to form a
   * complete certificate chain.
   *
   * @param certificate - output path for the signed certificate
   * @param certificateSigningRequestFilename - path to the CSR
   * @param params1 - signing parameters (validity, dates, SANs)
   * @returns the path to the signed certificate
   */
  async signCertificateRequest(certificate, certificateSigningRequestFilename, params1) {
    await ensure_openssl_installed();
    (0, import_node_assert7.default)(import_node_fs7.default.existsSync(certificateSigningRequestFilename));
    if (!certificateFileExist(certificate)) {
      return "";
    }
    adjustDate(params1);
    adjustApplicationUri(params1);
    processAltNames(params1);
    const options = { cwd: this.rootDir };
    const csr = await (0, import_node_opcua_crypto2.readCertificateSigningRequest)(certificateSigningRequestFilename);
    const csrInfo = (0, import_node_opcua_crypto2.exploreCertificateSigningRequest)(csr);
    const applicationUri = csrInfo.extensionRequest.subjectAltName.uniformResourceIdentifier ? csrInfo.extensionRequest.subjectAltName.uniformResourceIdentifier[0] : void 0;
    if (typeof applicationUri !== "string") {
      throw new Error("Cannot find applicationUri in CSR");
    }
    const dns = csrInfo.extensionRequest.subjectAltName.dNSName || [];
    let ip = csrInfo.extensionRequest.subjectAltName.iPAddress || [];
    ip = ip.map(octetStringToIpAddress);
    const params = {
      applicationUri,
      dns,
      ip
    };
    processAltNames(params);
    this._wireRevocationEnvVars();
    const configFile = generateStaticConfig("conf/caconfig.cnf", options);
    displaySubtitle("- then we ask the authority to sign the certificate signing request");
    const configOption = ` -config ${configFile}`;
    await execute_openssl(
      "ca " + configOption + " -startdate " + x509Date(params1.startDate) + " -enddate " + x509Date(params1.endDate) + " -batch -out " + q3(n4(certificate)) + " -in " + q3(n4(certificateSigningRequestFilename)),
      options
    );
    displaySubtitle("- dump the certificate for a check");
    await execute_openssl(`x509 -in ${q3(n4(certificate))}  -dates -fingerprint -purpose -noout`, options);
    displaySubtitle("- construct CA certificate with CRL");
    await this.constructCACertificateWithCRL();
    displaySubtitle("- construct certificate chain");
    await this.constructCertificateChain(certificate);
    displaySubtitle("- verify certificate against the root CA");
    await this.verifyCertificate(certificate);
    return certificate;
  }
  /**
   * Verify a certificate against this CA.
   *
   * @param certificate - path to the certificate file to verify
   */
  async verifyCertificate(certificate) {
    const isImplemented = false;
    if (isImplemented) {
      const options = { cwd: this.rootDir };
      const configFile = generateStaticConfig("conf/caconfig.cnf", options);
      setEnv("OPENSSL_CONF", makePath(configFile));
      const _configOption = ` -config ${configFile}`;
      _configOption;
      await execute_openssl_no_failure(
        `verify -verbose  -CAfile ${q3(n4(this.caCertificateWithCrl))} ${q3(n4(certificate))}`,
        options
      );
    }
  }
};

// packages/node-opcua-pki/lib/pki/certificate_manager.ts
var import_node_events = require("events");
var import_node_fs10 = __toESM(require("fs"));
var import_node_path6 = __toESM(require("path"));
var import_global_mutex = require("@ster5/global-mutex");
var import_chalk6 = __toESM(require("chalk"));
var import_chokidar = __toESM(require("chokidar"));
var import_node_opcua_crypto5 = require("node-opcua-crypto");

// packages/node-opcua-pki/lib/toolbox/without_openssl/create_certificate_signing_request.ts
var import_node_assert8 = __toESM(require("assert"));
var import_node_fs8 = __toESM(require("fs"));
var import_node_opcua_crypto3 = require("node-opcua-crypto");
async function createCertificateSigningRequestAsync(certificateSigningRequestFilename, params) {
  (0, import_node_assert8.default)(params);
  (0, import_node_assert8.default)(params.rootDir);
  (0, import_node_assert8.default)(params.configFile);
  (0, import_node_assert8.default)(params.privateKey);
  (0, import_node_assert8.default)(typeof params.privateKey === "string");
  (0, import_node_assert8.default)(import_node_fs8.default.existsSync(params.privateKey), `Private key must exist${params.privateKey}`);
  (0, import_node_assert8.default)(import_node_fs8.default.existsSync(params.rootDir), "RootDir key must exist");
  (0, import_node_assert8.default)(typeof certificateSigningRequestFilename === "string");
  const subject = params.subject ? new import_node_opcua_crypto3.Subject(params.subject).toString() : void 0;
  displaySubtitle("- Creating a Certificate Signing Request with subtile");
  const privateKeyPem = await import_node_fs8.default.promises.readFile(params.privateKey, "utf-8");
  const privateKey = await (0, import_node_opcua_crypto3.pemToPrivateKey)(privateKeyPem);
  const { csr } = await (0, import_node_opcua_crypto3.createCertificateSigningRequest)({
    privateKey,
    dns: params.dns,
    ip: params.ip,
    subject,
    applicationUri: params.applicationUri,
    purpose: params.purpose
  });
  await import_node_fs8.default.promises.writeFile(certificateSigningRequestFilename, csr, "utf-8");
  display(`- privateKey ${params.privateKey}`);
  display(`- certificateSigningRequestFilename ${certificateSigningRequestFilename}`);
}

// packages/node-opcua-pki/lib/toolbox/without_openssl/create_self_signed_certificate.ts
var import_node_assert9 = __toESM(require("assert"));
var import_node_fs9 = __toESM(require("fs"));
var import_node_opcua_crypto4 = require("node-opcua-crypto");
async function createSelfSignedCertificateAsync(certificate, params) {
  params.purpose = params.purpose || import_node_opcua_crypto4.CertificatePurpose.ForApplication;
  (0, import_node_assert9.default)(params.purpose, "Please provide a Certificate Purpose");
  (0, import_node_assert9.default)(import_node_fs9.default.existsSync(params.configFile));
  (0, import_node_assert9.default)(import_node_fs9.default.existsSync(params.rootDir));
  (0, import_node_assert9.default)(import_node_fs9.default.existsSync(params.privateKey));
  if (!params.subject) {
    throw Error("Missing subject");
  }
  (0, import_node_assert9.default)(typeof params.applicationUri === "string");
  (0, import_node_assert9.default)(Array.isArray(params.dns));
  adjustDate(params);
  (0, import_node_assert9.default)(Object.prototype.hasOwnProperty.call(params, "validity"));
  let subject = new import_node_opcua_crypto4.Subject(params.subject);
  subject = subject.toString();
  const purpose = params.purpose;
  displayTitle("Generate a certificate request");
  const privateKeyPem = await import_node_fs9.default.promises.readFile(params.privateKey, "utf-8");
  const privateKey = await (0, import_node_opcua_crypto4.pemToPrivateKey)(privateKeyPem);
  const { cert } = await (0, import_node_opcua_crypto4.createSelfSignedCertificate)({
    privateKey,
    notBefore: params.startDate,
    notAfter: params.endDate,
    validity: params.validity,
    dns: params.dns,
    ip: params.ip,
    subject,
    applicationUri: params.applicationUri,
    purpose
  });
  await import_node_fs9.default.promises.writeFile(certificate, cert, "utf-8");
}
async function createSelfSignedCertificate(certificate, params) {
  await createSelfSignedCertificateAsync(certificate, params);
}

// packages/node-opcua-pki/lib/pki/certificate_manager.ts
var configurationFileSimpleTemplate2 = simple_config_template_cnf_default;
var fsWriteFile = import_node_fs10.default.promises.writeFile;
function getOrComputeInfo(entry) {
  if (!entry.info) {
    entry.info = (0, import_node_opcua_crypto5.exploreCertificate)(entry.certificate);
  }
  return entry.info;
}
var VerificationStatus = /* @__PURE__ */ ((VerificationStatus2) => {
  VerificationStatus2["BadCertificateInvalid"] = "BadCertificateInvalid";
  VerificationStatus2["BadSecurityChecksFailed"] = "BadSecurityChecksFailed";
  VerificationStatus2["BadCertificatePolicyCheckFailed"] = "BadCertificatePolicyCheckFailed";
  VerificationStatus2["BadCertificateTimeInvalid"] = "BadCertificateTimeInvalid";
  VerificationStatus2["BadCertificateIssuerTimeInvalid"] = "BadCertificateIssuerTimeInvalid";
  VerificationStatus2["BadCertificateHostNameInvalid"] = "BadCertificateHostNameInvalid";
  VerificationStatus2["BadCertificateUriInvalid"] = "BadCertificateUriInvalid";
  VerificationStatus2["BadCertificateUseNotAllowed"] = "BadCertificateUseNotAllowed";
  VerificationStatus2["BadCertificateIssuerUseNotAllowed"] = "BadCertificateIssuerUseNotAllowed";
  VerificationStatus2["BadCertificateUntrusted"] = "BadCertificateUntrusted";
  VerificationStatus2["BadCertificateRevocationUnknown"] = "BadCertificateRevocationUnknown";
  VerificationStatus2["BadCertificateIssuerRevocationUnknown"] = "BadCertificateIssuerRevocationUnknown";
  VerificationStatus2["BadCertificateRevoked"] = "BadCertificateRevoked";
  VerificationStatus2["BadCertificateIssuerRevoked"] = "BadCertificateIssuerRevoked";
  VerificationStatus2["BadCertificateChainIncomplete"] = "BadCertificateChainIncomplete";
  VerificationStatus2["Good"] = "Good";
  return VerificationStatus2;
})(VerificationStatus || {});
function coerceCertificateChain(certificate) {
  if (Array.isArray(certificate)) {
    if (certificate.length === 0) return [];
    return certificate.reduce((acc, cert) => {
      return acc.concat((0, import_node_opcua_crypto5.split_der)(cert));
    }, []);
  }
  return (0, import_node_opcua_crypto5.split_der)(certificate);
}
function makeFingerprint(certificate) {
  const chain = coerceCertificateChain(certificate);
  return (0, import_node_opcua_crypto5.makeSHA1Thumbprint)(chain[0]).toString("hex");
}
function short(stringToShorten) {
  return stringToShorten.substring(0, 10);
}
var forbiddenChars = /[\x00-\x1F<>:"/\\|?*]/g;
function buildIdealCertificateName(certificate) {
  const chain = coerceCertificateChain(certificate);
  const fingerprint = makeFingerprint(chain);
  try {
    const commonName = (0, import_node_opcua_crypto5.exploreCertificate)(chain[0]).tbsCertificate.subject.commonName || "";
    const sanitizedCommonName = commonName.replace(forbiddenChars, "_");
    return `${sanitizedCommonName}[${fingerprint}]`;
  } catch (_err) {
    return `invalid_certificate_[${fingerprint}]`;
  }
}
function findMatchingIssuerKey(entries, wantedIssuerKey) {
  return entries.filter((entry) => {
    const info = getOrComputeInfo(entry);
    return info.tbsCertificate.extensions && info.tbsCertificate.extensions.subjectKeyIdentifier === wantedIssuerKey;
  });
}
function isSelfSigned2(info) {
  return info.tbsCertificate.extensions?.subjectKeyIdentifier === info.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier;
}
function isSelfSigned3(certificate) {
  const info = (0, import_node_opcua_crypto5.exploreCertificate)(certificate);
  return isSelfSigned2(info);
}
function _isIssuerInfo(info) {
  const basicConstraints = info.tbsCertificate.extensions?.basicConstraints;
  if (basicConstraints?.cA) {
    return true;
  }
  const keyUsage = info.tbsCertificate.extensions?.keyUsage;
  if (keyUsage?.keyCertSign) {
    return true;
  }
  return false;
}
function isIssuer(certificate) {
  try {
    const info = (0, import_node_opcua_crypto5.exploreCertificate)(certificate);
    return _isIssuerInfo(info);
  } catch (_err) {
    return false;
  }
}
function isIntermediateIssuer(certificate) {
  try {
    const info = (0, import_node_opcua_crypto5.exploreCertificate)(certificate);
    if (!_isIssuerInfo(info)) {
      return false;
    }
    return !isSelfSigned2(info);
  } catch (_err) {
    return false;
  }
}
function isRootIssuer(certificate) {
  try {
    const info = (0, import_node_opcua_crypto5.exploreCertificate)(certificate);
    if (!_isIssuerInfo(info)) {
      return false;
    }
    return isSelfSigned2(info);
  } catch (_err) {
    return false;
  }
}
function findIssuerCertificateInChain(certificate, chain) {
  const coercedCertificate = coerceCertificateChain(certificate);
  const firstCertificate = coercedCertificate[0];
  if (!firstCertificate) {
    return null;
  }
  const certInfo = (0, import_node_opcua_crypto5.exploreCertificate)(firstCertificate);
  if (isSelfSigned2(certInfo)) {
    return firstCertificate;
  }
  const wantedIssuerKey = certInfo.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier;
  if (!wantedIssuerKey) {
    debugLog("Certificate has no extension 3");
    return null;
  }
  const coercedChain = coerceCertificateChain(chain);
  const potentialIssuers = coercedChain.filter((c) => {
    const info = (0, import_node_opcua_crypto5.exploreCertificate)(c);
    return info.tbsCertificate.extensions && info.tbsCertificate.extensions.subjectKeyIdentifier === wantedIssuerKey;
  });
  if (potentialIssuers.length === 1) {
    return potentialIssuers[0];
  }
  if (potentialIssuers.length > 1) {
    debugLog("findIssuerCertificateInChain: certificate is not self-signed but has several issuers");
    return potentialIssuers[0];
  }
  return null;
}
var CertificateManagerState = /* @__PURE__ */ ((CertificateManagerState2) => {
  CertificateManagerState2[CertificateManagerState2["Uninitialized"] = 0] = "Uninitialized";
  CertificateManagerState2[CertificateManagerState2["Initializing"] = 1] = "Initializing";
  CertificateManagerState2[CertificateManagerState2["Initialized"] = 2] = "Initialized";
  CertificateManagerState2[CertificateManagerState2["Disposing"] = 3] = "Disposing";
  CertificateManagerState2[CertificateManagerState2["Disposed"] = 4] = "Disposed";
  return CertificateManagerState2;
})(CertificateManagerState || {});
var ChainCompletionStatus = /* @__PURE__ */ ((ChainCompletionStatus2) => {
  ChainCompletionStatus2["AlreadyComplete"] = "AlreadyComplete";
  ChainCompletionStatus2["ChainCompleted"] = "ChainCompleted";
  ChainCompletionStatus2["IssuerNotFound"] = "IssuerNotFound";
  ChainCompletionStatus2["EmptyChain"] = "EmptyChain";
  ChainCompletionStatus2["MaxDepthReached"] = "MaxDepthReached";
  return ChainCompletionStatus2;
})(ChainCompletionStatus || {});
var CertificateManager = class _CertificateManager extends import_node_events.EventEmitter {
  // ── Global instance registry ─────────────────────────────────
  // Tracks all initialized CertificateManager instances so their
  // file watchers can be closed automatically on process exit,
  // even if the consumer forgets to call dispose().
  static #activeInstances = /* @__PURE__ */ new Set();
  static #cleanupInstalled = false;
  static #installProcessCleanup() {
    if (_CertificateManager.#cleanupInstalled) return;
    _CertificateManager.#cleanupInstalled = true;
    const closeDanglingWatchers = () => {
      for (const cm of _CertificateManager.#activeInstances) {
        for (const w of cm.#watchers) {
          try {
            w.close();
          } catch {
          }
        }
        cm.#watchers.splice(0);
        cm.state = 4 /* Disposed */;
      }
      _CertificateManager.#activeInstances.clear();
    };
    process.on("beforeExit", closeDanglingWatchers);
    for (const signal of ["SIGINT", "SIGTERM"]) {
      process.once(signal, () => {
        closeDanglingWatchers();
        process.exit();
      });
    }
  }
  /**
   * Dispose **all** active CertificateManager instances,
   * closing their file watchers and freeing resources.
   *
   * This is mainly useful in test tear-down to ensure the
   * Node.js process can exit cleanly.
   */
  static async disposeAll() {
    const instances = [..._CertificateManager.#activeInstances];
    await Promise.all(instances.map((cm) => _CertificateManager.prototype.dispose.call(cm)));
  }
  /**
   * Assert that all CertificateManager instances have been
   * properly disposed. Throws an Error listing the locations
   * of any leaked instances.
   *
   * Intended for use in test `afterAll()` / `afterEach()`
   * hooks to catch missing `dispose()` calls early.
   *
   * @example
   * ```ts
   * after(() => {
   *     CertificateManager.checkAllDisposed();
   * });
   * ```
   */
  static checkAllDisposed() {
    if (_CertificateManager.#activeInstances.size === 0) return;
    const locations = [..._CertificateManager.#activeInstances].map((cm) => cm.rootDir);
    throw new Error(
      `${_CertificateManager.#activeInstances.size} CertificateManager instance(s) not disposed:
  - ${locations.join("\n  - ")}`
    );
  }
  // ─────────────────────────────────────────────────────────────
  /**
   * When `true` (the default), any certificate that is not
   * already in the trusted or rejected store is automatically
   * written to the rejected folder the first time it is seen.
   */
  untrustUnknownCertificate = true;
  /** Current lifecycle state of this instance. */
  state = 0 /* Uninitialized */;
  /** @deprecated Use {@link folderPollingInterval} instead (typo fix). */
  folderPoolingInterval = 5e3;
  /** Interval in milliseconds for file-system polling (when enabled). */
  get folderPollingInterval() {
    return this.folderPoolingInterval;
  }
  set folderPollingInterval(value) {
    this.folderPoolingInterval = value;
  }
  /** RSA key size used when generating the private key. */
  keySize;
  #location;
  #watchers = [];
  #pendingUnrefs = /* @__PURE__ */ new Set();
  #readCertificatesCalled = false;
  #filenameToHash = /* @__PURE__ */ new Map();
  #initializingPromise;
  #addCertValidation;
  #disableFileWatchers;
  #thumbs = {
    rejected: /* @__PURE__ */ new Map(),
    trusted: /* @__PURE__ */ new Map(),
    issuers: {
      certs: /* @__PURE__ */ new Map()
    },
    crl: /* @__PURE__ */ new Map(),
    issuersCrl: /* @__PURE__ */ new Map()
  };
  /**
   * Create a new CertificateManager.
   *
   * The constructor creates the root directory if it does not
   * exist but does **not** initialise the PKI store — call
   * {@link initialize} before using any other method.
   *
   * @param options - configuration options
   */
  constructor(options) {
    super();
    options.keySize = options.keySize || 2048;
    if (!options.location) {
      throw new Error("CertificateManager: missing 'location' option");
    }
    this.#location = makePath(options.location, "");
    this.keySize = options.keySize;
    const v = options.addCertificateValidationOptions ?? {};
    this.#addCertValidation = {
      acceptExpiredCertificate: v.acceptExpiredCertificate ?? false,
      acceptRevokedCertificate: v.acceptRevokedCertificate ?? false,
      ignoreMissingRevocationList: v.ignoreMissingRevocationList ?? false,
      maxChainLength: v.maxChainLength ?? 5
    };
    this.#disableFileWatchers = options.disableFileWatchers ?? process.env.OPCUA_PKI_DISABLE_FILE_WATCHERS === "true";
    mkdirRecursiveSync(options.location);
    if (!import_node_fs10.default.existsSync(this.#location)) {
      throw new Error(`CertificateManager cannot access location ${this.#location}`);
    }
  }
  /** Path to the OpenSSL configuration file. */
  get configFile() {
    return import_node_path6.default.join(this.rootDir, "own/openssl.cnf");
  }
  /** Root directory of the PKI store. */
  get rootDir() {
    return this.#location;
  }
  /** Path to the private key file (`own/private/private_key.pem`). */
  get privateKey() {
    return import_node_path6.default.join(this.rootDir, "own/private/private_key.pem");
  }
  /** Path to the OpenSSL random seed file. */
  get randomFile() {
    return import_node_path6.default.join(this.rootDir, "./random.rnd");
  }
  /**
   * Move a certificate to the rejected store.
   * If the certificate was previously trusted, it will be removed from the trusted folder.
   * @param certificateOrChain - the DER-encoded certificate or certificate chain
   */
  async rejectCertificate(certificateOrChain) {
    await this.#moveCertificate(certificateOrChain, "rejected");
  }
  /**
   * Move a certificate to the trusted store.
   * If the certificate was previously rejected, it will be removed from the rejected folder.
   * @param certificateOrChain - the DER-encoded certificate or certificate chain
   */
  async trustCertificate(certificateOrChain) {
    await this.#moveCertificate(certificateOrChain, "trusted");
  }
  /**
   * Check whether the trusted certificate store is empty.
   *
   * This inspects the in-memory index, which is kept in
   * sync with the `trusted/certs/` folder by file-system
   * watchers after {@link initialize} has been called.
   */
  isTrustListEmpty() {
    return this.#thumbs.trusted.size === 0;
  }
  /**
   * Return the number of certificates currently in the
   * trusted store.
   */
  getTrustedCertificateCount() {
    return this.#thumbs.trusted.size;
  }
  /** Path to the rejected certificates folder. */
  get rejectedFolder() {
    return import_node_path6.default.join(this.rootDir, "rejected");
  }
  /** Path to the trusted certificates folder. */
  get trustedFolder() {
    return import_node_path6.default.join(this.rootDir, "trusted/certs");
  }
  /** Path to the trusted CRL folder. */
  get crlFolder() {
    return import_node_path6.default.join(this.rootDir, "trusted/crl");
  }
  /** Path to the issuer (CA) certificates folder. */
  get issuersCertFolder() {
    return import_node_path6.default.join(this.rootDir, "issuers/certs");
  }
  /** Path to the issuer CRL folder. */
  get issuersCrlFolder() {
    return import_node_path6.default.join(this.rootDir, "issuers/crl");
  }
  /** Path to the own certificate folder. */
  get ownCertFolder() {
    return import_node_path6.default.join(this.rootDir, "own/certs");
  }
  get ownPrivateFolder() {
    return import_node_path6.default.join(this.rootDir, "own/private");
  }
  /**
   * Check if a certificate is in the trusted store.
   * If the certificate is unknown and `untrustUnknownCertificate` is set,
   * it will be written to the rejected folder.
   * @param certificate - the DER-encoded certificate
   * @returns `"Good"` if trusted, `"BadCertificateUntrusted"` if rejected/unknown,
   *   or `"BadCertificateInvalid"` if the certificate cannot be parsed.
   */
  async isCertificateTrusted(certificateOrCertificateChain) {
    try {
      const chain = coerceCertificateChain(certificateOrCertificateChain);
      const leafCertificate = chain[0];
      if (chain.length < 1) {
        return "BadCertificateInvalid";
      }
      let fingerprint;
      try {
        fingerprint = makeFingerprint(chain[0]);
      } catch (_err) {
        return "BadCertificateInvalid";
      }
      if (this.#thumbs.trusted.has(fingerprint)) {
        return "Good";
      }
      if (!this.#thumbs.rejected.has(fingerprint)) {
        if (!this.untrustUnknownCertificate) {
          return "Good";
        }
        try {
          (0, import_node_opcua_crypto5.exploreCertificateInfo)(chain[0]);
        } catch (_err) {
          return "BadCertificateInvalid";
        }
        const filename = import_node_path6.default.join(this.rejectedFolder, `${buildIdealCertificateName(leafCertificate)}.pem`);
        debugLog("certificate has never been seen before and is now rejected (untrusted) ", filename);
        await fsWriteFile(filename, (0, import_node_opcua_crypto5.toPem)(chain, "CERTIFICATE"));
        this.#thumbs.rejected.set(fingerprint, { certificate: leafCertificate, filename });
      }
      return "BadCertificateUntrusted";
    } catch (_err) {
      return "BadCertificateInvalid";
    }
  }
  async #innerVerifyCertificateAsync(certificateOrChain, _isIssuer, level, options) {
    if (level >= 5) {
      return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
    }
    const chain = coerceCertificateChain(certificateOrChain);
    debugLog("NB CERTIFICATE IN CHAIN = ", chain.length);
    const info = (0, import_node_opcua_crypto5.exploreCertificate)(chain[0]);
    let hasValidIssuer = false;
    let hasTrustedIssuer = false;
    const hasIssuerKey = info.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier;
    debugLog("Certificate as an Issuer Key", hasIssuerKey);
    if (hasIssuerKey) {
      const isSelfSigned = isSelfSigned2(info);
      debugLog("Is the Certificate self-signed  ?", isSelfSigned);
      if (!isSelfSigned) {
        debugLog(
          "Is issuer found in the list of know issuers ?",
          "\n subjectKeyIdentifier   = ",
          info.tbsCertificate.extensions?.subjectKeyIdentifier,
          "\n authorityKeyIdentifier = ",
          info.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier
        );
        let issuerCertificate = await this.findIssuerCertificate(chain[0]);
        if (!issuerCertificate) {
          issuerCertificate = findIssuerCertificateInChain(chain[0], chain);
          if (!issuerCertificate) {
            debugLog(
              " the issuer has not been found in the chain itself nor in the issuer.cert list => the chain is incomplete!"
            );
            return "BadCertificateChainIncomplete" /* BadCertificateChainIncomplete */;
          }
          debugLog(" the issuer certificate has been found in the chain itself ! the chain is complete !");
        } else {
          debugLog(" the issuer certificate has been found in the issuer.cert folder !");
        }
        const issuerStatus = await this.#innerVerifyCertificateAsync(issuerCertificate, true, level + 1, options);
        if (issuerStatus === "BadCertificateRevocationUnknown" /* BadCertificateRevocationUnknown */) {
          return "BadCertificateIssuerRevocationUnknown" /* BadCertificateIssuerRevocationUnknown */;
        }
        if (issuerStatus === "BadCertificateIssuerRevocationUnknown" /* BadCertificateIssuerRevocationUnknown */) {
          return "BadCertificateIssuerRevocationUnknown" /* BadCertificateIssuerRevocationUnknown */;
        }
        if (issuerStatus === "BadCertificateTimeInvalid" /* BadCertificateTimeInvalid */) {
          if (!options?.acceptOutDatedIssuerCertificate) {
            return "BadCertificateIssuerTimeInvalid" /* BadCertificateIssuerTimeInvalid */;
          }
        }
        if (issuerStatus === "BadCertificateUntrusted" /* BadCertificateUntrusted */) {
          debugLog("warning issuerStatus = ", issuerStatus.toString(), "the issuer certificate is not trusted");
        }
        if (issuerStatus !== "Good" /* Good */ && issuerStatus !== "BadCertificateUntrusted" /* BadCertificateUntrusted */) {
          return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
        }
        const isCertificateSignatureOK = (0, import_node_opcua_crypto5.verifyCertificateSignature)(chain[0], issuerCertificate);
        if (!isCertificateSignatureOK) {
          debugLog(" the certificate was not signed by the issuer as it claim to be ! Danger");
          return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
        }
        hasValidIssuer = true;
        let revokedStatus = await this.isCertificateRevoked(chain, issuerCertificate);
        if (revokedStatus === "BadCertificateRevocationUnknown" /* BadCertificateRevocationUnknown */) {
          if (options?.ignoreMissingRevocationList) {
            revokedStatus = "Good" /* Good */;
          }
        }
        if (revokedStatus !== "Good" /* Good */) {
          debugLog("revokedStatus", revokedStatus);
          return revokedStatus;
        }
        const issuerTrustedStatus = await this.#checkRejectedOrTrusted(issuerCertificate);
        debugLog("issuerTrustedStatus", issuerTrustedStatus);
        if (issuerTrustedStatus === "unknown") {
          hasTrustedIssuer = false;
        } else if (issuerTrustedStatus === "trusted") {
          hasTrustedIssuer = true;
        } else if (issuerTrustedStatus === "rejected") {
          return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
        }
      } else {
        const isCertificateSignatureOK = (0, import_node_opcua_crypto5.verifyCertificateSignature)(chain[0], chain[0]);
        if (!isCertificateSignatureOK) {
          debugLog("Self-signed Certificate signature is not valid");
          return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
        }
        const revokedStatus = await this.isCertificateRevoked(chain);
        debugLog("revokedStatus of self signed certificate:", revokedStatus);
      }
    }
    const status = await this.#checkRejectedOrTrusted(chain[0]);
    if (status === "rejected") {
      if (!(options.acceptCertificateWithValidIssuerChain && hasValidIssuer && hasTrustedIssuer)) {
        return "BadCertificateUntrusted" /* BadCertificateUntrusted */;
      }
    }
    const _c2 = chain[1] ? (0, import_node_opcua_crypto5.exploreCertificateInfo)(chain[1]) : "non";
    debugLog("chain[1] info=", _c2);
    const certificateInfo = (0, import_node_opcua_crypto5.exploreCertificateInfo)(chain[0]);
    const now = /* @__PURE__ */ new Date();
    let isTimeInvalid = false;
    if (certificateInfo.notBefore.getTime() > now.getTime()) {
      debugLog(
        `${import_chalk6.default.red("certificate is invalid : certificate is not active yet !")} not before date =${certificateInfo.notBefore}`
      );
      if (!options.acceptPendingCertificate) {
        isTimeInvalid = true;
      }
    }
    if (certificateInfo.notAfter.getTime() <= now.getTime()) {
      debugLog(
        `${import_chalk6.default.red("certificate is invalid : certificate has expired !")} not after date =${certificateInfo.notAfter}`
      );
      if (!options.acceptOutdatedCertificate) {
        isTimeInvalid = true;
      }
    }
    if (status === "trusted") {
      return isTimeInvalid ? "BadCertificateTimeInvalid" /* BadCertificateTimeInvalid */ : "Good" /* Good */;
    }
    if (hasIssuerKey) {
      if (!hasTrustedIssuer) {
        return "BadCertificateUntrusted" /* BadCertificateUntrusted */;
      }
      if (!hasValidIssuer) {
        return "BadCertificateUntrusted" /* BadCertificateUntrusted */;
      }
      if (!options.acceptCertificateWithValidIssuerChain) {
        return "BadCertificateUntrusted" /* BadCertificateUntrusted */;
      }
      return isTimeInvalid ? "BadCertificateTimeInvalid" /* BadCertificateTimeInvalid */ : "Good" /* Good */;
    } else {
      return "BadCertificateUntrusted" /* BadCertificateUntrusted */;
    }
  }
  /**
   * Internal verification hook called by {@link verifyCertificate}.
   *
   * Subclasses can override this to inject additional validation
   * logic (e.g. application-level policy checks) while still
   * delegating to the default chain/CRL/trust verification.
   *
   * @param certificate - the DER-encoded certificate to verify
   * @param options - verification options forwarded from the
   *   public API
   * @returns the verification status code
   */
  async verifyCertificateAsync(certificate, options) {
    const chain = coerceCertificateChain(certificate);
    for (const element of chain) {
      try {
        (0, import_node_opcua_crypto5.exploreCertificateInfo)(element);
      } catch (_err) {
        return "BadCertificateInvalid" /* BadCertificateInvalid */;
      }
    }
    const status1 = await this.#innerVerifyCertificateAsync(chain, false, 0, options);
    return status1;
  }
  /**
   * Verify a certificate against the PKI trust store.
   *
   * This performs a full validation including trust status,
   * issuer chain, CRL revocation checks, and time validity.
   *
   * @param certificate - the DER-encoded certificate to verify
   * @param options - optional flags to relax validation rules
   * @returns the verification status code
   */
  async verifyCertificate(certificate, options) {
    if (!certificate) {
      return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
    }
    try {
      const status = await this.verifyCertificateAsync(certificate, options || {});
      return status;
    } catch (error) {
      warningLog(`verifyCertificate error: ${error.message}`);
      return "BadCertificateInvalid" /* BadCertificateInvalid */;
    }
  }
  /**
   * Initialize the PKI directory structure, generate the
   * private key (if missing), and start file-system watchers.
   *
   * This method is idempotent — subsequent calls are no-ops.
   * It must be called before any certificate operations.
   */
  async initialize() {
    if (this.state !== 0 /* Uninitialized */) {
      return;
    }
    this.state = 1 /* Initializing */;
    this.#initializingPromise = this.#initialize();
    await this.#initializingPromise;
    this.#initializingPromise = void 0;
    this.state = 2 /* Initialized */;
    _CertificateManager.#activeInstances.add(this);
    _CertificateManager.#installProcessCleanup();
  }
  async #initialize() {
    this.state = 1 /* Initializing */;
    const pkiDir = this.#location;
    mkdirRecursiveSync(pkiDir);
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "own"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "own/certs"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "own/private"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "rejected"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "trusted"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "trusted/certs"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "trusted/crl"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "issuers"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "issuers/certs"));
    mkdirRecursiveSync(import_node_path6.default.join(pkiDir, "issuers/crl"));
    if (!import_node_fs10.default.existsSync(this.configFile) || !import_node_fs10.default.existsSync(this.privateKey)) {
      return await this.withLock2(async () => {
        if (this.state === 3 /* Disposing */ || this.state === 4 /* Disposed */) {
          return;
        }
        if (!import_node_fs10.default.existsSync(this.configFile)) {
          import_node_fs10.default.writeFileSync(this.configFile, configurationFileSimpleTemplate2);
        }
        if (!import_node_fs10.default.existsSync(this.privateKey)) {
          debugLog("generating private key ...");
          await (0, import_node_opcua_crypto5.generatePrivateKeyFile)(this.privateKey, this.keySize);
          await this.#readCertificates();
        } else {
          await this.#readCertificates();
        }
      });
    } else {
      await this.#readCertificates();
    }
  }
  /**
   * Dispose of the CertificateManager, releasing file watchers
   * and other resources. The instance should not be used after
   * calling this method.
   */
  async dispose() {
    if (this.state === 3 /* Disposing */) {
      throw new Error("Already disposing");
    }
    if (this.state === 0 /* Uninitialized */) {
      this.state = 4 /* Disposed */;
      return;
    }
    if (this.state === 1 /* Initializing */) {
      if (this.#initializingPromise) {
        await this.#initializingPromise;
      }
    }
    try {
      this.state = 3 /* Disposing */;
      await (0, import_global_mutex.drainPendingLocks)();
      for (const unreff of this.#pendingUnrefs) {
        unreff();
      }
      this.#pendingUnrefs.clear();
      await Promise.all(this.#watchers.map((w) => w.close()));
      this.#watchers.forEach((w) => {
        w.removeAllListeners();
      });
      this.#watchers.splice(0);
    } finally {
      this.state = 4 /* Disposed */;
      _CertificateManager.#activeInstances.delete(this);
    }
  }
  /**
   * Force a full re-scan of all PKI folders, rebuilding
   * the in-memory `_thumbs` index from scratch.
   *
   * Call this after external processes have modified the
   * PKI folders (e.g. via `writeTrustList` or CLI tools)
   * to ensure the CertificateManager sees the latest
   * state without waiting for file-system events.
   */
  async reloadCertificates() {
    await Promise.all(this.#watchers.map((w) => w.close()));
    for (const w of this.#watchers) {
      w.removeAllListeners();
    }
    this.#watchers.splice(0);
    this.#thumbs.rejected.clear();
    this.#thumbs.trusted.clear();
    this.#thumbs.issuers.certs.clear();
    this.#thumbs.crl.clear();
    this.#thumbs.issuersCrl.clear();
    this.#filenameToHash.clear();
    this.#readCertificatesCalled = false;
    await this.#readCertificates();
  }
  async withLock2(action) {
    const lockFileName = import_node_path6.default.join(this.rootDir, "mutex");
    return (0, import_global_mutex.withLock)({ fileToLock: lockFileName }, async () => {
      return await action();
    });
  }
  /**
   * Create a self-signed certificate for this PKI's private key.
   *
   * The certificate is written to `params.outputFile` or
   * `own/certs/self_signed_certificate.pem` by default.
   *
   * @param params - certificate parameters (subject, SANs,
   *   validity, etc.)
   */
  async createSelfSignedCertificate(params) {
    if (typeof params.applicationUri !== "string") {
      throw new Error("createSelfSignedCertificate: expecting applicationUri to be a string");
    }
    if (!import_node_fs10.default.existsSync(this.privateKey)) {
      throw new Error(`Cannot find private key ${this.privateKey}`);
    }
    let certificateFilename = import_node_path6.default.join(this.rootDir, "own/certs/self_signed_certificate.pem");
    certificateFilename = params.outputFile || certificateFilename;
    const _params = params;
    _params.rootDir = this.rootDir;
    _params.configFile = this.configFile;
    _params.privateKey = this.privateKey;
    _params.subject = params.subject || "CN=FIXME";
    await this.withLock2(async () => {
      await createSelfSignedCertificate(certificateFilename, _params);
    });
  }
  /**
   * Create a Certificate Signing Request (CSR) using this
   * PKI's private key and configuration.
   *
   * The CSR file is written to `own/certs/` with a timestamped
   * filename.
   *
   * @param params - CSR parameters (subject, SANs)
   * @returns the filesystem path to the generated CSR file
   */
  async createCertificateRequest(params) {
    if (!params) {
      throw new Error("params is required");
    }
    const _params = params;
    if (Object.prototype.hasOwnProperty.call(_params, "rootDir")) {
      throw new Error("rootDir should not be specified ");
    }
    _params.rootDir = import_node_path6.default.resolve(this.rootDir);
    _params.configFile = import_node_path6.default.resolve(this.configFile);
    _params.privateKey = import_node_path6.default.resolve(this.privateKey);
    return await this.withLock2(async () => {
      const now = /* @__PURE__ */ new Date();
      const today = `${now.toISOString().slice(0, 10)}_${now.getTime()}`;
      const certificateSigningRequestFilename = import_node_path6.default.join(this.rootDir, "own/certs", `certificate_${today}.csr`);
      await createCertificateSigningRequestAsync(certificateSigningRequestFilename, _params);
      return certificateSigningRequestFilename;
    });
  }
  /**
   * Add a CA (issuer) certificate to the issuers store.
   * If the certificate is already present, this is a no-op.
   * @param certificate - the DER-encoded CA certificate
   * @param validate - if `true`, verify the certificate before adding
   * @param addInTrustList - if `true`, also add to the trusted store
   * @returns `VerificationStatus.Good` on success
   */
  async addIssuer(certificate, validate = false, addInTrustList = false) {
    if (validate) {
      const status = await this.verifyCertificate(certificate);
      if (status !== "Good" /* Good */ && status !== "BadCertificateUntrusted" /* BadCertificateUntrusted */) {
        return status;
      }
    }
    const pemCertificate = (0, import_node_opcua_crypto5.toPem)(certificate, "CERTIFICATE");
    const fingerprint = makeFingerprint(certificate);
    if (this.#thumbs.issuers.certs.has(fingerprint)) {
      return "Good" /* Good */;
    }
    const filename = import_node_path6.default.join(this.issuersCertFolder, `issuer_${buildIdealCertificateName(certificate)}.pem`);
    await import_node_fs10.default.promises.writeFile(filename, pemCertificate, "ascii");
    this.#thumbs.issuers.certs.set(fingerprint, { certificate, filename });
    if (addInTrustList) {
      await this.trustCertificate(certificate);
    }
    return "Good" /* Good */;
  }
  /**
   * Add multiple CA (issuer) certificates to the issuers store.
   * @param certificates - the DER-encoded CA certificates
   * @param validate - if `true`, verify each certificate before adding
   * @param addInTrustList - if `true`, also add each certificate to the trusted store
   * @returns `VerificationStatus.Good` on success
   */
  async addIssuers(certificates, validate = false, addInTrustList = false) {
    for (const certificate of certificates) {
      if (!isIssuer(certificate)) {
        warningLog(`Certificate ${makeFingerprint(certificate)} is not a issuer certificate`);
        continue;
      }
      await this.addIssuer(certificate, validate, addInTrustList);
    }
    return "Good" /* Good */;
  }
  /**
   * Add a CRL to the certificate manager.
   * @param crl - the CRL to add
   * @param target - "issuers" (default) writes to issuers/crl, "trusted" writes to trusted/crl
   */
  async addRevocationList(crl, target = "issuers") {
    return await this.withLock2(async () => {
      try {
        const index = target === "trusted" ? this.#thumbs.crl : this.#thumbs.issuersCrl;
        const folder = target === "trusted" ? this.crlFolder : this.issuersCrlFolder;
        const crlInfo = (0, import_node_opcua_crypto5.exploreCertificateRevocationList)(crl);
        const key = crlInfo.tbsCertList.issuerFingerprint;
        if (!index.has(key)) {
          index.set(key, { crls: [], serialNumbers: {} });
        }
        const pemCertificate = (0, import_node_opcua_crypto5.toPem)(crl, "X509 CRL");
        const sanitizedKey = key.replace(/:/g, "");
        const filename = import_node_path6.default.join(folder, `crl_[${sanitizedKey}].pem`);
        await import_node_fs10.default.promises.writeFile(filename, pemCertificate, "ascii");
        await this.#onCrlFileAdded(index, filename);
        await this.#waitAndCheckCRLProcessingStatus();
        return "Good" /* Good */;
      } catch (err) {
        debugLog(err);
        return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
      }
    });
  }
  /**
   * Remove all CRL files from the specified folder(s) and clear the
   * corresponding in-memory index.
   * @param target - "issuers" clears issuers/crl, "trusted" clears
   *   trusted/crl, "all" clears both.
   */
  async clearRevocationLists(target) {
    const clearFolder = async (folder, index) => {
      try {
        const files = await import_node_fs10.default.promises.readdir(folder);
        for (const file of files) {
          const ext = import_node_path6.default.extname(file).toLowerCase();
          if (ext === ".crl" || ext === ".pem" || ext === ".der") {
            await import_node_fs10.default.promises.unlink(import_node_path6.default.join(folder, file));
          }
        }
      } catch (err) {
        if (err.code !== "ENOENT") {
          throw err;
        }
      }
      index.clear();
    };
    if (target === "issuers" || target === "all") {
      await clearFolder(this.issuersCrlFolder, this.#thumbs.issuersCrl);
    }
    if (target === "trusted" || target === "all") {
      await clearFolder(this.crlFolder, this.#thumbs.crl);
    }
  }
  /**
   * Check whether an issuer certificate with the given thumbprint
   * is already registered.
   * @param thumbprint - hex-encoded SHA-1 thumbprint (lowercase)
   */
  async hasIssuer(thumbprint) {
    await this.#readCertificates();
    const normalized = thumbprint.toLowerCase();
    return this.#thumbs.issuers.certs.has(normalized);
  }
  /**
   * Remove a trusted certificate identified by its SHA-1 thumbprint.
   * Deletes the file on disk and removes the entry from the
   * in-memory index.
   * @param thumbprint - hex-encoded SHA-1 thumbprint (lowercase)
   * @returns the removed certificate buffer, or `null` if not found
   */
  async removeTrustedCertificate(thumbprint) {
    await this.#readCertificates();
    const normalized = thumbprint.toLowerCase();
    const entry = this.#thumbs.trusted.get(normalized);
    if (!entry) {
      return null;
    }
    try {
      await import_node_fs10.default.promises.unlink(entry.filename);
    } catch (err) {
      if (err.code !== "ENOENT") {
        throw err;
      }
    }
    this.#thumbs.trusted.delete(normalized);
    return entry.certificate;
  }
  /**
   * Remove an issuer certificate identified by its SHA-1 thumbprint.
   * Deletes the file on disk and removes the entry from the
   * in-memory index.
   * @param thumbprint - hex-encoded SHA-1 thumbprint (lowercase)
   * @returns the removed certificate buffer, or `null` if not found
   */
  async removeIssuer(thumbprint) {
    await this.#readCertificates();
    const normalized = thumbprint.toLowerCase();
    const entry = this.#thumbs.issuers.certs.get(normalized);
    if (!entry) {
      return null;
    }
    try {
      await import_node_fs10.default.promises.unlink(entry.filename);
    } catch (err) {
      if (err.code !== "ENOENT") {
        throw err;
      }
    }
    this.#thumbs.issuers.certs.delete(normalized);
    return entry.certificate;
  }
  /**
   * Remove all CRL files that were issued by the given CA certificate
   * from the specified folder (or both).
   * @param issuerCertificate - the CA certificate whose CRLs to remove
   * @param target - "issuers", "trusted", or "all" (default "all")
   */
  async removeRevocationListsForIssuer(issuerCertificate, target = "all") {
    const issuerInfo = (0, import_node_opcua_crypto5.exploreCertificate)(issuerCertificate);
    const issuerFingerprint = issuerInfo.tbsCertificate.subjectFingerPrint;
    const processIndex = async (index) => {
      const crlData = index.get(issuerFingerprint);
      if (!crlData) return;
      for (const crlEntry of crlData.crls) {
        try {
          await import_node_fs10.default.promises.unlink(crlEntry.filename);
        } catch (err) {
          if (err.code !== "ENOENT") {
            throw err;
          }
        }
      }
      index.delete(issuerFingerprint);
    };
    if (target === "issuers" || target === "all") {
      await processIndex(this.#thumbs.issuersCrl);
    }
    if (target === "trusted" || target === "all") {
      await processIndex(this.#thumbs.crl);
    }
  }
  /**
   * Validate a certificate (optionally with its chain) and add
   * the leaf certificate to the trusted store.
   *
   * Performs OPC UA Part 4, Table 100 validation:
   *
   * 1. **Certificate Structure** — parse the DER encoding.
   * 2. **Build Certificate Chain** — walk from the leaf to a
   *    self-signed root CA, using the provided chain and the
   *    issuers store.
   * 3. **Signature** — verify each certificate's signature
   *    against its issuer.
   * 4. **Issuer Presence** — every issuer in the chain must
   *    already be registered in the issuers store (per GDS
   *    7.8.2.6).
   * 5. **Validity Period** — each certificate must be within
   *    its validity window (overridable via
   *    {@link AddCertificateValidationOptions.acceptExpiredCertificate}).
   * 6. **Revocation Check** — each certificate is checked
   *    against its issuer's CRL (overridable via
   *    {@link AddCertificateValidationOptions.acceptRevokedCertificate}
   *    and {@link AddCertificateValidationOptions.ignoreMissingRevocationList}).
   *
   * Only the leaf certificate is added to the trusted store.
   *
   * @param certificateChain - DER-encoded certificate or chain
   * @returns `VerificationStatus.Good` on success, or an error
   *  status indicating why the certificate was rejected.
   */
  async addTrustedCertificateFromChain(certificateChain) {
    try {
      return await this.#addTrustedCertificateFromChainImpl(certificateChain);
    } catch (_err) {
      warningLog("addTrustedCertificateFromChain: unexpected error", _err);
      return "BadCertificateInvalid" /* BadCertificateInvalid */;
    }
  }
  async #addTrustedCertificateFromChainImpl(certificateChain) {
    let certificates;
    try {
      certificates = coerceCertificateChain(certificateChain);
    } catch (_err) {
      return "BadCertificateInvalid" /* BadCertificateInvalid */;
    }
    if (certificates.length === 0) {
      return "BadCertificateInvalid" /* BadCertificateInvalid */;
    }
    const leafCertificate = certificates[0];
    const opts = this.#addCertValidation;
    let leafInfo;
    try {
      leafInfo = (0, import_node_opcua_crypto5.exploreCertificate)(leafCertificate);
    } catch (_err) {
      return "BadCertificateInvalid" /* BadCertificateInvalid */;
    }
    await this.#scanCertFolder(this.issuersCertFolder, this.#thumbs.issuers.certs);
    let currentCert = leafCertificate;
    let currentInfo = leafInfo;
    let depth = 0;
    while (true) {
      depth++;
      if (depth > opts.maxChainLength) {
        return "BadSecurityChecksFailed" /* BadSecurityChecksFailed */;
      }
      if (!opts.acceptExpiredCertificate) {
        let certDetails;
        try {
          certDetails = (0, import_node_opcua_crypto5.exploreCertificateInfo)(currentCert);
        } catch (_err) {
          return "BadCertificateInvalid" /* BadCertificateInvalid */;
        }
        const now = /* @__PURE__ */ new Date();
        if (certDetails.notBefore.getTime() > now.getTime()) {
          return "BadCertificateTimeInvalid" /* BadCertificateTimeInvalid */;
        }
        if (certDetails.notAfter.getTime() <= now.getTime()) {
          return depth === 1 ? "BadCertificateTimeInvalid" /* BadCertificateTimeInvalid */ : "BadCertificateIssuerTimeInvalid" /* BadCertificateIssuerTimeInvalid */;
        }
      }
      if (isSelfSigned2(currentInfo)) {
        try {
          if (!(0, import_node_opcua_crypto5.verifyCertificateSignature)(currentCert, currentCert)) {
            return "BadCertificateInvalid" /* BadCertificateInvalid */;
          }
        } catch (_err) {
          return "BadCertificateInvalid" /* BadCertificateInvalid */;
        }
        break;
      }
      let issuerCert = await this.findIssuerCertificate(currentCert);
      if (!issuerCert) {
        issuerCert = findIssuerCertificateInChain(currentCert, certificates);
        if (!issuerCert || issuerCert === currentCert) {
          return "BadCertificateChainIncomplete" /* BadCertificateChainIncomplete */;
        }
      }
      try {
        if (!(0, import_node_opcua_crypto5.verifyCertificateSignature)(currentCert, issuerCert)) {
          return "BadCertificateInvalid" /* BadCertificateInvalid */;
        }
      } catch (_err) {
        return "BadCertificateInvalid" /* BadCertificateInvalid */;
      }
      const issuerThumbprint = makeFingerprint(issuerCert);
      if (!await this.hasIssuer(issuerThumbprint)) {
        return "BadCertificateChainIncomplete" /* BadCertificateChainIncomplete */;
      }
      const revokedStatus = await this.isCertificateRevoked(currentCert, issuerCert);
      if (revokedStatus === "BadCertificateRevoked" /* BadCertificateRevoked */) {
        if (!opts.acceptRevokedCertificate) {
          return "BadCertificateRevoked" /* BadCertificateRevoked */;
        }
      } else if (revokedStatus === "BadCertificateRevocationUnknown" /* BadCertificateRevocationUnknown */) {
        if (!opts.ignoreMissingRevocationList) {
          return "BadCertificateRevocationUnknown" /* BadCertificateRevocationUnknown */;
        }
      }
      currentCert = issuerCert;
      try {
        currentInfo = (0, import_node_opcua_crypto5.exploreCertificate)(currentCert);
      } catch (_err) {
        return "BadCertificateInvalid" /* BadCertificateInvalid */;
      }
    }
    await this.trustCertificate(certificates);
    return "Good" /* Good */;
  }
  /**
   * Check whether an issuer certificate is still needed by any
   * certificate in the trusted store.
   *
   * This is used before removing an issuer to ensure that
   * doing so would not break the chain of any trusted
   * certificate.
   *
   * @param issuerCertificate - the CA certificate to check
   * @returns `true` if at least one trusted certificate was
   *   signed by this issuer.
   */
  async isIssuerInUseByTrustedCertificate(issuerCertificate) {
    await this.#readCertificates();
    for (const entry of this.#thumbs.trusted.values()) {
      if (!entry.certificate) continue;
      try {
        if ((0, import_node_opcua_crypto5.verifyCertificateSignature)(entry.certificate, issuerCertificate)) {
          return true;
        }
      } catch (_err) {
      }
    }
    return false;
  }
  /**
   *  find the issuer certificate among the trusted  issuer certificates.
   *
   *  The findIssuerCertificate method is an asynchronous method that attempts to find
   *  the issuer certificate for a given certificate from the list of issuer certificate declared in the PKI
   *
   *  - If the certificate is self-signed, it returns the certificate itself.
   *
   *  - If the certificate has no extension 3, it is assumed to be generated by an old system, and a null value is returned.
   *
   *  - the method checks both issuer and trusted certificates and returns the appropriate issuercertificate,
   *    if found. If multiple matching certificates are found, a warning is logged to the console.
   *
   */
  async findIssuerCertificate(certificate) {
    const firstCertificate = coerceCertificateChain(certificate)[0];
    const certInfo = (0, import_node_opcua_crypto5.exploreCertificate)(firstCertificate);
    if (isSelfSigned2(certInfo)) {
      return firstCertificate;
    }
    const wantedIssuerKey = certInfo.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier;
    if (!wantedIssuerKey) {
      debugLog("Certificate has no extension 3");
      return null;
    }
    const issuerCertificates = [...this.#thumbs.issuers.certs.values()];
    const selectedIssuerCertificates = findMatchingIssuerKey(issuerCertificates, wantedIssuerKey);
    if (selectedIssuerCertificates.length > 0) {
      if (selectedIssuerCertificates.length > 1) {
        warningLog("Warning more than one issuer certificate exists with subjectKeyIdentifier ", wantedIssuerKey);
      }
      return selectedIssuerCertificates[0].certificate || null;
    }
    const trustedCertificates = [...this.#thumbs.trusted.values()];
    const selectedTrustedCertificates = findMatchingIssuerKey(trustedCertificates, wantedIssuerKey);
    if (selectedTrustedCertificates.length > 1) {
      warningLog(
        "Warning more than one certificate exists with subjectKeyIdentifier in trusted certificate list ",
        wantedIssuerKey,
        selectedTrustedCertificates.length
      );
    }
    return selectedTrustedCertificates.length > 0 ? selectedTrustedCertificates[0].certificate : null;
  }
  /**
   * Outcome status for {@link CertificateManager.completeCertificateChain}.
   */
  static ChainCompletionStatus = ChainCompletionStatus;
  /**
   * Complete a certificate chain by walking the issuer store.
   *
   * Starting from the last certificate in the provided chain, this method
   * repeatedly calls {@link findIssuerCertificate} to locate the parent
   * certificate until it reaches a self-signed root or can no longer find
   * an issuer.
   *
   * @param chain - the (potentially partial) certificate chain, leaf first
   * @param maxDepth - maximum number of issuers to append (default: 10)
   * @returns a {@link ChainCompletionResult} containing the (possibly completed)
   *          chain, a status code, and an optional diagnostic message.
   */
  async completeCertificateChain(chain, maxDepth = 10) {
    if (chain.length === 0) {
      return {
        chain,
        status: "EmptyChain" /* EmptyChain */,
        message: "Input chain is empty \u2014 nothing to complete."
      };
    }
    await this.#scanCertFolder(this.issuersCertFolder, this.#thumbs.issuers.certs);
    const result = [...chain];
    let depth = 0;
    while (depth < maxDepth) {
      const lastCert = result[result.length - 1];
      const lastInfo = (0, import_node_opcua_crypto5.exploreCertificate)(lastCert);
      if (isSelfSigned2(lastInfo)) {
        const wasExtended = result.length > chain.length;
        return {
          chain: result,
          status: wasExtended ? "ChainCompleted" /* ChainCompleted */ : "AlreadyComplete" /* AlreadyComplete */,
          message: wasExtended ? `Chain completed: ${result.length - chain.length} issuer(s) appended, ending at self-signed root "${lastInfo.tbsCertificate.subject.commonName}".` : `Chain is already complete (self-signed root "${lastInfo.tbsCertificate.subject.commonName}").`
        };
      }
      const issuerCert = await this.findIssuerCertificate(lastCert);
      if (!issuerCert) {
        const cn = lastInfo.tbsCertificate.subject.commonName ?? "?";
        const akid = lastInfo.tbsCertificate.extensions?.authorityKeyIdentifier?.keyIdentifier ?? "?";
        const msg = `Cannot find issuer for "${cn}" (authorityKeyIdentifier: ${akid}). Ensure the CA certificate is present in the issuers/certs folder.`;
        warningLog(`completeCertificateChain: ${msg}`);
        return {
          chain: result,
          status: "IssuerNotFound" /* IssuerNotFound */,
          message: msg
        };
      }
      const issuerFingerprint = makeFingerprint(issuerCert);
      const alreadyInChain = result.some((c) => makeFingerprint(c) === issuerFingerprint);
      if (alreadyInChain) {
        return {
          chain: result,
          status: "AlreadyComplete" /* AlreadyComplete */,
          message: `Chain ends at root "${(0, import_node_opcua_crypto5.exploreCertificate)(issuerCert).tbsCertificate.subject.commonName}" (already present in chain).`
        };
      }
      result.push(issuerCert);
      depth++;
    }
    return {
      chain: result,
      status: "MaxDepthReached" /* MaxDepthReached */,
      message: `Chain completion stopped after ${maxDepth} iterations \u2014 possible circular chain or very deep hierarchy.`
    };
  }
  /**
   *
   * check if the certificate explicitly appear in the trust list, the reject list or none.
   * In case of being in the reject and trusted list at the same time is consider: rejected.
   * @internal
   * @private
   */
  async #checkRejectedOrTrusted(certificate) {
    const firstCertificate = coerceCertificateChain(certificate)[0];
    const fingerprint = makeFingerprint(firstCertificate);
    debugLog("#checkRejectedOrTrusted fingerprint ", short(fingerprint));
    await this.#readCertificates();
    if (this.#thumbs.rejected.has(fingerprint)) {
      return "rejected";
    }
    if (this.#thumbs.trusted.has(fingerprint)) {
      return "trusted";
    }
    return "unknown";
  }
  async #moveCertificate(certificateOrChain, newStatus) {
    await this.withLock2(async () => {
      const chain = coerceCertificateChain(certificateOrChain);
      const certificate = chain[0];
      const fingerprint = makeFingerprint(certificate);
      let status = await this.#checkRejectedOrTrusted(certificate);
      if (status === "unknown") {
        const pem = (0, import_node_opcua_crypto5.toPem)(chain, "CERTIFICATE");
        const filename = import_node_path6.default.join(this.rejectedFolder, `${buildIdealCertificateName(certificate)}.pem`);
        await import_node_fs10.default.promises.writeFile(filename, pem);
        this.#thumbs.rejected.set(fingerprint, { certificate, filename });
        status = "rejected";
      }
      debugLog("#moveCertificate", fingerprint.substring(0, 10), "from", status, "to", newStatus);
      if (status !== "rejected" && status !== "trusted") {
        throw new Error(`#moveCertificate: unexpected status '${status}' for certificate ${fingerprint.substring(0, 10)}`);
      }
      if (status !== newStatus) {
        const indexSrc = status === "rejected" ? this.#thumbs.rejected : this.#thumbs.trusted;
        const srcEntry = indexSrc.get(fingerprint);
        if (!srcEntry) {
          debugLog(" cannot find certificate ", fingerprint.substring(0, 10), " in", status);
          throw new Error(`#moveCertificate: certificate ${fingerprint.substring(0, 10)} not found in ${status} index`);
        }
        const destFolder = newStatus === "trusted" ? this.trustedFolder : this.rejectedFolder;
        const certificateDest = import_node_path6.default.join(destFolder, import_node_path6.default.basename(srcEntry.filename));
        debugLog("#moveCertificate", fingerprint.substring(0, 10), "old name", srcEntry.filename);
        debugLog("#moveCertificate", fingerprint.substring(0, 10), "new name", certificateDest);
        await import_node_fs10.default.promises.rename(srcEntry.filename, certificateDest);
        indexSrc.delete(fingerprint);
        const indexDest = newStatus === "trusted" ? this.#thumbs.trusted : this.#thumbs.rejected;
        indexDest.set(fingerprint, { certificate, filename: certificateDest });
      }
    });
  }
  #findAssociatedCRLs(issuerCertificate) {
    const issuerCertificateInfo = (0, import_node_opcua_crypto5.exploreCertificate)(issuerCertificate);
    const key = issuerCertificateInfo.tbsCertificate.subjectFingerPrint;
    return this.#thumbs.issuersCrl.get(key) ?? this.#thumbs.crl.get(key) ?? null;
  }
  /**
   * Check whether a certificate has been revoked by its issuer's CRL.
   *
   * - Self-signed certificates are never considered revoked.
   * - If no `issuerCertificate` is provided, the method attempts
   *   to find it via {@link findIssuerCertificate}.
   *
   * @param certificate - the DER-encoded certificate to check
   * @param issuerCertificate - optional issuer certificate; looked
   *   up automatically when omitted
   * @returns `Good` if not revoked, `BadCertificateRevoked` if the
   *   serial number appears in a CRL,
   *   `BadCertificateRevocationUnknown` if no CRL is available,
   *   or `BadCertificateChainIncomplete` if the issuer cannot be
   *   found.
   */
  async isCertificateRevoked(certificate, issuerCertificate) {
    const chain = coerceCertificateChain(certificate);
    const firstCertificate = chain[0];
    if (isSelfSigned3(firstCertificate)) {
      return "Good" /* Good */;
    }
    if (!issuerCertificate) {
      issuerCertificate = await this.findIssuerCertificate(firstCertificate);
    }
    if (!issuerCertificate) {
      issuerCertificate = findIssuerCertificateInChain(firstCertificate, chain);
    }
    if (!issuerCertificate) {
      return "BadCertificateChainIncomplete" /* BadCertificateChainIncomplete */;
    }
    const crls = this.#findAssociatedCRLs(issuerCertificate);
    if (!crls) {
      return "BadCertificateRevocationUnknown" /* BadCertificateRevocationUnknown */;
    }
    const certInfo = (0, import_node_opcua_crypto5.exploreCertificate)(firstCertificate);
    const serialNumber = certInfo.tbsCertificate.serialNumber || certInfo.tbsCertificate.extensions?.authorityKeyIdentifier?.serial || "";
    const key = certInfo.tbsCertificate.extensions?.authorityKeyIdentifier?.authorityCertIssuerFingerPrint || "<unknown>";
    const crl2 = this.#thumbs.crl.get(key) ?? null;
    if (crls.serialNumbers[serialNumber] || crl2?.serialNumbers[serialNumber]) {
      return "BadCertificateRevoked" /* BadCertificateRevoked */;
    }
    return "Good" /* Good */;
  }
  #pendingCrlToProcess = 0;
  #onCrlProcessWaiters = [];
  #queue = [];
  #onCrlFileAdded(index, filename) {
    this.#queue.push({ index, filename });
    this.#pendingCrlToProcess += 1;
    if (this.#pendingCrlToProcess === 1) {
      this.#processNextCrl();
    }
  }
  async #processNextCrl() {
    try {
      const nextCRL = this.#queue.shift();
      if (!nextCRL) return;
      const { index, filename } = nextCRL;
      const crl = await (0, import_node_opcua_crypto5.readCertificateRevocationList)(filename);
      const crlInfo = (0, import_node_opcua_crypto5.exploreCertificateRevocationList)(crl);
      debugLog(import_chalk6.default.cyan("add CRL in folder "), filename);
      const fingerprint = crlInfo.tbsCertList.issuerFingerprint;
      if (!index.has(fingerprint)) {
        index.set(fingerprint, { crls: [], serialNumbers: {} });
      }
      const data = index.get(fingerprint) || { crls: [], serialNumbers: {} };
      data.crls.push({ crlInfo, filename });
      for (const revokedCertificate of crlInfo.tbsCertList.revokedCertificates) {
        const serialNumber = revokedCertificate.userCertificate;
        if (!data.serialNumbers[serialNumber]) {
          data.serialNumbers[serialNumber] = revokedCertificate.revocationDate;
        }
      }
      debugLog(import_chalk6.default.cyan("CRL"), fingerprint, "serial numbers = ", Object.keys(data.serialNumbers));
    } catch (err) {
      debugLog("CRL filename error =");
      debugLog(err);
    }
    this.#pendingCrlToProcess -= 1;
    if (this.#pendingCrlToProcess === 0) {
      for (const waiter of this.#onCrlProcessWaiters) {
        waiter();
      }
      this.#onCrlProcessWaiters.length = 0;
    } else {
      this.#processNextCrl();
    }
  }
  async #readCertificates() {
    if (this.#readCertificatesCalled) {
      return;
    }
    this.#readCertificatesCalled = true;
    const usePolling = process.env.OPCUA_PKI_USE_POLLING === "true";
    const envInterval = process.env.OPCUA_PKI_POLLING_INTERVAL ? parseInt(process.env.OPCUA_PKI_POLLING_INTERVAL, 10) : void 0;
    const pollingInterval = Math.min(10 * 60 * 1e3, Math.max(100, envInterval ?? this.folderPollingInterval));
    const chokidarOptions = {
      usePolling,
      ...usePolling ? { interval: pollingInterval } : {},
      persistent: false
    };
    const allCapturedHandles = [];
    const origWatch = import_node_fs10.default.watch;
    let watcherReadyCount = 0;
    const totalWatchers = 5;
    import_node_fs10.default.watch = ((...args) => {
      const handle = origWatch.apply(import_node_fs10.default, args);
      handle.setMaxListeners(handle.getMaxListeners() + 1);
      handle.on("error", () => {
      });
      allCapturedHandles.push(handle);
      return handle;
    });
    const createUnreffedWatcher = (folder) => {
      const startIdx = allCapturedHandles.length;
      const w = import_chokidar.default.watch(folder, chokidarOptions);
      const unreffAll = () => {
        for (let i = startIdx; i < allCapturedHandles.length; i++) {
          allCapturedHandles[i].unref();
        }
        watcherReadyCount++;
        if (watcherReadyCount >= totalWatchers) {
          import_node_fs10.default.watch = origWatch;
        }
      };
      return { w, capturedHandles: allCapturedHandles.slice(startIdx), unreffAll };
    };
    await Promise.all([
      this.#scanCertFolder(this.trustedFolder, this.#thumbs.trusted),
      this.#scanCertFolder(this.issuersCertFolder, this.#thumbs.issuers.certs),
      this.#scanCertFolder(this.rejectedFolder, this.#thumbs.rejected),
      this.#scanCrlFolder(this.crlFolder, this.#thumbs.crl),
      this.#scanCrlFolder(this.issuersCrlFolder, this.#thumbs.issuersCrl)
    ]);
    if (this.#disableFileWatchers) {
      import_node_fs10.default.watch = origWatch;
    } else {
      this.#startWatcher(this.trustedFolder, this.#thumbs.trusted, createUnreffedWatcher, "trusted");
      this.#startWatcher(this.issuersCertFolder, this.#thumbs.issuers.certs, createUnreffedWatcher, "issuersCerts");
      this.#startWatcher(this.rejectedFolder, this.#thumbs.rejected, createUnreffedWatcher, "rejected");
      this.#startCrlWatcher(this.crlFolder, this.#thumbs.crl, createUnreffedWatcher, "crl");
      this.#startCrlWatcher(this.issuersCrlFolder, this.#thumbs.issuersCrl, createUnreffedWatcher, "issuersCrl");
    }
  }
  /**
   * Scan a certificate folder and populate the in-memory index.
   * Uses async readdir/stat to yield the event loop between
   * file reads, preventing main-loop stalls with large folders.
   */
  async #scanCertFolder(folder, index) {
    if (!import_node_fs10.default.existsSync(folder)) return;
    const files = await import_node_fs10.default.promises.readdir(folder);
    for (const file of files) {
      const filename = import_node_path6.default.join(folder, file);
      try {
        const stat = await import_node_fs10.default.promises.stat(filename);
        if (!stat.isFile()) continue;
        const certs = await (0, import_node_opcua_crypto5.readCertificateChainAsync)(filename);
        if (certs.length === 0) continue;
        const certificate = certs[0];
        if (certs.length > 1) {
          try {
            await import_node_fs10.default.promises.writeFile(filename, (0, import_node_opcua_crypto5.toPem)(certs, "CERTIFICATE"), "ascii");
          } catch (writeErr) {
            debugLog(`scanCertFolder: could not rewrite legacy PEM ${filename} (read-only fs?)`, writeErr);
          }
          for (let i = 1; i < certs.length; i++) {
            if (isIssuer(certs[i])) {
              try {
                await this.addIssuer(certs[i]);
              } catch (issuerErr) {
                debugLog(`scanCertFolder: could not auto-register issuer from ${filename}`, issuerErr);
              }
            }
          }
        }
        const info = (0, import_node_opcua_crypto5.exploreCertificate)(certificate);
        const fingerprint = makeFingerprint(certificate);
        index.set(fingerprint, { certificate, filename, info });
        this.#filenameToHash.set(filename, fingerprint);
      } catch (err) {
        debugLog(`scanCertFolder: skipping ${filename}`, err);
      }
    }
  }
  /**
   * Scan a CRL folder and populate the in-memory CRL index.
   */
  async #scanCrlFolder(folder, index) {
    if (!import_node_fs10.default.existsSync(folder)) return;
    const files = await import_node_fs10.default.promises.readdir(folder);
    for (const file of files) {
      const filename = import_node_path6.default.join(folder, file);
      try {
        const stat = await import_node_fs10.default.promises.stat(filename);
        if (!stat.isFile()) continue;
        this.#onCrlFileAdded(index, filename);
      } catch (err) {
        debugLog(`scanCrlFolder: skipping ${filename}`, err);
      }
    }
    await this.#waitAndCheckCRLProcessingStatus();
  }
  /**
   * Start a chokidar watcher for a CRL folder.
   * Non-blocking — does NOT await "ready".
   */
  #startCrlWatcher(folder, index, createUnreffedWatcher, store) {
    const { w, unreffAll } = createUnreffedWatcher(folder);
    w.on("error", (err) => {
      debugLog(`chokidar CRL watcher error on ${folder}:`, err);
    });
    let ready = false;
    w.on("unlink", (filename) => {
      for (const [key, data] of index.entries()) {
        data.crls = data.crls.filter((c) => c.filename !== filename);
        if (data.crls.length === 0) {
          index.delete(key);
        }
      }
      if (ready) {
        this.emit("crlRemoved", { store, filename });
      }
    });
    w.on("add", (filename) => {
      if (ready) {
        this.#onCrlFileAdded(index, filename);
        this.emit("crlAdded", { store, filename });
      }
    });
    w.on("change", (changedPath) => {
      debugLog("change in folder ", folder, changedPath);
    });
    this.#watchers.push(w);
    this.#pendingUnrefs.add(unreffAll);
    w.on("ready", () => {
      ready = true;
      this.#pendingUnrefs.delete(unreffAll);
      unreffAll();
    });
  }
  /**
   * Start a chokidar watcher for a certificate folder.
   * Non-blocking — does NOT await "ready".
   */
  #startWatcher(folder, index, createUnreffedWatcher, store) {
    const { w, unreffAll } = createUnreffedWatcher(folder);
    w.on("error", (err) => {
      debugLog(`chokidar cert watcher error on ${folder}:`, err);
    });
    let ready = false;
    w.on("unlink", (filename) => {
      debugLog(import_chalk6.default.cyan(`unlink in folder ${folder}`), filename);
      const h = this.#filenameToHash.get(filename);
      if (h && index.has(h)) {
        index.delete(h);
        this.emit("certificateRemoved", { store, fingerprint: h, filename });
      }
    });
    w.on("add", (filename) => {
      debugLog(import_chalk6.default.cyan(`add in folder ${folder}`), filename);
      try {
        const certificate = (0, import_node_opcua_crypto5.readCertificateChain)(filename)[0];
        const info = (0, import_node_opcua_crypto5.exploreCertificate)(certificate);
        const fingerprint = makeFingerprint(certificate);
        const isNew = !index.has(fingerprint);
        index.set(fingerprint, { certificate, filename, info });
        this.#filenameToHash.set(filename, fingerprint);
        debugLog(
          import_chalk6.default.magenta("CERT"),
          info.tbsCertificate.subjectFingerPrint,
          info.tbsCertificate.serialNumber,
          info.tbsCertificate.extensions?.authorityKeyIdentifier?.authorityCertIssuerFingerPrint
        );
        if (ready || isNew) {
          this.emit("certificateAdded", { store, certificate, fingerprint, filename });
        }
      } catch (err) {
        debugLog(`Walk files in folder ${folder} with file ${filename}`);
        debugLog(err);
      }
    });
    w.on("change", (changedPath) => {
      debugLog(import_chalk6.default.cyan(`change in folder ${folder}`), changedPath);
      try {
        const certificate = (0, import_node_opcua_crypto5.readCertificateChain)(changedPath)[0];
        const newFingerprint = makeFingerprint(certificate);
        const oldHash = this.#filenameToHash.get(changedPath);
        if (oldHash && oldHash !== newFingerprint) {
          index.delete(oldHash);
        }
        index.set(newFingerprint, { certificate, filename: changedPath, info: (0, import_node_opcua_crypto5.exploreCertificate)(certificate) });
        this.#filenameToHash.set(changedPath, newFingerprint);
        this.emit("certificateChange", { store, certificate, fingerprint: newFingerprint, filename: changedPath });
      } catch (err) {
        debugLog(`change event: failed to re-read ${changedPath}`, err);
      }
    });
    this.#watchers.push(w);
    this.#pendingUnrefs.add(unreffAll);
    w.on("ready", () => {
      ready = true;
      this.#pendingUnrefs.delete(unreffAll);
      unreffAll();
      debugLog("ready");
      debugLog([...index.keys()].map((k) => k.substring(0, 10)));
    });
  }
  // make sure that all crls have been processed.
  async #waitAndCheckCRLProcessingStatus() {
    return new Promise((resolve, _reject) => {
      if (this.#pendingCrlToProcess === 0) {
        setImmediate(resolve);
        return;
      }
      this.#onCrlProcessWaiters.push(resolve);
    });
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CertificateAuthority,
  CertificateManager,
  CertificateManagerState,
  ChainCompletionStatus,
  Subject,
  VerificationStatus,
  adjustApplicationUri,
  adjustDate,
  coerceCertificateChain,
  convertPFXtoPEM,
  createPFX,
  dumpPFX,
  extractAllFromPFX,
  extractCACertificatesFromPFX,
  extractCertificateFromPFX,
  extractPrivateKeyFromPFX,
  findIssuerCertificateInChain,
  install_prerequisite,
  isIntermediateIssuer,
  isIssuer,
  isRootIssuer,
  makeFingerprint,
  quote
});
