// node-opcua-constants/dist/well_known_roles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WellKnownRoles = void 0;
var WellKnownRoles;
(function (WellKnownRoles) {
    /** Very limited access for anonymous Sessions. */
    WellKnownRoles[WellKnownRoles["Anonymous"] = 15644] = "Anonymous";
    /** Limited access for valid non-anonymous credentials. */
    WellKnownRoles[WellKnownRoles["AuthenticatedUser"] = 15656] = "AuthenticatedUser";
    /** Browse, read live/historical data, subscribe. */
    WellKnownRoles[WellKnownRoles["Observer"] = 15668] = "Observer";
    /** Observer + write some live data, call some Methods. */
    WellKnownRoles[WellKnownRoles["Operator"] = 15680] = "Operator";
    /** Browse, read/write config, read historical, call Methods, subscribe. */
    WellKnownRoles[WellKnownRoles["Engineer"] = 16036] = "Engineer";
    /** Browse, read live/historical data, call Methods, subscribe. */
    WellKnownRoles[WellKnownRoles["Supervisor"] = 15692] = "Supervisor";
    /** Change non-security related configuration. */
    WellKnownRoles[WellKnownRoles["ConfigureAdmin"] = 15716] = "ConfigureAdmin";
    /** Change security related settings. */
    WellKnownRoles[WellKnownRoles["SecurityAdmin"] = 15704] = "SecurityAdmin";
})(WellKnownRoles || (exports.WellKnownRoles = WellKnownRoles = {}));
