// node-opcua-service-filter/dist/check_where_clause.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateContentFilter = validateContentFilter;
exports.checkFilter = checkFilter;
const node_opcua_types_1 = require("node-opcua-types");
const node_opcua_variant_1 = require("node-opcua-variant");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const node_opcua_status_code_1 = require("node-opcua-status-code");
const node_opcua_debug_1 = require("node-opcua-debug");
const resolve_operand_1 = require("./resolve_operand");
const warningLog = (0, node_opcua_debug_1.make_warningLog)("Filter");
function _coerceToBoolean(value) {
    if (value instanceof node_opcua_variant_1.Variant) {
        return !!value.value;
    }
    return !!value;
}
function _coerceToCompareable(value) {
    if (value instanceof node_opcua_variant_1.Variant) {
        switch (value.dataType) {
            case node_opcua_variant_1.DataType.String:
                return value.value;
            case node_opcua_variant_1.DataType.Byte:
            case node_opcua_variant_1.DataType.SByte:
            case node_opcua_variant_1.DataType.Int16:
            case node_opcua_variant_1.DataType.Int32:
            case node_opcua_variant_1.DataType.UInt16:
            case node_opcua_variant_1.DataType.UInt32:
                return value.value;
            case node_opcua_variant_1.DataType.Double:
                return value.value;
            case node_opcua_variant_1.DataType.DateTime:
                return value.value.getTime();
            case node_opcua_variant_1.DataType.Guid:
                return value.value.toLowerCase();
            case node_opcua_variant_1.DataType.ByteString:
                return value.value.toString("hex");
            case node_opcua_variant_1.DataType.XmlElement:
                return value.value;
            case node_opcua_variant_1.DataType.LocalizedText:
                return value.value.text;
            case node_opcua_variant_1.DataType.QualifiedName:
                return value.value.name; // not sure about this one
            default:
                return "";
        }
    }
    return "";
}
function _coerceToEqualable(value) {
    return value;
}
function evaluateOperand(filterContext, filter, operand, coerce) {
    if (operand instanceof node_opcua_types_1.AttributeOperand) {
        return coerce((0, resolve_operand_1.resolveOperand)(filterContext, operand));
    }
    else if (operand instanceof node_opcua_types_1.SimpleAttributeOperand) {
        return coerce((0, resolve_operand_1.resolveOperand)(filterContext, operand));
    }
    else if (operand instanceof node_opcua_types_1.LiteralOperand) {
        return coerce(operand.value);
    }
    else if (operand instanceof node_opcua_types_1.ElementOperand) {
        const index = operand.index;
        return coerce(checkFilterAtIndex(filterContext, filter, index));
    }
    // istanbul ignore
    return coerce(null);
}
function checkOfType(filterContext, ofType) {
    // c8 ignore next
    if (!ofType || !(ofType instanceof node_opcua_types_1.LiteralOperand)) {
        warningLog("checkOfType : unsupported case ! ofType is not a LiteralOperand , ofType = ", ofType?.toString());
        return false;
    }
    // c8 ignore next
    if (ofType.value.dataType !== node_opcua_variant_1.DataType.NodeId) {
        warningLog("invalid operand type (expecting NodeId); got " + node_opcua_variant_1.DataType[ofType.value.dataType]);
        return false;
    }
    const ofTypeNode = ofType.value.value;
    // c8 ignore next
    if (!ofTypeNode) {
        return false; // the ofType node is not known, we don't know what to do
    }
    const ofTypeNodeNodeClass = filterContext.getNodeClass(ofTypeNode);
    // c8 ignore next
    if (ofTypeNodeNodeClass !== node_opcua_data_model_1.NodeClass.ObjectType &&
        ofTypeNodeNodeClass !== node_opcua_data_model_1.NodeClass.VariableType &&
        ofTypeNodeNodeClass !== node_opcua_data_model_1.NodeClass.DataType &&
        ofTypeNodeNodeClass !== node_opcua_data_model_1.NodeClass.ReferenceType) {
        warningLog("operand should be a ObjectType " + ofTypeNode.toString());
        return false;
    }
    if (!filterContext.eventSource || filterContext.eventSource.isEmpty()) {
        return false;
    }
    let sourceTypeDefinition = filterContext.eventSource;
    const sourceNodeClass = filterContext.getNodeClass(filterContext.eventSource);
    if (sourceNodeClass === node_opcua_data_model_1.NodeClass.Object || sourceNodeClass === node_opcua_data_model_1.NodeClass.Variable) {
        sourceTypeDefinition = filterContext.getTypeDefinition(filterContext.eventSource);
        if (!sourceTypeDefinition) {
            return false;
        }
    }
    return filterContext.isSubtypeOf(sourceTypeDefinition, ofTypeNode);
}
function checkNot(filterContext, filter, filteredOperands) {
    const operandA = evaluateOperand(filterContext, filter, filteredOperands[0], _coerceToBoolean);
    return !operandA;
}
function checkOr(filterContext, filter, filteredOperands) {
    const operandA = evaluateOperand(filterContext, filter, filteredOperands[0], _coerceToBoolean);
    const operandB = evaluateOperand(filterContext, filter, filteredOperands[1], _coerceToBoolean);
    return operandA || operandB;
}
/**
 *
 * TRUE if operand[0] and operand[1] are TRUE.
 * The following restrictions apply to the operands:
 *  [0]: Any operand that resolves to a Boolean.
 *
 *   [1]: Any operand that resolves to a Boolean.
 * If any operand cannot be resolved to a Boolean it is considered a NULL. See below for a discussion on the handling of NULL.
 */
function checkAnd(filterContext, filter, filteredOperands) {
    const operandA = evaluateOperand(filterContext, filter, filteredOperands[0], _coerceToBoolean);
    const operandB = evaluateOperand(filterContext, filter, filteredOperands[1], _coerceToBoolean);
    return operandA && operandB;
}
function checkLessThan(filterContext, filter, filteredOperands) {
    const operandA = evaluateOperand(filterContext, filter, filteredOperands[0], _coerceToCompareable);
    const operandB = evaluateOperand(filterContext, filter, filteredOperands[1], _coerceToCompareable);
    return operandA < operandB;
}
function checkLessThanOrEqual(filterContext, filter, filteredOperands) {
    const operandA = evaluateOperand(filterContext, filter, filteredOperands[0], _coerceToCompareable);
    const operandB = evaluateOperand(filterContext, filter, filteredOperands[1], _coerceToCompareable);
    return operandA <= operandB;
}
function checkGreaterThan(filterContext, filter, filteredOperands) {
    const operandA = evaluateOperand(filterContext, filter, filteredOperands[0], _coerceToCompareable);
    const operandB = evaluateOperand(filterContext, filter, filteredOperands[1], _coerceToCompareable);
    return operandA > operandB;
}
function checkGreaterThanOrEqual(filterContext, filter, filteredOperands) {
    const operandA = evaluateOperand(filterContext, filter, filteredOperands[0], _coerceToCompareable);
    const operandB = evaluateOperand(filterContext, filter, filteredOperands[1], _coerceToCompareable);
    return operandA >= operandB;
}
const isVariantEqual = (a, b) => {
    switch (a.dataType) {
        case node_opcua_variant_1.DataType.Null:
            return b.dataType === node_opcua_variant_1.DataType.Null;
        case node_opcua_variant_1.DataType.Boolean:
            return a.value === b.value;
        case node_opcua_variant_1.DataType.Byte:
        case node_opcua_variant_1.DataType.SByte:
        case node_opcua_variant_1.DataType.Int16:
        case node_opcua_variant_1.DataType.Int32:
        case node_opcua_variant_1.DataType.UInt16:
        case node_opcua_variant_1.DataType.UInt32:
            return a.value === b.value;
        case node_opcua_variant_1.DataType.Double:
            return a.value === b.value;
        case node_opcua_variant_1.DataType.String:
            return a.value === b.value;
        case node_opcua_variant_1.DataType.NodeId:
            return (0, node_opcua_nodeid_1.sameNodeId)(a.value, b.value);
        case node_opcua_variant_1.DataType.DateTime:
            return a.value?.getTime() === b.value.getTime();
        case node_opcua_variant_1.DataType.Guid:
            return a.value.toLowerCase() === ("" + (b.value || "")).toLowerCase();
        case node_opcua_variant_1.DataType.ByteString:
            if (b.dataType !== node_opcua_variant_1.DataType.ByteString) {
                return false;
            }
            return a.value.toString("hex") === b.value.toString("hex");
        case node_opcua_variant_1.DataType.XmlElement:
            return a.value === b.value;
        case node_opcua_variant_1.DataType.LocalizedText:
            return a.value?.text === b.value?.text;
        case node_opcua_variant_1.DataType.QualifiedName:
            return a.value?.namespaceIndex === b.value?.namespaceIndex && a.value?.name === b.value?.name;
        case node_opcua_variant_1.DataType.ExtensionObject:
            console.log("isVariantEqual: Not implemented for DataType.ExtensionObject");
        default:
            return false; // not sure how to do
    }
};
function checkEquals(filterContext, filter, filteredOperands) {
    const operandA = evaluateOperand(filterContext, filter, filteredOperands[0], _coerceToEqualable);
    const operandB = evaluateOperand(filterContext, filter, filteredOperands[1], _coerceToEqualable);
    return isVariantEqual(operandA, operandB);
}
/**
 *
 * TRUE if operand[0] is greater or equal to operand[1] and less than or equal to operand[2].
 * The following restrictions apply to the operands:
 *   [0]: Any operand that resolves to an ordered value.
 *   [1]: Any operand that resolves to an ordered value.
 *   [2]: Any operand that resolves to an ordered value.
 * If the operands are of different types, the system shall perform any implicit conversion to match
 * all operands to a common type. If no implicit conversion is available and the operands are of different
 * types, the particular result is FALSE. See the discussion on data type precedence in Table 123
 * for more information how to convert operands of different types.
 */
function checkBetween(filterContext, filter, filteredOperands) {
    const operandA = evaluateOperand(filterContext, filter, filteredOperands[0], _coerceToCompareable);
    const operandLow = evaluateOperand(filterContext, filter, filteredOperands[1], _coerceToCompareable);
    const operandHigh = evaluateOperand(filterContext, filter, filteredOperands[2], _coerceToCompareable);
    return operandA >= operandLow && operandA <= operandHigh;
}
/**
 *
 * InList
 * TRUE if operand[0] is equal to one or more of the remaining operands.
 * The Equals Operator is evaluated for operand[0] and each remaining operand in the
 * list. If any Equals evaluation is TRUE, InList returns TRUE
 x*/
function checkInList(context, filterOperands) {
    const operand0 = filterOperands[0];
    // c8 ignore next
    if (!(operand0 instanceof node_opcua_types_1.SimpleAttributeOperand)) {
        // unsupported case
        warningLog("FilterOperator.InList operand0 is not a SimpleAttributeOperand " + operand0.constructor.name);
        return false;
    }
    const value = (0, resolve_operand_1.resolveOperand)(context, operand0);
    // c8 ignore next
    if (value.dataType !== node_opcua_variant_1.DataType.NodeId) {
        return false;
    }
    const nodeId = value.value;
    // c8 ignore next
    if (!nodeId) {
        return false;
    }
    function _is(nodeId1, operandX) {
        if (operandX.value.dataType !== node_opcua_variant_1.DataType.NodeId) {
            return false;
        }
        const nodeId2 = operandX.value.value;
        const nodeClass = context.getNodeClass(nodeId2);
        if (nodeClass === node_opcua_data_model_1.NodeClass.Unspecified) {
            return false;
        }
        return (0, node_opcua_nodeid_1.sameNodeId)(nodeId1, nodeId2);
    }
    for (let i = 1; i < filterOperands.length; i++) {
        const filterOperand = filterOperands[i];
        if (filterOperand instanceof node_opcua_types_1.LiteralOperand && _is(nodeId, filterOperand)) {
            return true;
        }
    }
    return false;
}
// eslint-disable-next-line complexity
function checkFilterAtIndex(filterContext, filter, index) {
    if (!filter.elements || filter.elements.length === 0) {
        return true;
    }
    const element = filter.elements[index];
    // c8 ignore next
    if (!element) {
        return true;
    }
    const filterOperands = element.filterOperands || [];
    switch (element.filterOperator) {
        case node_opcua_types_1.FilterOperator.Equals:
            return checkEquals(filterContext, filter, filterOperands);
        case node_opcua_types_1.FilterOperator.LessThan:
            return checkLessThan(filterContext, filter, filterOperands);
        case node_opcua_types_1.FilterOperator.LessThanOrEqual:
            return checkLessThanOrEqual(filterContext, filter, filterOperands);
        case node_opcua_types_1.FilterOperator.GreaterThan:
            return checkGreaterThan(filterContext, filter, filterOperands);
        case node_opcua_types_1.FilterOperator.GreaterThanOrEqual:
            return checkGreaterThanOrEqual(filterContext, filter, filterOperands);
        case node_opcua_types_1.FilterOperator.Between:
            return checkBetween(filterContext, filter, filterOperands);
        case node_opcua_types_1.FilterOperator.And:
            return checkAnd(filterContext, filter, filterOperands);
        case node_opcua_types_1.FilterOperator.Or:
            return checkOr(filterContext, filter, filterOperands);
        case node_opcua_types_1.FilterOperator.Not:
            return checkNot(filterContext, filter, filterOperands);
        case node_opcua_types_1.FilterOperator.OfType:
            return checkOfType(filterContext, element.filterOperands[0]);
        case node_opcua_types_1.FilterOperator.InList:
            return checkInList(filterContext, filterOperands);
        case node_opcua_types_1.FilterOperator.RelatedTo:
        case node_opcua_types_1.FilterOperator.Like:
        case node_opcua_types_1.FilterOperator.BitwiseAnd:
        case node_opcua_types_1.FilterOperator.BitwiseOr:
        case node_opcua_types_1.FilterOperator.Cast:
        case node_opcua_types_1.FilterOperator.InView:
        case node_opcua_types_1.FilterOperator.IsNull:
        default:
            // from Spec  OPC Unified Architecture, Part 4 133 Release 1.04
            //  Any basic FilterOperator in Table 119 may be used in the whereClause, however, only the
            //  OfType_14 FilterOperator from Table 120 is permitted.
            warningLog(`checkFilter: operator ${node_opcua_types_1.FilterOperator[element.filterOperator]} is currently not supported in filter`);
            return false;
    }
}
// Number of operands each FilterOperator expects, as per OPC UA Part 4 - Table 118 / Table 119.
// `max = Infinity` is used for the variadic InList operator (2..n). Operators not listed here are
// not subject to an operand-count check.
const operandCountByOperator = {
    [node_opcua_types_1.FilterOperator.Equals]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.IsNull]: { min: 1, max: 1 },
    [node_opcua_types_1.FilterOperator.GreaterThan]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.LessThan]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.GreaterThanOrEqual]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.LessThanOrEqual]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.Like]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.Not]: { min: 1, max: 1 },
    [node_opcua_types_1.FilterOperator.Between]: { min: 3, max: 3 },
    [node_opcua_types_1.FilterOperator.InList]: { min: 2, max: Number.POSITIVE_INFINITY },
    [node_opcua_types_1.FilterOperator.And]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.Or]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.Cast]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.BitwiseAnd]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.BitwiseOr]: { min: 2, max: 2 },
    [node_opcua_types_1.FilterOperator.InView]: { min: 1, max: 1 },
    [node_opcua_types_1.FilterOperator.OfType]: { min: 1, max: 1 },
    [node_opcua_types_1.FilterOperator.RelatedTo]: { min: 6, max: 6 }
};
/**
 * Validate the structure of a ContentFilter, returning a precise StatusCode (Good when valid).
 *
 * The following OPC UA Part 4 - 7.7 rules are enforced over the elements that are reachable from the
 * starting element (index 0); elements that cannot be traced back to the starting element are
 * ignored (7.7.1):
 *  - each operator carries the number of operands defined by Table 118 / Table 119, otherwise
 *    `BadFilterOperandCountMismatch` ("Extra operands for any operator shall result in an error");
 *  - each ElementOperand references an existing element (in-bounds), otherwise
 *    `BadFilterElementInvalid`;
 *  - the reference graph is acyclic, otherwise `BadFilterElementInvalid` ("The evaluation shall not
 *    introduce loops").
 *
 * The traversal is iterative so that validating a deeply-nested filter cannot itself exhaust the
 * stack.
 */
function validateContentFilter(filter) {
    const elements = filter.elements || [];
    const n = elements.length;
    if (n === 0) {
        return node_opcua_status_code_1.StatusCodes.Good;
    }
    // inspect a single element: validate its operand count and collect the in-bounds indices it
    // references through ElementOperands. `status` is non-Good when the element is malformed.
    const inspect = (elementIndex) => {
        const element = elements[elementIndex];
        // an element of the array may be null/undefined (e.g. a null ExtensionObject); see checkFilterAtIndex
        if (!element) {
            return { targets: [], status: node_opcua_status_code_1.StatusCodes.Good };
        }
        const operands = element.filterOperands || [];
        const expected = operandCountByOperator[element.filterOperator];
        if (expected && (operands.length < expected.min || operands.length > expected.max)) {
            warningLog(`checkFilter: operator ${node_opcua_types_1.FilterOperator[element.filterOperator]} of element #${elementIndex} ` +
                `has ${operands.length} operand(s), expected ${expected.min}${expected.max === expected.min ? "" : `..${expected.max === Number.POSITIVE_INFINITY ? "n" : expected.max}`}`);
            return { targets: [], status: node_opcua_status_code_1.StatusCodes.BadFilterOperandCountMismatch };
        }
        const targets = [];
        for (const operand of operands) {
            if (operand instanceof node_opcua_types_1.ElementOperand) {
                const index = operand.index;
                if (index < 0 || index >= n) {
                    warningLog(`checkFilter: ElementOperand of element #${elementIndex} references out-of-bounds element #${index} ` +
                        `(elements.length = ${n})`);
                    return { targets: [], status: node_opcua_status_code_1.StatusCodes.BadFilterElementInvalid };
                }
                targets.push(index);
            }
        }
        return { targets, status: node_opcua_status_code_1.StatusCodes.Good };
    };
    // 0 = unvisited, 1 = on the current DFS path, 2 = fully explored
    const color = new Array(n).fill(0);
    const root = inspect(0);
    if (root.status.isNot(node_opcua_status_code_1.StatusCodes.Good)) {
        return root.status;
    }
    color[0] = 1;
    const stack = [{ node: 0, targets: root.targets, pos: 0 }];
    while (stack.length > 0) {
        const frame = stack[stack.length - 1];
        if (frame.pos >= frame.targets.length) {
            color[frame.node] = 2;
            stack.pop();
            continue;
        }
        const next = frame.targets[frame.pos++];
        if (color[next] === 1) {
            // a back edge to an element still on the current path -> cycle
            warningLog(`checkFilter: cyclic ElementOperand reference detected involving element #${next}`);
            return node_opcua_status_code_1.StatusCodes.BadFilterElementInvalid;
        }
        if (color[next] === 0) {
            const r = inspect(next);
            if (r.status.isNot(node_opcua_status_code_1.StatusCodes.Good)) {
                return r.status;
            }
            color[next] = 1;
            stack.push({ node: next, targets: r.targets, pos: 0 });
        }
    }
    return node_opcua_status_code_1.StatusCodes.Good;
}
function checkFilter(filterContext, contentFilter) {
    if (validateContentFilter(contentFilter).isNot(node_opcua_status_code_1.StatusCodes.Good)) {
        return false;
    }
    return checkFilterAtIndex(filterContext, contentFilter, 0);
}
