// node-opcua-common/dist/certificate_chain_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticCertificateChainProvider = void 0;
/**
 * Holds a certificate chain and private key in memory.
 *
 * Used as the default provider when push certificate management
 * is NOT installed. The chain can be replaced in-place via `update()`.
 */
class StaticCertificateChainProvider {
    #chain;
    #key;
    constructor(chain, key) {
        this.#chain = chain;
        this.#key = key;
    }
    get certificateFile() {
        return "<in-memory>";
    }
    get privateKeyFile() {
        return "<in-memory>";
    }
    getCertificate() {
        return this.#chain[0];
    }
    getCertificateChain() {
        return this.#chain;
    }
    getPrivateKey() {
        return this.#key;
    }
    /**
     * No-op for static provider — the chain is already in memory.
     * Use `update()` to replace the chain explicitly.
     */
    invalidate() {
        // nothing to invalidate for a static provider
    }
    /**
     * Replace the certificate chain and optionally the private key.
     *
     * This immediately affects all consumers that call
     * `getCertificateChain()` on this provider (including
     * endpoint descriptions with dynamic `serverCertificate` getters).
     */
    update(chain, key) {
        if (chain.length === 0) {
            throw new Error("StaticCertificateChainProvider.update: chain must not be empty");
        }
        this.#chain = chain;
        if (key !== undefined) {
            this.#key = key;
        }
    }
    // Prevent secrets from leaking through JSON serialization
    toJSON() {
        return { provider: "StaticCertificateChainProvider" };
    }
    // Prevent secrets from leaking through console.log / util.inspect
    [Symbol.for("nodejs.util.inspect.custom")]() {
        return "StaticCertificateChainProvider { <in-memory> }";
    }
}
exports.StaticCertificateChainProvider = StaticCertificateChainProvider;
