// node-opcua-common/dist/disk_certificate_key_pair_provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiskCertificateKeyPairProvider = void 0;
/**
 * Disk-backed certificate/key provider.
 *
 * Reads certificate chains and private keys lazily from the filesystem
 * and caches them in `#`-private fields so they never appear in
 * `JSON.stringify`, `console.log`, `util.inspect`, or debugger panels.
 *
 * @module node-opcua-common
 */
const node_fs_1 = __importDefault(require("node:fs"));
const node_opcua_crypto_1 = require("node-opcua-crypto");
class DiskCertificateKeyPairProvider {
    #certificateChain = null;
    #privateKey = null;
    #certificateFile;
    #privateKeyFile;
    constructor(certificateFile, privateKeyFile) {
        this.#certificateFile = certificateFile;
        this.#privateKeyFile = privateKeyFile;
    }
    get certificateFile() {
        return this.#certificateFile;
    }
    get privateKeyFile() {
        return this.#privateKeyFile;
    }
    getCertificate() {
        return this.getCertificateChain()[0];
    }
    getCertificateChain() {
        if (!this.#certificateChain) {
            if (!node_fs_1.default.existsSync(this.#certificateFile)) {
                throw new Error(`Certificate file not found: ${this.#certificateFile}`);
            }
            const chain = (0, node_opcua_crypto_1.readCertificateChain)(this.#certificateFile);
            if (!chain || chain.length === 0) {
                throw new Error(`Invalid certificate chain (length=0) ${this.#certificateFile}`);
            }
            this.#certificateChain = chain;
        }
        return this.#certificateChain;
    }
    getPrivateKey() {
        if (!this.#privateKey) {
            if (!node_fs_1.default.existsSync(this.#privateKeyFile)) {
                throw new Error(`Private key file not found: ${this.#privateKeyFile}`);
            }
            const key = (0, node_opcua_crypto_1.readPrivateKey)(this.#privateKeyFile);
            if (key instanceof Buffer) {
                throw new Error(`Invalid private key ${this.#privateKeyFile}. Should not be a buffer`);
            }
            this.#privateKey = key;
        }
        return this.#privateKey;
    }
    /**
     * Clears cached secrets so the GC can reclaim sensitive material.
     * After calling dispose the holder will re-read from disk on next access.
     */
    dispose() {
        this.#certificateChain = null;
        this.#privateKey = null;
    }
    /**
     * Alias for {@link dispose}.
     * Implements `ICertificateChainProvider.invalidate()`.
     */
    invalidate() {
        this.dispose();
    }
    // Prevent secrets from leaking through JSON serialization
    toJSON() {
        return {
            provider: "DiskCertificateKeyPairProvider",
            certificateFile: this.#certificateFile,
            privateKeyFile: this.#privateKeyFile
        };
    }
    // Prevent secrets from leaking through console.log / util.inspect
    [Symbol.for("nodejs.util.inspect.custom")]() {
        return `DiskCertificateKeyPairProvider { certificateFile: "${this.#certificateFile}", privateKeyFile: "${this.#privateKeyFile}" }`;
    }
}
exports.DiskCertificateKeyPairProvider = DiskCertificateKeyPairProvider;
