// node-opcua-common/dist/in_memory_certificate_key_pair_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryCertificateKeyPairProvider = void 0;
/**
 * In-memory certificate/key provider with on-demand provisioning.
 *
 * Can be constructed with pre-existing cert+key, or can
 * auto-provision a self-signed certificate entirely in-memory
 * using `node-opcua-crypto/web` — no disk, no openssl.
 *
 * @module node-opcua-common
 */
const web_1 = require("node-opcua-crypto/web");
class InMemoryCertificateKeyPairProvider {
    #chain;
    #privateKey;
    constructor(chain, key) {
        this.#chain = chain ?? null;
        this.#privateKey = key ?? null;
    }
    get certificateFile() {
        return "<in-memory>";
    }
    get privateKeyFile() {
        return "<in-memory>";
    }
    getCertificate() {
        if (!this.#chain) {
            throw new Error("InMemoryCertificateKeyPairProvider: no certificate available — call ensureCertificateExists() first");
        }
        return this.#chain[0];
    }
    getCertificateChain() {
        if (!this.#chain) {
            throw new Error("InMemoryCertificateKeyPairProvider: no certificate chain available — call ensureCertificateExists() first");
        }
        return this.#chain;
    }
    getPrivateKey() {
        if (!this.#privateKey) {
            throw new Error("InMemoryCertificateKeyPairProvider: no private key available — call ensureCertificateExists() first");
        }
        return this.#privateKey;
    }
    /**
     * Clears cached secrets.
     */
    invalidate() {
        /* no-op — in-memory certs don't need reload from disk */
    }
    /**
     * Auto-provision a self-signed certificate + private key
     * entirely in-memory using `node-opcua-crypto/web`.
     *
     * If a cert+key pair was already provided at construction,
     * this is a no-op.
     *
     * No disk, no openssl, no `node:fs`.
     */
    async ensureCertificateExists(params) {
        if (this.#chain && this.#privateKey) {
            return;
        }
        const cryptoKey = await (0, web_1.generatePrivateKey)(2048);
        const { cert } = await (0, web_1.createSelfSignedCertificate)({
            privateKey: cryptoKey,
            subject: params.subject,
            applicationUri: params.applicationUri,
            dns: params.dns,
            ip: params.ip,
            validity: params.validity ?? 365 * 10,
            notBefore: new Date(),
            purpose: web_1.CertificatePurpose.ForApplication
        });
        this.#chain = [(0, web_1.convertPEMtoDER)(cert)];
        const { privPem } = await (0, web_1.privateKeyToPEM)(cryptoKey);
        this.#privateKey = (0, web_1.makePrivateKeyFromPem)(privPem);
    }
    // Prevent secrets from leaking through JSON serialization
    toJSON() {
        return {
            provider: "InMemoryCertificateKeyPairProvider",
            certificateFile: "<in-memory>",
            privateKeyFile: "<in-memory>"
        };
    }
    // Prevent secrets from leaking through console.log / util.inspect
    [Symbol.for("nodejs.util.inspect.custom")]() {
        return `InMemoryCertificateKeyPairProvider { <in-memory> }`;
    }
}
exports.InMemoryCertificateKeyPairProvider = InMemoryCertificateKeyPairProvider;
