// node-opcua-common/dist/in_memory_certificate_store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryCertificateStore = void 0;
/**
 * In-memory certificate trust store.
 *
 * Manages trust/rejection using `Set<thumbprint>` — no disk,
 * no file watchers, no PKI folder structure.
 *
 * @module node-opcua-common
 */
const web_1 = require("node-opcua-crypto/web");
const node_opcua_status_code_1 = require("node-opcua-status-code");
function thumbprint(certificate) {
    const cert = Array.isArray(certificate) ? certificate[0] : certificate;
    return (0, web_1.makeSHA1Thumbprint)(cert).toString("hex");
}
class InMemoryCertificateStore {
    #trusted = new Set();
    #rejected = new Set();
    #issuers = new Set();
    #crls = new Set();
    #autoAccept;
    referenceCounter = 0;
    constructor(options) {
        this.#autoAccept = options?.autoAcceptUnknown ?? true;
    }
    async initialize() {
        /* no-op */
    }
    async dispose() {
        if (this.referenceCounter > 0) {
            this.referenceCounter--;
            if (this.referenceCounter > 0) {
                return;
            }
        }
        this.#trusted.clear();
        this.#rejected.clear();
        this.#issuers.clear();
        this.#crls.clear();
    }
    async checkCertificate(certificate) {
        const tp = thumbprint(certificate);
        if (this.#trusted.has(tp)) {
            return node_opcua_status_code_1.StatusCodes.Good;
        }
        if (this.#rejected.has(tp)) {
            return node_opcua_status_code_1.StatusCodes.BadCertificateUntrusted;
        }
        // Unknown certificate
        if (this.#autoAccept) {
            this.#trusted.add(tp);
            return node_opcua_status_code_1.StatusCodes.Good;
        }
        this.#rejected.add(tp);
        return node_opcua_status_code_1.StatusCodes.BadCertificateUntrusted;
    }
    async verifyCertificate(certificate, _options) {
        const tp = thumbprint(certificate);
        if (this.#trusted.has(tp)) {
            return "Good";
        }
        if (this.#rejected.has(tp)) {
            return "BadCertificateUntrusted";
        }
        // Unknown certificate
        if (this.#autoAccept) {
            this.#trusted.add(tp);
            return "Good";
        }
        return "BadCertificateUntrusted";
    }
    async trustCertificate(certificate) {
        const tp = thumbprint(certificate);
        this.#rejected.delete(tp);
        this.#trusted.add(tp);
    }
    async rejectCertificate(certificate) {
        const tp = thumbprint(certificate);
        this.#trusted.delete(tp);
        this.#rejected.add(tp);
    }
    async getTrustStatus(certificate) {
        const tp = thumbprint(certificate);
        return this.#trusted.has(tp)
            ? node_opcua_status_code_1.StatusCodes.Good
            : node_opcua_status_code_1.StatusCodes.BadCertificateUntrusted;
    }
    async addIssuer(certificate, _validate, _addInTrustList) {
        this.#issuers.add(thumbprint(certificate));
    }
    async addRevocationList(crl, _target) {
        this.#crls.add(thumbprint(crl));
    }
}
exports.InMemoryCertificateStore = InMemoryCertificateStore;
