// node-opcua-common/dist/index.browser.js
"use strict";
/**
 * @module node-opcua-common/browser
 *
 * Browser-safe subset of `node-opcua-common`. Selected automatically by
 * bundlers (esbuild, webpack, vite, rollup) via the `"browser"` condition
 * in this package's `exports` map.
 *
 * Omits modules whose top-level imports drag Node-only built-ins into the
 * bundle:
 *   - `applicationurn`                    ŌĆö uses `crypto.createHash`
 *   - `disk_certificate_key_pair_provider` ŌĆö uses `node:fs`
 *   - `opcua_secure_object`               ŌĆö uses `node:events` and re-exports
 *                                            `DiskCertificateKeyPairProvider`
 *
 * Browser consumers that need a stable identity URN should build it
 * themselves (e.g. from a stored UUID) ŌĆö the OPC UA spec does not require
 * the SHA-1 host-derived form.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeZoneDataType = exports.SubscriptionDiagnosticsDataType = exports.SessionSecurityDiagnosticsDataType = exports.SessionDiagnosticsDataType = exports.ServiceCounterDataType = exports.ServerStatusDataType = exports.ServerState = exports.ServerDiagnosticsSummaryDataType = exports.SemanticChangeStructureDataType = exports.SamplingIntervalDiagnosticsDataType = exports.RedundantServerDataType = exports.ModelChangeStructureDataType = exports.EnumValueType = exports.DataTypeDefinition = exports.BuildInfo = void 0;
// Pure interfaces ŌĆö compile away to nothing at runtime.
__exportStar(require("./i_certificate_store"), exports);
// Runtime class `StaticCertificateChainProvider` + interfaces; only depends
// on `node-opcua-crypto/web` and the type-only re-export from
// `./opcua_secure_object` (erased at compile time).
__exportStar(require("./certificate_chain_provider"), exports);
// Runtime ŌĆö uses only `node-opcua-crypto/web`. The
// `import type { ICertificateKeyPairProviderWithLocation } from "./opcua_secure_object"`
// is `import type` and erased; no runtime require is emitted.
__exportStar(require("./in_memory_certificate_key_pair_provider"), exports);
// Runtime ŌĆö uses only `node-opcua-crypto/web` and `node-opcua-status-code`.
__exportStar(require("./in_memory_certificate_store"), exports);
// Pure string utility.
__exportStar(require("./make_subject"), exports);
// Plain type re-exports from `node-opcua-types`. These are interfaces /
// generated classes that do not pull Node-only modules.
var node_opcua_types_1 = require("node-opcua-types");
Object.defineProperty(exports, "BuildInfo", { enumerable: true, get: function () { return node_opcua_types_1.BuildInfo; } });
Object.defineProperty(exports, "DataTypeDefinition", { enumerable: true, get: function () { return node_opcua_types_1.DataTypeDefinition; } });
Object.defineProperty(exports, "EnumValueType", { enumerable: true, get: function () { return node_opcua_types_1.EnumValueType; } });
Object.defineProperty(exports, "ModelChangeStructureDataType", { enumerable: true, get: function () { return node_opcua_types_1.ModelChangeStructureDataType; } });
Object.defineProperty(exports, "RedundantServerDataType", { enumerable: true, get: function () { return node_opcua_types_1.RedundantServerDataType; } });
Object.defineProperty(exports, "SamplingIntervalDiagnosticsDataType", { enumerable: true, get: function () { return node_opcua_types_1.SamplingIntervalDiagnosticsDataType; } });
Object.defineProperty(exports, "SemanticChangeStructureDataType", { enumerable: true, get: function () { return node_opcua_types_1.SemanticChangeStructureDataType; } });
Object.defineProperty(exports, "ServerDiagnosticsSummaryDataType", { enumerable: true, get: function () { return node_opcua_types_1.ServerDiagnosticsSummaryDataType; } });
Object.defineProperty(exports, "ServerState", { enumerable: true, get: function () { return node_opcua_types_1.ServerState; } });
Object.defineProperty(exports, "ServerStatusDataType", { enumerable: true, get: function () { return node_opcua_types_1.ServerStatusDataType; } });
Object.defineProperty(exports, "ServiceCounterDataType", { enumerable: true, get: function () { return node_opcua_types_1.ServiceCounterDataType; } });
Object.defineProperty(exports, "SessionDiagnosticsDataType", { enumerable: true, get: function () { return node_opcua_types_1.SessionDiagnosticsDataType; } });
Object.defineProperty(exports, "SessionSecurityDiagnosticsDataType", { enumerable: true, get: function () { return node_opcua_types_1.SessionSecurityDiagnosticsDataType; } });
Object.defineProperty(exports, "SubscriptionDiagnosticsDataType", { enumerable: true, get: function () { return node_opcua_types_1.SubscriptionDiagnosticsDataType; } });
Object.defineProperty(exports, "TimeZoneDataType", { enumerable: true, get: function () { return node_opcua_types_1.TimeZoneDataType; } });
