// node-opcua-common/dist/opcua_secure_object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OPCUASecureObject = void 0;
exports.getPartialCertificateChain = getPartialCertificateChain;
/**
 * @module node-opcua-common
 */
const node_events_1 = require("node:events");
const node_opcua_assert_1 = require("node-opcua-assert");
const web_1 = require("node-opcua-crypto/web");
const disk_certificate_key_pair_provider_1 = require("./disk_certificate_key_pair_provider");
/**
 * Wrap a bare {@link ICertificateKeyPairProvider} (without location
 * properties) into an {@link ICertificateKeyPairProviderWithLocation}.
 *
 * If the provider already implements the extended interface, it is
 * returned as-is. Otherwise a thin wrapper adds `"<unknown>"` defaults.
 */
function ensureProviderHasLocation(provider) {
    if ("certificateFile" in provider && "privateKeyFile" in provider) {
        return provider;
    }
    return {
        get certificateFile() {
            return "<unknown>";
        },
        get privateKeyFile() {
            return "<unknown>";
        },
        getCertificate: () => provider.getCertificate(),
        getCertificateChain: () => provider.getCertificateChain(),
        getPrivateKey: () => provider.getPrivateKey(),
        invalidate: () => {
            if ("invalidate" in provider && typeof provider.invalidate === "function") {
                provider.invalidate();
            }
        }
    };
}
/**
 * Extract a partial certificate chain from a certificate chain so that the
 * total size of the chain does not exceed maxSize.
 * If maxSize is not provided, the full certificate chain is returned.
 * If the first certificate in the chain already exceeds maxSize, an error is thrown.
 *
 * @param certificateChain - full certificate chain (single DER buffer or array)
 * @param maxSize          - optional byte budget
 * @returns the truncated chain as an array of individual certificates
 */
function getPartialCertificateChain(certificateChain, maxSize) {
    if (!certificateChain ||
        (Array.isArray(certificateChain) && certificateChain.length === 0) ||
        (certificateChain instanceof Buffer && certificateChain.length === 0)) {
        return [];
    }
    const certificates = Array.isArray(certificateChain) ? certificateChain : (0, web_1.split_der)(certificateChain);
    if (maxSize === undefined) {
        return certificates;
    }
    // at least include first certificate
    const chainToReturn = [certificates[0]];
    let cumulatedLength = certificates[0].length;
    // Throw if first certificate already exceed maxSize
    if (cumulatedLength > maxSize) {
        throw new Error(`getPartialCertificateChain not enough space for leaf certificate ${maxSize} < ${cumulatedLength}`);
    }
    let index = 1;
    while (index < certificates.length && cumulatedLength + certificates[index].length <= maxSize) {
        chainToReturn.push(certificates[index]);
        cumulatedLength += certificates[index].length;
        index++;
    }
    return chainToReturn;
}
/**
 * An object that provides a certificate and a privateKey.
 *
 * All certificate/key access is delegated to an internal
 * {@link ICertificateKeyPairProviderWithLocation} provider.
 *
 * - When constructed with file paths, a {@link DiskCertificateKeyPairProvider}
 *   is created automatically.
 * - When constructed with an injected provider, it is used directly.
 * - The provider can be replaced at runtime via {@link setProvider}.
 */
// biome-ignore lint/suspicious/noExplicitAny: EventEmitter use any
class OPCUASecureObject extends node_events_1.EventEmitter {
    #provider;
    constructor(options) {
        super();
        if (options.certificateKeyPairProvider) {
            // Injected provider — wrap if missing location properties
            this.#provider = ensureProviderHasLocation(options.certificateKeyPairProvider);
        }
        else {
            // Auto-create disk provider from file paths
            (0, node_opcua_assert_1.assert)(typeof options.certificateFile === "string", "certificateFile or certificateKeyPairProvider is required");
            (0, node_opcua_assert_1.assert)(typeof options.privateKeyFile === "string", "privateKeyFile or certificateKeyPairProvider is required");
            this.#provider = new disk_certificate_key_pair_provider_1.DiskCertificateKeyPairProvider(options.certificateFile || "invalid certificate file", options.privateKeyFile || "invalid private key file");
        }
    }
    /** File path of the certificate (or `"<in-memory>"`). */
    get certificateFile() {
        return this.#provider.certificateFile;
    }
    /** File path of the private key (or `"<in-memory>"`). */
    get privateKeyFile() {
        return this.#provider.privateKeyFile;
    }
    getCertificate() {
        return this.#provider.getCertificate();
    }
    getCertificateChain() {
        return this.#provider.getCertificateChain();
    }
    getPrivateKey() {
        return this.#provider.getPrivateKey();
    }
    /**
     * Replace the internal provider.
     * Accepts any {@link ICertificateKeyPairProvider} — wraps it
     * with location defaults if it lacks `certificateFile`/`privateKeyFile`.
     */
    setProvider(provider) {
        this.#provider = ensureProviderHasLocation(provider);
    }
    /**
     * Invalidate cached certificate chain and private key so the next
     * `getCertificate()` / `getPrivateKey()` call re-reads from the
     * underlying source. For in-memory providers, this is a no-op.
     */
    invalidateCachedCertificates() {
        this.#provider.invalidate?.();
    }
}
exports.OPCUASecureObject = OPCUASecureObject;
