// node-opcua-address-space-base/dist/proofOfConcept.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UAVariableImpl = exports.UAObjectImpl = exports.BaseNodeImpl = exports.TypedEventEmitter = void 0;
const node_events_1 = require("node:events");
const node_opcua_data_model_1 = require("node-opcua-data-model");
class TypedEventEmitter {
    emitter = new node_events_1.EventEmitter();
    on(event, listener) {
        this.emitter.on(event, listener);
        return this;
    }
    once(event, listener) {
        this.emitter.once(event, listener);
        return this;
    }
    emit(event, ...args) {
        this.emitter.emit(event, ...args);
        return this;
    }
    off(event, listener) {
        this.emitter.off(event, listener);
        return this;
    }
}
exports.TypedEventEmitter = TypedEventEmitter;
class BaseNodeImpl extends TypedEventEmitter {
    get addressSpace() {
        return this._addressSpace;
    }
    _addressSpace = { str: "example" };
}
exports.BaseNodeImpl = BaseNodeImpl;
class UAObjectImpl extends BaseNodeImpl {
    nodeClass = node_opcua_data_model_1.NodeClass.Object;
    someUAObjectMethod() {
        // ...
    }
}
exports.UAObjectImpl = UAObjectImpl;
class UAVariableImpl extends BaseNodeImpl {
    nodeClass = node_opcua_data_model_1.NodeClass.Variable;
    someUAVariableMethod() {
        // ...
        this.emit("trucmuch", { value: 42 });
    }
}
exports.UAVariableImpl = UAVariableImpl;
const a = new UAVariableImpl();
a.on("trucmuch", (dataValue) => {
    console.log("trucmuch changed", dataValue.value);
});
a.emit("trucmuch", { value: 42 });
function onNodeFunc(node) {
    node.on("event", () => { });
}
function main() {
    const uaObject = new UAObjectImpl();
    uaObject.on("RolePermissions_changed", (dataValue) => {
        console.log("RolePermissions changed", dataValue.value);
    });
    const uaVariable = new UAVariableImpl();
    uaVariable.on("trucmuch", (dataValue, _indexRange) => {
        console.log("trucmuch changed", dataValue.value);
    });
    onNodeFunc(uaObject);
    onNodeFunc(uaVariable);
}
main();
