// node-opcua-object-registry/dist/objectRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectRegistry = void 0;
/**
 * @module node-opcua-object-registry
 */
const node_opcua_debug_1 = require("node-opcua-debug");
const gRegistries = [];
let hashCounter = 1;
class ObjectRegistry {
    static doDebug = false;
    static registries = gRegistries;
    _objectTypeName;
    _cache;
    constructor() {
        this._cache = {};
        gRegistries.push(this);
    }
    getClassName() {
        return this._objectTypeName ? this._objectTypeName : "<???>";
    }
    register(obj) {
        if (!this._objectTypeName) {
            this._objectTypeName = obj.constructor.name;
        }
        if (!obj._____hash) {
            obj._____hash = hashCounter;
            hashCounter += 1;
            this._cache[obj._____hash] = typeof WeakRef !== "undefined" ? new WeakRef(obj) : obj;
        }
        // c8 ignore next
        if (ObjectRegistry.doDebug) {
            // we capture the stack without processing it immediately to avoid massive CPU overhead at runtime
            obj._____trace = new Error("Trace captured at registration");
        }
    }
    unregister(obj) {
        if (obj._____hash === undefined)
            return;
        delete this._cache[obj._____hash];
        obj._____trace = undefined;
    }
    *_yieldAliveObjects() {
        for (const [key, cachedObject] of Object.entries(this._cache)) {
            const obj = cachedObject instanceof WeakRef ? cachedObject.deref() : cachedObject;
            if (obj) {
                yield [key, obj];
            }
            else {
                delete this._cache[key]; // Autoclean dead weak refs!
            }
        }
    }
    count() {
        let cnt = 0;
        for (const _ of this._yieldAliveObjects()) {
            cnt++;
        }
        return cnt;
    }
    toString() {
        const className = this.getClassName();
        let str = ` className :${className} found => ${this.count()} object leaking\n`;
        for (const [, obj] of this._yieldAliveObjects()) {
            str += `${obj.constructor.name} ${obj.toString ? obj.toString() : ""}\n`;
        }
        if (ObjectRegistry.doDebug) {
            for (const [key, obj] of this._yieldAliveObjects()) {
                const trace = obj._____trace ? (0, node_opcua_debug_1.traceFromThisProjectOnly)(obj._____trace) : "<no trace>";
                str += `   ${key}${trace}\n`;
            }
        }
        return str;
    }
}
exports.ObjectRegistry = ObjectRegistry;
ObjectRegistry.doDebug = typeof process === "object" ? !!process?.env?.NODEOPCUA_REGISTRY?.match(/DEBUG/) : false;
if (ObjectRegistry.doDebug) {
    console.log("ObjectRegistry.doDebug = ", ObjectRegistry.doDebug);
}
