// node-opcua-hostname/dist/hostname.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractFullyQualifiedDomainName = extractFullyQualifiedDomainName;
exports.prepareFQDN = prepareFQDN;
exports.getFullyQualifiedDomainName = getFullyQualifiedDomainName;
exports.getHostname = getHostname;
exports.resolveFullyQualifiedDomainName = resolveFullyQualifiedDomainName;
exports.getIpAddresses = getIpAddresses;
exports.isIPAddress = isIPAddress;
exports.ipv4ToHex = ipv4ToHex;
/**
 * @module node-opcua-hostname
 */
const node_dns_1 = __importDefault(require("node:dns"));
const node_net_1 = __importDefault(require("node:net"));
const node_os_1 = __importDefault(require("node:os"));
const node_util_1 = require("node:util");
function trim(str, length) {
    if (!length) {
        return str;
    }
    return str.substring(0, Math.min(str.length, length));
}
function fqdn(callback) {
    const uqdn = node_os_1.default.hostname();
    node_dns_1.default.lookup(uqdn, { hints: node_dns_1.default.ADDRCONFIG }, (err1, ip) => {
        if (err1) {
            return callback(err1);
        }
        node_dns_1.default.lookupService(ip, 0, (err2, _fqdn) => {
            if (err2) {
                return callback(err2);
            }
            _fqdn = _fqdn.replace(".localdomain", "");
            callback(null, _fqdn);
        });
    });
}
let _fullyQualifiedDomainNameInCache;
/**
 * extract FullyQualifiedDomainName of this computer
 */
async function extractFullyQualifiedDomainName() {
    if (_fullyQualifiedDomainNameInCache) {
        return _fullyQualifiedDomainNameInCache;
    }
    if (process.platform === "win32") {
        // http://serverfault.com/a/73643/251863
        const env = process.env;
        _fullyQualifiedDomainNameInCache =
            env.COMPUTERNAME + (env.USERDNSDOMAIN && env.USERDNSDOMAIN.length > 0 ? `.${env.USERDNSDOMAIN}` : "");
    }
    else {
        try {
            _fullyQualifiedDomainNameInCache = await (0, node_util_1.promisify)(fqdn)();
            if (_fullyQualifiedDomainNameInCache === "localhost") {
                throw new Error("localhost not expected");
            }
            if (/sethostname/.test(_fullyQualifiedDomainNameInCache)) {
                throw new Error("Detecting fqdn  on windows !!!");
            }
        }
        catch (_err) {
            // fall back to old method
            _fullyQualifiedDomainNameInCache = node_os_1.default.hostname();
        }
    }
    return _fullyQualifiedDomainNameInCache || "";
}
async function prepareFQDN() {
    _fullyQualifiedDomainNameInCache = await extractFullyQualifiedDomainName();
}
function getFullyQualifiedDomainName(optional_max_length) {
    if (!_fullyQualifiedDomainNameInCache) {
        throw new Error("FullyQualifiedDomainName computation is not completed yet");
    }
    return _fullyQualifiedDomainNameInCache ? trim(_fullyQualifiedDomainNameInCache, optional_max_length) : "%FQDN%";
}
function getHostname() {
    return node_os_1.default.hostname();
}
function resolveFullyQualifiedDomainName(str) {
    if (!_fullyQualifiedDomainNameInCache) {
        throw new Error("FullyQualifiedDomainName computation is not completed yet");
    }
    str = str.replace("%FQDN%", _fullyQualifiedDomainNameInCache);
    str = str.replace("{FQDN}", _fullyQualifiedDomainNameInCache);
    str = str.replace("{hostname}", getHostname());
    return str;
}
/**
 * Return all non-internal IPv4 addresses detected on this host.
 *
 * The result is sorted for deterministic output.
 */
function getIpAddresses() {
    const ipAddresses = [];
    const netInterfaces = node_os_1.default.networkInterfaces();
    for (const interfaceName of Object.keys(netInterfaces)) {
        const ifaces = netInterfaces[interfaceName];
        if (!ifaces) {
            continue;
        }
        for (const iface of ifaces) {
            if (iface.family === "IPv4" && !iface.internal) {
                ipAddresses.push(iface.address);
            }
        }
    }
    return ipAddresses.sort();
}
/**
 * Check whether a string is an IP address (v4 or v6).
 *
 * Returns `4` for IPv4, `6` for IPv6, or `0` if the string
 * is a hostname (not an IP literal).
 *
 * Useful for segregating `alternateHostname` and
 * `advertisedEndpoints` values into SAN `dNSName` vs
 * `iPAddress` entries.
 */
function isIPAddress(value) {
    return node_net_1.default.isIP(value);
}
/**
 * Convert a dotted-decimal IPv4 address to a hex string.
 *
 * This matches the format returned by `exploreCertificate`'s
 * `iPAddress` SAN entries (e.g. `"192.168.1.69"` → `"c0a80145"`).
 */
function ipv4ToHex(ip) {
    return ip
        .split(".")
        .map((octet) => Number.parseInt(octet, 10).toString(16).padStart(2, "0"))
        .join("");
}
// note : under windows ... echo %COMPUTERNAME%.%USERDNSDOMAIN%
prepareFQDN();
