// @ster5/global-mutex/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// source/index.ts
var source_exports = {};
__export(source_exports, {
  defaultStaleDuration: () => defaultStaleDuration2,
  drainPendingLocks: () => drainPendingLocks,
  isLocked: () => isLocked,
  setLockProvider: () => setLockProvider,
  withLock: () => withLock
});
module.exports = __toCommonJS(source_exports);
var import_node_fs2 = __toESM(require("fs"));
var import_node_path = __toESM(require("path"));

// source/native-provider.ts
var import_node_fs = __toESM(require("fs"));
var defaultStaleDuration = 2 * 60 * 1e3;
function normalizeRetry(retries) {
  if (retries === void 0) {
    return {
      maxAttempts: Number.POSITIVE_INFINITY,
      minTimeout: 100,
      maxTimeout: 2e3,
      randomize: true,
      maxRetryTime: Number.POSITIVE_INFINITY
    };
  }
  if (typeof retries === "number") {
    return {
      maxAttempts: retries,
      minTimeout: 1e3,
      maxTimeout: Number.POSITIVE_INFINITY,
      randomize: false,
      maxRetryTime: Number.POSITIVE_INFINITY
    };
  }
  const maxAttempts = retries.forever ? Number.POSITIVE_INFINITY : retries.retries ?? 10;
  return {
    maxAttempts,
    minTimeout: retries.minTimeout ?? 1e3,
    maxTimeout: retries.maxTimeout ?? Number.POSITIVE_INFINITY,
    randomize: retries.randomize ?? false,
    maxRetryTime: retries.maxRetryTime ?? Number.POSITIVE_INFINITY
  };
}
function retryDelay(r, attempt) {
  let d = Math.min(r.minTimeout * 2 ** attempt, r.maxTimeout);
  if (r.randomize) {
    d = Math.floor(d * (1 + Math.random()) / 2);
  }
  return d;
}
function lockPath(fileToLock) {
  return `${fileToLock}.lock`;
}
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
async function isStale(lp, staleDuration) {
  try {
    const stat = await import_node_fs.default.promises.stat(lp);
    return Date.now() - stat.mtimeMs > staleDuration;
  } catch {
    return true;
  }
}
var NativeLockProvider = class {
  activeLocks = /* @__PURE__ */ new Map();
  nextLockId = 0;
  exitHandlerRegistered = false;
  cleanupLocksOnExit = () => {
    for (const [, entry] of this.activeLocks) {
      if (!entry.released) {
        entry.released = true;
        clearInterval(entry.timer);
        try {
          import_node_fs.default.rmSync(entry.lockPath, { recursive: true, force: true });
        } catch {
        }
      }
    }
    this.activeLocks.clear();
  };
  ensureExitHandler() {
    if (this.exitHandlerRegistered) return;
    this.exitHandlerRegistered = true;
    process.on("exit", this.cleanupLocksOnExit);
  }
  async acquireLock(lp, stale, retry, update, onCompromised) {
    const startTime = Date.now();
    let attempt = 0;
    for (; ; ) {
      try {
        await import_node_fs.default.promises.mkdir(lp);
        this.ensureExitHandler();
        const lockId = this.nextLockId++;
        const timer = setInterval(() => {
          const entry = this.activeLocks.get(lockId);
          if (!entry || entry.released) return;
          try {
            const now = /* @__PURE__ */ new Date();
            import_node_fs.default.utimesSync(lp, now, now);
          } catch (err) {
            if (onCompromised) {
              onCompromised(err);
            }
          }
        }, update);
        timer.unref();
        this.activeLocks.set(lockId, { lockPath: lp, timer, released: false });
        return lockId;
      } catch (err) {
        const code = err.code;
        let isSwallowedEperm = false;
        if (code === "EEXIST") {
        } else if (code === "EPERM") {
          try {
            import_node_fs.default.statSync(lp);
          } catch {
            isSwallowedEperm = true;
          }
        } else {
          throw err;
        }
        if (!isSwallowedEperm) {
          if (await isStale(lp, stale)) {
            try {
              await import_node_fs.default.promises.rm(lp, { recursive: true, force: true });
            } catch {
            }
            continue;
          }
        }
        if (attempt >= retry.maxAttempts) {
          throw new Error("Lock file is already being held");
        }
        if (Date.now() - startTime >= retry.maxRetryTime) {
          throw new Error("Lock file is already being held");
        }
        await sleep(retryDelay(retry, attempt));
        attempt++;
      }
    }
  }
  releaseLock(lockId) {
    const entry = this.activeLocks.get(lockId);
    if (entry) {
      entry.released = true;
      clearInterval(entry.timer);
      this.activeLocks.delete(lockId);
      import_node_fs.default.rmSync(entry.lockPath, { recursive: true, force: true });
    }
  }
  async acquire(options) {
    const { fileToLock, stale = defaultStaleDuration, update, retries, onCompromised } = options;
    const retry = normalizeRetry(retries);
    const updateInterval = update ?? Math.floor(stale / 2);
    const lp = lockPath(fileToLock);
    const lockId = await this.acquireLock(lp, stale, retry, updateInterval, onCompromised);
    return async () => {
      this.releaseLock(lockId);
    };
  }
  async isLocked(fileToLock) {
    const lp = lockPath(fileToLock);
    try {
      await import_node_fs.default.promises.access(lp);
      return !await isStale(lp, defaultStaleDuration);
    } catch {
      return false;
    }
  }
  async drainPendingLocks() {
  }
};

// source/proper-lockfile-provider.ts
var ProperLockfileProvider = class {
  // ---------------------------------------------------------------------------
  // Public API
  // ---------------------------------------------------------------------------
  async acquire(options) {
    const { lock: properLock } = await import("proper-lockfile");
    const { fileToLock, stale, retries, onCompromised } = options;
    let properRetries;
    if (retries === void 0) {
      properRetries = {
        retries: 1e4,
        // effectively forever (approx 5.5 hours at 2s max timeout)
        factor: 2,
        minTimeout: 100,
        maxTimeout: 2e3,
        randomize: true
      };
    } else if (typeof retries === "number") {
      properRetries = { retries, factor: 1, minTimeout: 1e3 };
    } else if (retries && !retries.forever) {
      properRetries = {
        retries: retries.retries ?? 10,
        factor: 2,
        minTimeout: retries.minTimeout,
        maxTimeout: retries.maxTimeout,
        randomize: retries.randomize
      };
    } else if (retries?.forever) {
      properRetries = {
        retries: 1e4,
        // effectively forever (approx 5.5 hours at 2s max timeout)
        factor: 2,
        minTimeout: retries.minTimeout ?? 100,
        maxTimeout: retries.maxTimeout ?? 2e3,
        randomize: retries.randomize ?? true
      };
    }
    const properOptions = {
      stale: stale ?? 2 * 60 * 1e3,
      // default 2 mins
      retries: properRetries,
      onCompromised: onCompromised || (() => {
      })
    };
    const release = await properLock(fileToLock, properOptions);
    return async () => {
      await release();
    };
  }
  async isLocked(fileToLock) {
    const { check: properCheck } = await import("proper-lockfile");
    return properCheck(fileToLock, { stale: 2 * 60 * 1e3 });
  }
  async drainPendingLocks() {
  }
};

// source/index.ts
var defaultStaleDuration2 = 2 * 60 * 1e3;
var overriddenProviderType;
var _resolvedProvider;
function setLockProvider(type) {
  overriddenProviderType = type;
  _resolvedProvider = void 0;
}
async function getProvider() {
  if (_resolvedProvider) return _resolvedProvider;
  let targetType = overriddenProviderType;
  if (!targetType) {
    const envType = process.env.GLOBAL_MUTEX_PROVIDER;
    if (envType === "native" || envType === "proper-lockfile") {
      targetType = envType;
      console.log(`[global-mutex] Forcing lock provider to '${envType}' as specified by GLOBAL_MUTEX_PROVIDER env variable.`);
    }
  }
  if (!targetType) {
    targetType = "proper-lockfile";
  }
  if (targetType === "proper-lockfile") {
    try {
      await import("proper-lockfile");
      _resolvedProvider = new ProperLockfileProvider();
      return _resolvedProvider;
    } catch (_e) {
      if (overriddenProviderType === "proper-lockfile" || process.env.GLOBAL_MUTEX_PROVIDER === "proper-lockfile") {
        console.warn(
          "[global-mutex] proper-lockfile was explicitly requested but is not installed. Falling back to native provider."
        );
      }
      _resolvedProvider = new NativeLockProvider();
      return _resolvedProvider;
    }
  }
  _resolvedProvider = new NativeLockProvider();
  return _resolvedProvider;
}
var pendingLocks = /* @__PURE__ */ new Set();
async function withLock(options, action) {
  const { fileToLock } = options;
  const lockPromise = (async () => {
    try {
      await import_node_fs2.default.promises.access(import_node_path.default.dirname(fileToLock));
    } catch {
      throw new Error(`Invalid lockfile: ${fileToLock}`);
    }
    try {
      await import_node_fs2.default.promises.writeFile(fileToLock, "", { flag: "wx" });
    } catch (e) {
      if (e.code !== "EEXIST") throw e;
    }
    const provider = await getProvider();
    const release = await provider.acquire(options);
    try {
      return await action();
    } finally {
      try {
        await release();
      } catch (err) {
        console.warn("Error in Unlock !!!", err.message);
      }
    }
  })();
  pendingLocks.add(lockPromise);
  const cleanup = () => pendingLocks.delete(lockPromise);
  lockPromise.then(cleanup, cleanup);
  return lockPromise;
}
async function isLocked(fileToLock) {
  const provider = await getProvider();
  return provider.isLocked(fileToLock);
}
async function drainPendingLocks() {
  if (_resolvedProvider) {
    await _resolvedProvider.drainPendingLocks();
  }
  await Promise.allSettled([...pendingLocks]);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  defaultStaleDuration,
  drainPendingLocks,
  isLocked,
  setLockProvider,
  withLock
});
