// node-opcua-client/dist/private/reconnection/reconnection.js
"use strict";
/**
 * @module node-opcua-client-private
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._shouldNotContinue = _shouldNotContinue;
exports._shouldNotContinue2 = _shouldNotContinue2;
exports.repair_client_session = repair_client_session;
exports.repair_client_sessions = repair_client_sessions;
// tslint:disable:only-arrow-functions
const chalk_1 = __importDefault(require("chalk"));
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_client_dynamic_extension_object_1 = require("node-opcua-client-dynamic-extension-object");
const node_opcua_debug_1 = require("node-opcua-debug");
const node_opcua_service_subscription_1 = require("node-opcua-service-subscription");
const node_opcua_types_1 = require("node-opcua-types");
const client_publish_engine_reconnection_1 = require("./client_publish_engine_reconnection");
const client_subscription_reconnection_1 = require("./client_subscription_reconnection");
const debugLog = (0, node_opcua_debug_1.make_debugLog)("RECONNECTION");
const doDebug = (0, node_opcua_debug_1.checkDebugFlag)("RECONNECTION");
const errorLog = (0, node_opcua_debug_1.make_errorLog)("RECONNECTION");
const warningLog = (0, node_opcua_debug_1.make_warningLog)("RECONNECTION");
function _shouldNotContinue3(client) {
    if (!client._secureChannel) {
        return new Error("Failure during reconnection : client or session is not usable anymore");
    }
    return null;
}
function _shouldNotContinue(session) {
    if (!session._client || session.hasBeenClosed() || !session._client._secureChannel || session._client.isUnusable()) {
        return new Error("Failure during reconnection : client or session is not usable anymore");
    }
    return null;
}
function _shouldNotContinue2(subscription) {
    if (!subscription.hasSession) {
        return new Error("Failure during reconnection : client or session is not usable anymore");
    }
    return _shouldNotContinue(subscription.session);
}
function _throwIfShouldNotContinue(session) {
    const err = _shouldNotContinue(session);
    if (err)
        throw err;
}
//
// a new secure channel has be created, we need to reactivate the corresponding session,
// and reestablish the subscription and restart the publish engine.
//
//
// see OPC UA part 4 ( version 1.03 ) figure 34 page 106
// 6.5 Reestablishing subscription....
//
//
//
//                      +---------------------+
//                      | CreateSecureChannel |
//                      | CreateSession       |
//                      | ActivateSession     |
//                      +---------------------+
//                                |
//                                |
//                                v
//                      +---------------------+
//                      | CreateSubscription  |<-------------------------------------------------------------+
//                      +---------------------+                                                              |
//                                |                                                                         (1)
//                                |
//                                v
//                      +---------------------+
//     (2)------------->| StartPublishEngine  |
//                      +---------------------+
//                                |
//                                V
//                      +---------------------+
//             +------->| Monitor Connection  |
//             |        +---------------------+
//             |                    |
//             |                    v
//             |          Good    /   \
//             +-----------------/ SR? \______Broken_____+
//                               \     /                 |
//                                \   /                  |
//                                                       |
//                                                       v
//                                                 +---------------------+
//                                                 |                     |
//                                                 | CreateSecureChannel |<-----+
//                                                 |                     |      |
//                                                 +---------------------+      |
//                                                         |                    |
//                                                         v                    |
//                                                       /   \                  |
//                                                      / SR? \______Bad________+
//                                                      \     /
//                                                       \   /
//                                                         |
//                                                         |Good
//                                                         v
//                                                 +---------------------+
//                                                 |                     |
//                                                 | ActivateSession     |
//                                                 |                     |
//                                                 +---------------------+
//                                                         |
//                                  +----------------------+
//                                  |
//                                  v                    +-------------------+       +----------------------+
//                                /   \                  | CreateSession     |       |                      |
//                               / SR? \______Bad_____>  | ActivateSession   |----->  | TransferSubscription |
//                               \     /                 |                   |       |                      |       (1)
//                                \   /                  +-------------------+       +----------------------+        ^
//                                  | Good                                                      |                    |
//                                  v   (for each subscription)                                 |                    |
//                          +--------------------+                                            /   \                  |
//                          |                    |                                     OK    / OK? \______Bad________+
//                          | RePublish          |<----------------------------------------- \     /
//                      +-->|                    |                                            \   /
//                      |   +--------------------+
//                      |           |
//                      |           v
//                      | GOOD    /   \
//                      +------  / SR? \______Bad SubscriptionInvalidId______>(1)
// (2)                           \     /
//  ^                             \   /
//  |                               |
//  |                               |
//  |      BadMessageNotAvailable   |
//  +-------------------------------+
function _ask_for_subscription_republish(session, callback) {
    // prettier-ignore
    {
        const err = _shouldNotContinue(session);
        if (err) {
            return callback(err);
        }
    }
    doDebug && debugLog(chalk_1.default.bgCyan.yellow.bold("_ask_for_subscription_republish "));
    // assert(session.getPublishEngine().nbPendingPublishRequests === 0,
    //   "at this time, publish request queue shall still be empty");
    const engine = session.getPublishEngine();
    (0, client_publish_engine_reconnection_1.republish)(engine, (err) => {
        doDebug && debugLog("_ask_for_subscription_republish :  republish sent");
        // prettier-ignore
        {
            const err = _shouldNotContinue(session);
            if (err) {
                return callback(err);
            }
        }
        doDebug && debugLog(chalk_1.default.bgCyan.green.bold("_ask_for_subscription_republish done "), err ? err.message : "OK");
        if (err) {
            warningLog("republish has failed with error :", err.message);
            doDebug && debugLog("_ask_for_subscription_republish has :  recreating subscription");
            return repair_client_session_by_recreating_a_new_session(session._client, session, callback);
        }
        // Republish has completed: resume normal Publish handling. During the connection break the
        // in-flight PublishRequests have failed and the publish-request queue has drained, so we must
        // replenish it now, otherwise the client would stop sending PublishRequests and never receive
        // live notifications again. (https://github.com/node-opcua/node-opcua/issues/1524)
        // Note: on the new-session repair path the engine is still suspended here, so this is a no-op
        // there and the queue is replenished later by suspend(false).
        // Guard on subscriptionCount: with no subscription, send_publish_request() would still emit
        // spurious PublishRequests (it does not check the subscription count itself), adding useless
        // round-trips on every reconnection of a subscription-less session.
        if (engine.subscriptionCount > 0) {
            engine.replenish_publish_request_queue();
        }
        callback(err);
    });
}
function create_session_and_repeat_if_failed(client, session) {
    return new Promise((resolve, reject) => {
        _throwIfShouldNotContinue(session);
        doDebug && debugLog(chalk_1.default.bgWhite.red("    => creating a new session ...."));
        // create new session, based on old session,
        // so we can reuse subscriptions data
        client.__createSession_step2(session, (err, session1) => {
            const abortErr = _shouldNotContinue(session);
            if (abortErr)
                return reject(abortErr);
            if (!err && session1) {
                (0, node_opcua_assert_1.assert)(session === session1, "session should have been recycled");
                resolve(session);
            }
            else {
                doDebug && debugLog("Cannot complete subscription republish err = ", err?.message);
                reject(err || new Error("Failed to create session"));
            }
        });
    });
}
function activate_session(client, session, newSession) {
    return new Promise((resolve, reject) => {
        _throwIfShouldNotContinue(session);
        doDebug && debugLog(chalk_1.default.bgWhite.red("    => activating a new session ...."));
        client._activateSession(newSession, newSession.userIdentityInfo, (err, _session1) => {
            // c8 ignore next
            doDebug && debugLog("    =>  activating a new session .... Done err=", err ? err.message : "null");
            if (err) {
                doDebug &&
                    debugLog("reactivation of the new session has failed: let be smart and close it before failing this repair attempt");
                // but just on the server side, not on the client side
                const closeSessionRequest = new node_opcua_types_1.CloseSessionRequest({
                    requestHeader: {
                        authenticationToken: newSession.authenticationToken
                    },
                    deleteSubscriptions: true
                });
                newSession._client?.performMessageTransaction(closeSessionRequest, (err2) => {
                    if (err2) {
                        warningLog("closing session", err2.message);
                    }
                    // c8 ignore next
                    doDebug && debugLog("the temporary replacement session is now closed");
                    // c8 ignore next
                    doDebug && debugLog(" err ", err.message, "propagated upwards");
                    reject(err);
                });
            }
            else {
                resolve();
            }
        });
    });
}
async function recreateSubscriptions(session, newSession, subscriptionsToRecreate) {
    for (const subscriptionId of subscriptionsToRecreate) {
        _throwIfShouldNotContinue(session);
        if (!session.getPublishEngine().hasSubscription(subscriptionId)) {
            doDebug && debugLog(chalk_1.default.red("          => CANNOT RECREATE SUBSCRIPTION  "), subscriptionId);
            continue;
        }
        const subscription = session.getPublishEngine().getSubscription(subscriptionId);
        doDebug && debugLog(chalk_1.default.red("          => RECREATING SUBSCRIPTION  "), subscriptionId);
        (0, node_opcua_assert_1.assert)(subscription.session === newSession, "must have the new session");
        try {
            await (0, client_subscription_reconnection_1.recreateSubscriptionAndMonitoredItem)(subscription);
            doDebug &&
                debugLog(chalk_1.default.cyan("          => RECREATING SUBSCRIPTION  AND MONITORED ITEM DONE subscriptionId="), subscriptionId);
        }
        catch (err) {
            doDebug && debugLog(`_recreateSubscription failed !${err.message}`);
        }
    }
    _throwIfShouldNotContinue(session);
}
function attempt_subscription_transfer(_client, session, newSession) {
    return new Promise((resolve, reject) => {
        _throwIfShouldNotContinue(session);
        // get the old subscriptions id from the old session
        const subscriptionsIds = session.getPublishEngine().getSubscriptionIds();
        doDebug && debugLog("  session subscriptionCount = ", newSession.getPublishEngine().subscriptionCount);
        if (subscriptionsIds.length === 0) {
            doDebug && debugLog(" No subscriptions => skipping transfer subscriptions");
            return resolve(); // no need to transfer subscriptions
        }
        doDebug && debugLog("    => asking server to transfer subscriptions = [", subscriptionsIds.join(", "), "]");
        // Transfer subscriptions - ask for initial values....
        const subscriptionsToTransfer = new node_opcua_service_subscription_1.TransferSubscriptionsRequest({
            sendInitialValues: true,
            subscriptionIds: subscriptionsIds
        });
        if (newSession.getPublishEngine().nbPendingPublishRequests !== 0) {
            warningLog("Warning : we should not be publishing here");
        }
        newSession.transferSubscriptions(subscriptionsToTransfer, (err, transferSubscriptionsResponse) => {
            // may be the connection with server has been disconnected
            const abortErr = _shouldNotContinue(session);
            if (abortErr)
                return reject(abortErr);
            if (err || !transferSubscriptionsResponse) {
                warningLog(chalk_1.default.bgCyan("May be the server is not supporting this feature"));
                // when transfer subscription has failed, we have no other choice but
                // recreate the subscriptions on the server side
                const subscriptionsToRecreate = [...(subscriptionsToTransfer.subscriptionIds || [])];
                warningLog(chalk_1.default.bgCyan("We need to recreate entirely the subscription"));
                recreateSubscriptions(session, newSession, subscriptionsToRecreate)
                    .then(() => resolve())
                    .catch((e) => reject(e));
                return;
            }
            const results = transferSubscriptionsResponse.results || [];
            // c8 ignore next
            if (doDebug) {
                debugLog(chalk_1.default.cyan("    =>  transfer subscriptions  done"), results.map((x) => x.statusCode.toString()).join(" "));
            }
            const subscriptionsToRecreate = [];
            // some subscriptions may be marked as invalid on the server side ..
            // those one need to be recreated and repaired ....
            for (let i = 0; i < results.length; i++) {
                const statusCode = results[i].statusCode;
                if (statusCode.isNotGood()) {
                    // the subscription could not be transferred (BadSubscriptionIdInvalid when the
                    // subscription no longer exists, BadUserAccessDenied when the server refuses the
                    // transfer per OPC UA Part 4 §5.14.7, ...). Whatever the reason, recreate it here
                    // directly on the new session instead of relying on the subsequent Republish step
                    // to notice and repair it (which would otherwise cost an extra failed round-trip).
                    doDebug &&
                        debugLog(chalk_1.default.red("         WARNING SUBSCRIPTION  "), subscriptionsIds[i], chalk_1.default.red(" SHOULD BE RECREATED ("), statusCode.toString(), chalk_1.default.red(")"));
                    subscriptionsToRecreate.push(subscriptionsIds[i]);
                }
                else {
                    const availableSequenceNumbers = results[i].availableSequenceNumbers;
                    doDebug &&
                        debugLog(chalk_1.default.green("         SUBSCRIPTION "), subscriptionsIds[i], chalk_1.default.green(" CAN BE REPAIRED AND AVAILABLE "), availableSequenceNumbers);
                    // should be Good.
                }
            }
            doDebug && debugLog("  new session subscriptionCount = ", newSession.getPublishEngine().subscriptionCount);
            recreateSubscriptions(session, newSession, subscriptionsToRecreate)
                .then(() => resolve())
                .catch((e) => reject(e));
        });
    });
}
async function repair_client_session_by_recreating_a_new_session_async(client, session) {
    _throwIfShouldNotContinue(session);
    // As we don't know if server has been rebooted or not,
    // and may be upgraded in between, we have to invalidate the extra data type manager
    (0, node_opcua_client_dynamic_extension_object_1.invalidateExtraDataTypeManager)(session);
    // c8 ignore next
    if (doDebug) {
        debugLog(" repairing client session by_recreating a new session for old session ", session.sessionId.toString());
    }
    const listenerCountBefore = session.listenerCount("");
    // Step 1: suspend old session publish engine
    _throwIfShouldNotContinue(session);
    // c8 ignore next
    doDebug && debugLog(chalk_1.default.bgWhite.red("    => suspend old session publish engine...."));
    session.getPublishEngine().suspend(true);
    // Step 2: create new session
    const newSession = await create_session_and_repeat_if_failed(client, session);
    _throwIfShouldNotContinue(session);
    // Step 3: activate new session
    await activate_session(client, session, newSession);
    _throwIfShouldNotContinue(session);
    // Step 4: before subscription repair hook
    if (client.beforeSubscriptionRecreate) {
        const hookErr = await client.beforeSubscriptionRecreate(newSession);
        _throwIfShouldNotContinue(session);
        if (hookErr) {
            throw hookErr;
        }
    }
    // Step 5: attempt subscription transfer
    await attempt_subscription_transfer(client, session, newSession);
    _throwIfShouldNotContinue(session);
    // Step 6: ask for subscription republish
    await new Promise((resolve, reject) => {
        _ask_for_subscription_republish(newSession, (err) => {
            if (err) {
                warningLog("warning: Subscription republished has failed ", err.message);
                reject(err);
            }
            else {
                resolve();
            }
        });
    });
    _throwIfShouldNotContinue(session);
    // Step 7: start publishing as normal
    newSession.getPublishEngine().suspend(false);
    const listenerCountAfter = session.listenerCount("");
    (0, node_opcua_assert_1.assert)(newSession === session);
    doDebug && debugLog("listenerCountBefore =", listenerCountBefore, "listenerCountAfter = ", listenerCountAfter);
}
function repair_client_session_by_recreating_a_new_session(client, session, callback) {
    repair_client_session_by_recreating_a_new_session_async(client, session)
        .then(() => callback())
        .catch((err) => {
        doDebug && debugLog("repair_client_session_by_recreating_a_new_session failed with ", err.message);
        callback(err);
    });
}
function _repair_client_session(client, session, callback) {
    const callback2 = (err2) => {
        doDebug &&
            debugLog("Session repair completed with err: ", err2 ? err2.message : "<no error>", session.sessionId.toString());
        if (!err2) {
            session.emit("session_repaired");
        }
        else {
            session.emit("session_repaired_failed", err2);
        }
        callback(err2);
    };
    if (doDebug) {
        doDebug && debugLog(chalk_1.default.yellow("  TRYING TO REACTIVATE EXISTING SESSION"), session.sessionId.toString());
        doDebug && debugLog("   SubscriptionIds :", session.getPublishEngine().getSubscriptionIds());
    }
    // prettier-ignore
    {
        const err = _shouldNotContinue(session);
        if (err) {
            return callback(err);
        }
    }
    client._activateSession(session, session.userIdentityInfo, (err, _session2) => {
        // prettier-ignore
        {
            const err = _shouldNotContinue(session);
            if (err) {
                return callback(err);
            }
        }
        //
        // Note: current limitation :
        //  - The reconnection doesn't work yet, if connection break is caused by a server that crashes and restarts.
        //
        doDebug && debugLog("   ActivateSession : ", err ? chalk_1.default.red(err.message) : chalk_1.default.green(" SUCCESS !!! "));
        if (err) {
            //  activate old session has failed => let's  recreate a new Channel and transfer the subscription
            return repair_client_session_by_recreating_a_new_session(client, session, callback2);
        }
        else {
            // activate old session has succeeded => let's call Republish
            return _ask_for_subscription_republish(session, callback2);
        }
    });
}
function repair_client_session(client, session, callback) {
    if (!client) {
        doDebug && debugLog("Aborting reactivation of old session because user requested session to be close");
        return callback();
    }
    doDebug && debugLog(chalk_1.default.yellow("Starting client session repair"));
    const privateSession = session;
    privateSession._reconnecting = privateSession._reconnecting || { reconnecting: false, pendingCallbacks: [] };
    if (session.hasBeenClosed()) {
        privateSession._reconnecting.reconnecting = false;
        doDebug && debugLog("Aborting reactivation of old session because session has been closed");
        return callback();
    }
    if (privateSession._reconnecting.reconnecting) {
        doDebug && debugLog(chalk_1.default.bgCyan("Reconnection is already happening for session"), session.sessionId.toString());
        privateSession._reconnecting.pendingCallbacks.push(callback);
        return;
    }
    privateSession._reconnecting.reconnecting = true;
    // get old transaction queue ...
    const transactionQueue = privateSession._reconnecting.pendingTransactions.splice(0);
    const repeatedAction = (callback) => {
        // prettier-ignore
        {
            const err = _shouldNotContinue(session);
            if (err) {
                return callback(err);
            }
        }
        _repair_client_session(client, session, (err) => {
            // prettier-ignore
            {
                const err = _shouldNotContinue(session);
                if (err) {
                    return callback(err);
                }
            }
            if (err) {
                errorLog(chalk_1.default.red("session restoration has failed! err ="), err.message, session.sessionId.toString(), " => Let's retry");
                if (!session.hasBeenClosed()) {
                    const delay = 2000;
                    errorLog(chalk_1.default.red(`... will retry session repair... in ${delay} ms`));
                    setTimeout(() => {
                        {
                            const err = _shouldNotContinue(session);
                            if (err) {
                                warningLog("cancelling session repair");
                                return callback(err);
                            }
                        }
                        errorLog(chalk_1.default.red("Retrying session repair..."));
                        repeatedAction(callback);
                    }, delay);
                    return;
                }
                else {
                    errorLog(chalk_1.default.red("session restoration should be interrupted because session has been closed forcefully"));
                }
                // session does not need to be repaired anymore
                callback();
                return;
            }
            // c8 ignore next
            doDebug && debugLog(chalk_1.default.yellow("session has been restored"), session.sessionId.toString());
            session.emit("session_restored");
            callback(err);
        });
    };
    repeatedAction((err) => {
        privateSession._reconnecting.reconnecting = false;
        const otherCallbacks = privateSession._reconnecting.pendingCallbacks.splice(0);
        // re-inject element in queue
        // c8 ignore next
        if (transactionQueue.length > 0) {
            doDebug && debugLog(chalk_1.default.yellow("re-injecting transaction queue"), transactionQueue.length);
            transactionQueue.forEach((e) => privateSession._reconnecting.pendingTransactions.push(e));
        }
        otherCallbacks.forEach((c) => c(err));
        callback(err);
    });
}
async function repair_client_sessions(client, callback) {
    // repair session
    const sessions = client.getSessions();
    doDebug && debugLog(chalk_1.default.red.bgWhite(" Starting sessions reactivation", sessions.length));
    const allErrors = [];
    for (const session of sessions) {
        await new Promise((resolve) => {
            repair_client_session(client, session, (err) => {
                allErrors.push(err);
                resolve();
            });
        });
    }
    // prettier-ignore
    {
        const err = _shouldNotContinue3(client);
        if (err) {
            return callback(err);
        }
    }
    const firstError = allErrors.find((e) => e != null);
    callback(firstError);
}
